{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "Migrating from Classical Machine Learning to Quantum Machine Learning: An overview and case study on Drug Discovery",
      "provenance": [],
      "collapsed_sections": []
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "gXXkON3VV2Hn",
        "colab_type": "text"
      },
      "source": [
        "# Migrating from Classical Machine Learning to Quantum Machine Learning: An overview and case study on Drug Discovery\n",
        "\n",
        "\n",
        "Yash Patel , Manipal University Jaipur , yash.169105220@muj.manipal.edu\n",
        "\n",
        "\n",
        "Carlos Toxtli , West Virginia University , carlos.toxtli@mail.wvu.edu , https://www.carlostoxtli.com/#quantum"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "0gdeTPn7VJal",
        "colab_type": "text"
      },
      "source": [
        "## ABSTRACT"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Z0MDqJXiRTja",
        "colab_type": "text"
      },
      "source": [
        "Recent advances in Quantum Computing have opened up new dimensions in various fields for their use. This encourages teams in different fields to explore how their existing models could be ported to a quantum approach. Getting started in this field is not a straightforward task and requires domain knowledge in quantum computing. In this work, we explore how to migrate an existing model to a quantum approach. In order to demonstrate the process, we focus on the Drug Discovery field where the capabilities of Quantum Computers can be used to the fullest. Drug discovery is the process of discovering and designing new drugs and finding their potential protein targets. Several Machine Learning (ML) approaches have been used for Drug Discovery, but Quantum Machine Learning can be used to find atypical patterns in the data which Classical Machine Learning algorithms fail to find. Current approaches for Drug discovery usually take a long time and require significative resources. Quantum Machine Learning algorithms can lead to quicker discoveries. Here we discuss some methods that are used in the Drug Discovery pipeline, mainly focusing on Drug-Target Identification, along with Machine Learning and their migration to Quantum Machine Learning equivalent approaches.\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "oNJX09quVQU_",
        "colab_type": "text"
      },
      "source": [
        "## INTRODUCTION"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "y2rkrYWRR-Lu",
        "colab_type": "text"
      },
      "source": [
        "Drug discovery is the process where new drugs are discovered or created. Drug discovery mainly involves chemistry, pharmacology, and clinical services. Cost of Drug Development is the total cost needed from drug discovery to clinical and approval. Some costs, such as out-of-pocket expenses remain unaccounted in the total cost. Also, the time duration for drug development is usually long, generally more than ten years for a single drug, which carries a huge cost. Drug development mainly has four steps: Target Identification and Validation, Compound Screening and Lead Discovery, Preclinical Development, and Clinical Development. Drug discovery comprises of Target Identification and validation along with compound screening. Drug discovery amounts to the maximum utilization of resources. The process can involve scientists to determine germs, viruses, and bacteria that cause a specific disease or illness. The time frame can range from 3–20 years, and costs can range between several million to billions of dollars. \n",
        "\n",
        "Target Identification is the initial and most important process in the drug discovery pipeline. It concerns collecting effects of a certain drug on a particular biological entity. The biological entity which is affected by a drug is termed as a target, generally a protein sequence. Identification of suitable targets and their corresponding validations increases the chances of success during the discovery process and also allows one to foresee the effects and side effects associated with the modulation of the target. \n",
        "\n",
        "Compound screening and lead discovery comprise of HTS (high throughput screening), which is an experimentation method for screening of compounds to check for any activity against a biological target. This process over the years has become automated using certain applications that use score functions that provide ranks depending on the amount of activity.\n",
        "\n",
        "Recent advances in Artificial Intelligence (AI) has led to an increase in its applications to solve new problems. Application of artificial intelligence in drug discovery is a topic that is being explored, and a substantial amount of work has been done in the past few years. This has lead to an increase in the application of AI in many pharmaceutical companies. Due to the incorporation of AI by pharmaceutical companies, many datasets have been curated in the past few years such as DrugBank, PubChem, ZINC, ChEMBL, and many more. These datasets have compounds represented in the form of SMILES which is a popular line notation of representing chemical structures. SMILES are then converted to molecular descriptors which are numerical vectors (feature vectors) that portray certain characteristics of a compound. These are generated using various geometrical, structural, and physicochemical properties. Few libraries have been developed for forming feature vectors from SMILES. Machine Learning algorithms are then used on the extracted feature vectors.\n",
        "\n",
        "Advancements in Quantum Computing have enabled to implement certain Machine Learning techniques on Quantum Computers. Quantum Computers by IBM, DWave, Rigetti are available for developers and researchers to be used, but the use is restricted. Toolkits such as PennyLane, Strawberry Fields, qiskit can be used to develop quantum computer algorithms. These toolkits have enabled developers to implement machine learning techniques to quantum computers and find new applications for their usage.\n",
        "\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "lY6Wk6BwVUDe",
        "colab_type": "text"
      },
      "source": [
        "## ML TO QML TRANSITION THEORY"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "RvU7JOXtS64o",
        "colab_type": "text"
      },
      "source": [
        "Machine Learning is the study of algorithms that use statistical models to find patterns in data without providing an explicit set of rules. Machine Learning parses through the provided data, learns from it, and tries to predict the state of any new data provided. Machine Learning algorithms improve their performance if the quantity of data is increased and the data is diversified. Hence, these algorithms are best used for huge datasets. Machine Learning approaches are broadly classified into two types: Supervised and Unsupervised Learning. For the context of the case study, we relied on Supervised Learning algorithms, mainly Support Vector Machines (SVMs) and Random Forest. It has been found and stated by many researchers that SVMs and Random Forest perform the best for target identification while using the feature-based approach.\n",
        "\n",
        "Support Vector Machine is a supervised Machine Learning algorithm that is widely used for both regression and classification tasks. SVM's plot the features in a n-dimensional space with each value being a particular coordinate in the plane. These points are separated by a hyperplane, splitting them into different classes. Random Forest uses a different approach to classification. Random forest is a collection of many Decision Trees, which are graph-like structures, a model of decisions and their possible outcomes, resource cost, and utility. Both of the algorithms are best suitable for classification and regression purposes. Various case studies and research outcomes have shown that these two methods are best suitable for Drug Target Identification studies.\n",
        "\n",
        "Deep learning algorithms have been widely used in recent years, demonstrating good results for applications in different fields of study. Deep Neural Networks are Artificial Neural Networks that have multiple hidden layers. The basic building block of a neural network is a neuron which is nothing but an activation function whose coefficients are determined during the training process. A feed-forward DNN can be considered as the most basic DNN architecture, having multiple hidden layers. These are widely used for drug target identification purposes because of their ability to predict multiple numbers of tasks in a single model. We have used a simple feed-forward DNN to compare its results with the QNN counterpart.\n",
        "Classical computers work on the phenomenon of bit manipulation based on the classical gates such as the XOR, NOT, AND etc. But the Quantum Computer use a different set of gates called quantum gates. A quantum gate is a basic quantum circuit that runs on a smaller number of qubits. Quantum gates possess the property of reversibility which the classical gates lack. This enables the quantum gates to retain information throughout the operation. It is though possible to perform classical computing using the quantum reversible gates. In a classical computer, the information is represented in the form of 0's and 1's as bits, whereas the quantum computer stores information in the form of qubits. Quantum computers leverage special phenomenon of superposition which allows a qubit to be in two states simultaneously until the value of the qubit is explicitly measured. Also, another feature that quantum gates are capable of using is of quantum entanglement. Qubits that are entangled on their way into the quantum gate remain entangled on the way out, keeping their information safely sealed throughout the transition. Quantum gates are conventionally represented in the form of unitary matrices. The number of qubits in the input and output of the gate must be equal; a gate which acts on n qubits is represented by a 2^n\\times 2^n unitary matrix.\n",
        "\n",
        "Quantum Computing is the process of manipulation of qubits to process information. The ability of qubits to be in superposition principle leads to a substantial speedup in the amount of information that can be processed. Quantum Machine Learning is the domain where the principles of quantum computing are leveraged to develop algorithms that can be used to solve the typical machine learning algorithms. This can be achieved by adapting the classical algorithms to run on a quantum computer. In theory, we can call quantum machine learning as a method of quantum information processing that learns input-output relations from data provided during the training phase using a quantum function or a quantum approach to learn the data and finding patterns. There are still questions and discussions going on the efficiency of such algorithms on different types of data. There are still methods to be discovered to perfect the art of using quantum computers to solve machine learning problems. Quantum Machine Learning can also be used to analyze quantum data. Quantum Machine Learning typically requires one to encode the given classical data set into a quantum computer to make it accessible for quantum information processing. This encoded data can be processed using a quantum circuit to analyze it. \n",
        "\n",
        "Certain Quantum Machine Learning approaches have already been developed, and many still are under development. One such algorithm is the Quantum Support Vector Machines. It is available in the qiskit suite for direct implementation in python. Machine Learning represents data in a matrix format and is the basis for quantum information processing. Quantum machine learning can take two forms: where classical machine learning algorithms are transformed into their quantum counterparts; to be implemented on a quantum information processor, or taking some of the computationally expensive classical machine learning subroutines and implementing them on the quantum computer. In a Quantum Support Vector Machine algorithm, the classical data is converted to the quantum state for each training example. This is then followed by hyperplane parameter optimization.\n",
        "\n",
        "Neural Networks are a series of algorithms that try to recognize underlying relationships in a set of data that mimics the process in the human brain. Neural networks are multi-layer networks of neurons that we use to classify things, make predictions. Recent advances in Quantum Computing have lead to the implementation of neural networks using the principles of Quantum Mechanics and also to be deployed to run in a Quantum Computer. Quantum Neural Networks comprise a quantum perceptron as the basic unit in the network. A quantum perceptron must satisfy linear and non-linear properties because Quantum Computing uses a combination of linear and non-linear properties of Quantum Mechanics to function. QNN operates much like a classical ANN composed of several layers of perceptrons – an input layer, one or more hidden layers, and an output layer. Each layer is fully connected to the previous layer. Each hidden layer computes a weighted sum of the outputs of the previous layer. If this is sum above a threshold, the node goes high; otherwise, it stays low. The output layer does the same thing as the hidden layer(s), except that it also checks its accuracy against the target output of the network. The network as a whole computes a function by checking which output bit is high. There are no checks to make sure exactly one output is high.\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "q6jjqwN9VYXs",
        "colab_type": "text"
      },
      "source": [
        "## Case Study: Drug Discovery (Drug Target Identification)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "FbNZ6WRIwDdD",
        "colab_type": "text"
      },
      "source": [
        "Drug discovery is basically the discovery of new drugs. Drug Target identification is the process of identifying an entity on which the drug compound goes and binds to, usually a protein or gene. Drug research and development starts with the identification of biomolecular targets; this is the preliminary choice, as choosing a target will result in the modulation of the disease. After the target is identified, it then validated by physiological models. Such validation is nowadays is performed using computer software where each drug that is validated is awarded a score. Though the final validation comes only after clinical tests on subjects, have been performed.\n",
        "Drug Target Identification (DTI) is both a significant step and a challenge to be solved in the Drug Discovery pipeline and has attracted the attention of researches in the past few years. Traditional methods of DTI are labor-intensive and also resources intensive; hence, the focus is now being shifted to computational methods for drug target identification. Machine Learning methods are being explored for Drug Target Identification because of the ability to learn from datasets and also find patterns in the underlying data. \n",
        "\n",
        "For the purpose of the case study, we have used the DrugBank (https://www.drugbank.ca/) dataset. Drugbank is a freely available dataset containing drug and drug target information. It compromises of detailed drug data as well as detailed target data. The latest release of the Drugbank (version 5.1.4) compromises of 13,363 drug entries including 2,606 approved small molecule drugs, 1,299 approved biotech (protein/peptide) drugs, 130 nutraceuticals, and over 6,311 experimental drugs. Additionally, 5,164 non-redundant protein (i.e. drug target/enzyme/transporter/carrier) sequences are linked to these drug entries. Each DrugCard entry contains more than 200 data fields with half of the information being devoted to drug/chemical data and the other half devoted to drug target or protein data.\n",
        "\n",
        " Machine Learning methods for drug target identification can be classified into three types: feature-based approach, similarity-based approach, and other approaches. The input to the machine learning model in a feature-based approach is a feature vector, where each value of the feature vector represents an essential chemical or biological property of the drug or the protein. Feature vectors are generated by concatenating the molecular descriptors of the drug compound and the protein descriptors of the target sequence. This can be passed as an input to a standard machine learning model such as Support Vector Machines or Random Forest. In a similarity-based approach, a similarity matrix is generated where the (x,y)th element of the matrix is the similarity of drug x and drug y. Similarly, a target similarity matrix is generated by a protein sequence alignment. Other approaches use different information about the drug, such as the pharmacological information or the biomedical information. From these drug-protein relations are extracted. There might be cases where this information might not be the real drug-target interaction.\n",
        "\n",
        "For implementing the classical Machine Learning solution, we used the feature-based approach, as this data can also be used easily with the Quantum Machine Learning algorithms. Some of the bioinformatics libraries that we used are provided in table.\n",
        "\n",
        "![](https://docs.google.com/drawings/d/e/2PACX-1vQLcZ5H5Ogw8iDzV7aZehcC9zJt5qPbKwITsmZ_p6SqI70HBC6r13qHXU77O8B47vjRHWpaK_Dp30Oj/pub?w=960&h=720)\n",
        "\n",
        "These libraries were then used to generate the molecular and the protein descriptors of the drug and targets, respectively. These generated molecular descriptors provide information about a certain property of the drug or protein compound. The molecular descriptors and the protein descriptors are then appended to form a feature vector. For example, the size of the molecular descriptor vector be 1xn, and the size of the protein descriptor vector be 1xm, then the feature vector will have the dimensionality of 1x(n+m). This method is applied for positive drug target pair as well as a negative drug target pair. A positive drug target pair is the one where the drug combines to a particular protein sequence. For example, we have a drug A, and it combines to 4 proteins or targets in the dataset, then we have 4 positive drug target pairs and rest of the proteins will form a negative drug target pair with drug A. Positive drug target pair is labeled as one and the negative drug target pair is labeled as zero. This collection of feature vectors is then used as the input to SVM and Random Forest machine learning methods. Also, we implemented a feed-forward DNN for the same classification task.\n",
        "\n",
        "\n",
        "The recent advances in Quantum Computing and Quantum Computing have many developers, and researchers try solving real-world problems using technology. We also had a similar kind of motivation to try out a Quantum Machine Learning algorithm in the domain of drug discovery. There are four different approaches to combine the discipline of quantum computing and machine learning shown in the figure below.\n",
        "\n",
        "\n",
        "![](https://upload.wikimedia.org/wikipedia/commons/thumb/1/1b/Qml_approaches.tif/lossless-page1-296px-Qml_approaches.tif.png)\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "AU37eMgYXNN_",
        "colab_type": "text"
      },
      "source": [
        "### Methodology"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "cYKrl9AIXO8o",
        "colab_type": "text"
      },
      "source": [
        "We are using the Classical Quantum approach where the data is of classical nature and the algorithm to be used is of Quantum nature. In this type of approach, the classical data is converted to quantum data. This enables the implementation of the quantum algorithms on real-world quantum computers. Also, the ability of quantum computers to leverage the phenomenon of quantum mechanics such as superposition and entanglement made us want to try out implementing the drug target identification using machine learning on quantum computers. Also, since quantum gates have the ability to be reversible, it makes them fast to process data because the previous state of the machine is known and all of the information is retained.\n",
        "\n",
        "Hypothesis- \n",
        "After comparing the way classical and quantum computer process information, we hypothesize that machine learning on quantum computers will lead to competitive results (±10%) in comparison with the classical approaches.\n",
        "\n",
        "Limitations - \n",
        "While using the toolkits for implementing quantum computing, we encountered various problems while using the libraries. While shifting from the classical approach to the quantum approach, we had to reduce the number of features that we could select to train the algorithm. This was mainly due to the lack of quantum computing power available for commercial use. On IBM-Q experience services, we could only get a maximum of 32 qubits to be used for simulation purposes and a maximum of 16 qubits to run the algorithm on a real quantum device. In both of the QSVM and QNN approaches, the number of qubits needs to be equal to the number of features that are to be trained. Our initial feature vector used in the classical machine learning approaches for a drug-target pair was of size 1x1821, this was then to be reduced to the number of qubits that we wanted to use.\n",
        "\n",
        "To reduce the number of features, we had to select the highest-ranking features from the given feature vector. This was implemented using the Principle Component Analysis (PCA). PCA is a technique used for dimensionality reduction. Dimensionality reduction is reducing the dimensions of the feature space. Feature elimination by dimensionality reduction helps to retain the simplicity of the data and also maintains the interpretability of your variables. One drawback of PCA is that we are also eliminating any benefits that the original features would bring as compared to the reduced number of dimensions. This transformation is defined in such a way that the first principal component has the largest possible variance (that is, accounts for as much of the variability in the data as possible), and each succeeding component, in turn, has the highest variance possible under the constraint that it is orthogonal to the preceding components. The resulting vectors (each being a linear combination of the variables and containing n observations) are an uncorrelated orthogonal basis set. PCA is sensitive to the relative scaling of the original variables. Also, we had to scale the data in the range [-1 , 1]. The reduction of sale will lead to changing some important descriptor values, which in their original values would have led to the finding of different patterns of correlations."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "ZLrsIPpUXTc-",
        "colab_type": "text"
      },
      "source": [
        "### Simulators"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "HUCdaDteWbXK",
        "colab_type": "text"
      },
      "source": [
        "Quantum Simulators are controllable quantum systems that can be used to simulate other quantum systems. Quantum simulators can help us tackle problems of quantum computing on a classical computer using an interface where quantum data is converted to classical and vice-versa. They form a bridge between the Classical-Classical algorithms and Classical-Quantum or Quantum-Classical algorithms. This data then can be used to implement Quantum Computing algorithms on a classical computer. Many toolkits are available to implement quantum simulators such as the qiskit by IBM, Pennylane by Xanadu, and many more. Quantum Simulators behave like real-world quantum computers, but they are run on classical machines."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "zt995w1tXWfo",
        "colab_type": "text"
      },
      "source": [
        "### Results"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "S44OgmQNVxvD",
        "colab_type": "text"
      },
      "source": [
        "In our initial hypothesis, we were expecting competitive results for the QSVM and QNN implementations when compared with their classical counterparts i.e., SVM, and DNN. But the results were not competitive at all; the QSVM and QNN did not perform even close to their classical counterparts. This might have been due to the limitations of the number of features that one can select when implementing a quantum computing approach because of the limitation of the number of qubits one can use. Quantum supremacy is still quite far away, but we can hope that once we achieve quantum supremacy, we can expect to see better results for the drug target identification using Quantum Computing algorithms.\n",
        "\n",
        "![](https://docs.google.com/drawings/d/e/2PACX-1vTPM9zLsTxotQpA5_JcMPaWhL8GE26do4WKY377rM-rGPvLdpkRKS-NvBL_3vyTS_pGSyio7oLRit75/pub?w=792&h=373)\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "mLtsA4cyVelD",
        "colab_type": "text"
      },
      "source": [
        "## DISCUSSION"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "N4FmJKBigISu",
        "colab_type": "text"
      },
      "source": [
        "In this work, we focus on covering a type of application that promises to have impactful results through the use of quantum computing. It is important to mention that we only focus on a single subarea of drug discovery that is Drug-Target Identification and other sub-areas such as Hit search or Lead search could also have a great impact. Future work can explore the algorithm migration processes for other drug discovery techniques and also for other fields of study.\n",
        "\n",
        "We did not reject the null hypothesis of considering the quantum results as competitive. It was mainly due to the limited data that the algorithms could process due to the limited number of qubits. The implemented QSVM algorithm used each qubit to process each feature, so the number of features was limited by the device that executed it, it ranged from 5 to 32 qubits depending on the IBM Q device. The same limitation affected the neural network implementation by using the Xanadu PenyLane library over a Rigetti Forest environment with ten qubits. Fairer experimentation would be performed by using the same number of features as the classical approach, but in general, the drug discovery subareas are highly demanding in features.\n",
        "\n",
        "Although the results were not competitive, there are some lessons learned that could reduce the learning curve of future implementations. Techniques that allow the best selection and reduction of features are indispensable for this limited environments. The quantum machines available in the cloud have a system of task queuing, so executions can take a long time, not because of the processing time, but because of the ability of these devices to prepare and dispatch the queued processes. It is important to note that emulators consume large amounts of memory, which depends on the information entered. It is important to make tests with different volumes of data to prevent memory saturation.\n",
        "\n",
        "It is not yet clear what type of implementations are the most benefited from the use of quantum machine learning. Future work can perform a comparison of techniques by different areas and find which front is closest to the quantum supremacy. This can lead future research work to create a simple baseline for experimentation. This type of baselines is well known in other areas such as IRIS for machine learning, MNIST for character recognition or CIFAR for classification of objects in images.\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "nORsWtMpXjgA",
        "colab_type": "text"
      },
      "source": [
        "## CONCLUSIONS"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "49u75kx6gYac",
        "colab_type": "text"
      },
      "source": [
        "Quantum computing is an area that promises to solve the great problems of humanity by achieving a performance far superior to today's computers. Although no cases have been demonstrated in which quantum computing exceeds the classical computing in a specific task (quantum supremacy), scientists in this area ensure that we are close. In this work, we set out to understand how these types of algorithms work and how it is possible to convert conventional to quantum algorithms. To demonstrate the process of transforming a solution, we focus on a case study that promises impactful results that is drug discovery. The subarea of drug discovery in which we focus on was Drug-Target Identification. For this type of problem, classification algorithms are required. In classical computing, these problems are usually solved with machine learning algorithms. To demonstrate how to apply machine learning algorithms for quantum machine learning, we experimented with a conventional machine learning method (Support Vector Machine) and a deep learning method (Multi-Layer Perceptron). In this paper, we show step by step the process for both the conventional and quantum solutions, from the preprocessing of information to training and evaluation. The algorithms were designed to run in variational circuits. The experiments were performed in quantum computing simulators and on quantum cloud services. The results were superior in the classic solutions as expected. The contribution of this work focuses on documenting the migration process, showing the requirements, and the limitations of current quantum systems. \n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "S_0koxahwljj",
        "colab_type": "text"
      },
      "source": [
        "# CODE\n",
        "\n",
        "Interactive paper"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "CIEM4J8o52WD",
        "colab_type": "text"
      },
      "source": [
        "## VM Setup"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "iTdXDxSO58yM",
        "colab_type": "code",
        "outputId": "a5dbd0cb-733e-4172-e290-b3d4bb907670",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        }
      },
      "source": [
        "%%time\n",
        "!wget -c https://repo.continuum.io/archive/Anaconda3-5.1.0-Linux-x86_64.sh\n",
        "!chmod +x Anaconda3-5.1.0-Linux-x86_64.sh\n",
        "!bash ./Anaconda3-5.1.0-Linux-x86_64.sh -b -f -p /usr/local\n",
        "import sys\n",
        "sys.path.append('/usr/local/lib/python3.6/site-packages/')\n",
        "!conda install -y --prefix /usr/local -c rdkit python=3.6\n",
        "!conda install -y --prefix /usr/local -c pip python=3.6\n",
        "!pip3 install mordred\n",
        "!pip3 install oddt pubchempy\n",
        "!pip3 install biopython\n",
        "!conda install -y -c conda-forge rdkit\n",
        "!apt install -y  openbabel\n",
        "!pip install biopython\n",
        "!conda uninstall -y sympy\n",
        "!pip install qiskit\n",
        "!python -m pip install pennylane"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "--2020-06-03 06:03:33--  https://repo.continuum.io/archive/Anaconda3-5.1.0-Linux-x86_64.sh\n",
            "Resolving repo.continuum.io (repo.continuum.io)... 104.18.201.79, 104.18.200.79, 2606:4700::6812:c94f, ...\n",
            "Connecting to repo.continuum.io (repo.continuum.io)|104.18.201.79|:443... connected.\n",
            "HTTP request sent, awaiting response... 301 Moved Permanently\n",
            "Location: https://repo.anaconda.com/archive/Anaconda3-5.1.0-Linux-x86_64.sh [following]\n",
            "--2020-06-03 06:03:33--  https://repo.anaconda.com/archive/Anaconda3-5.1.0-Linux-x86_64.sh\n",
            "Resolving repo.anaconda.com (repo.anaconda.com)... 104.16.130.3, 104.16.131.3, 2606:4700::6810:8203, ...\n",
            "Connecting to repo.anaconda.com (repo.anaconda.com)|104.16.130.3|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 577996269 (551M) [application/x-sh]\n",
            "Saving to: ‘Anaconda3-5.1.0-Linux-x86_64.sh’\n",
            "\n",
            "Anaconda3-5.1.0-Lin 100%[===================>] 551.22M   227MB/s    in 2.4s    \n",
            "\n",
            "2020-06-03 06:03:36 (227 MB/s) - ‘Anaconda3-5.1.0-Linux-x86_64.sh’ saved [577996269/577996269]\n",
            "\n",
            "PREFIX=/usr/local\n",
            "installing: python-3.6.4-hc3d631a_1 ...\n",
            "Python 3.6.4 :: Anaconda, Inc.\n",
            "installing: ca-certificates-2017.08.26-h1d4fec5_0 ...\n",
            "installing: conda-env-2.6.0-h36134e3_1 ...\n",
            "installing: intel-openmp-2018.0.0-hc7b2577_8 ...\n",
            "installing: libgcc-ng-7.2.0-h7cc24e2_2 ...\n",
            "installing: libgfortran-ng-7.2.0-h9f7466a_2 ...\n",
            "installing: libstdcxx-ng-7.2.0-h7a57d05_2 ...\n",
            "installing: bzip2-1.0.6-h9a117a8_4 ...\n",
            "installing: expat-2.2.5-he0dffb1_0 ...\n",
            "installing: gmp-6.1.2-h6c8ec71_1 ...\n",
            "installing: graphite2-1.3.10-hf63cedd_1 ...\n",
            "installing: icu-58.2-h9c2bf20_1 ...\n",
            "installing: jbig-2.1-hdba287a_0 ...\n",
            "installing: jpeg-9b-h024ee3a_2 ...\n",
            "installing: libffi-3.2.1-hd88cf55_4 ...\n",
            "installing: libsodium-1.0.15-hf101ebd_0 ...\n",
            "installing: libtool-2.4.6-h544aabb_3 ...\n",
            "installing: libxcb-1.12-hcd93eb1_4 ...\n",
            "installing: lzo-2.10-h49e0be7_2 ...\n",
            "installing: mkl-2018.0.1-h19d6760_4 ...\n",
            "installing: ncurses-6.0-h9df7e31_2 ...\n",
            "installing: openssl-1.0.2n-hb7f436b_0 ...\n",
            "installing: patchelf-0.9-hf79760b_2 ...\n",
            "installing: pcre-8.41-hc27e229_1 ...\n",
            "installing: pixman-0.34.0-hceecf20_3 ...\n",
            "installing: tk-8.6.7-hc745277_3 ...\n",
            "installing: unixodbc-2.3.4-hc36303a_1 ...\n",
            "installing: xz-5.2.3-h55aa19d_2 ...\n",
            "installing: yaml-0.1.7-had09818_2 ...\n",
            "installing: zlib-1.2.11-ha838bed_2 ...\n",
            "installing: glib-2.53.6-h5d9569c_2 ...\n",
            "installing: hdf5-1.10.1-h9caa474_1 ...\n",
            "installing: libedit-3.1-heed3624_0 ...\n",
            "installing: libpng-1.6.34-hb9fc6fc_0 ...\n",
            "installing: libssh2-1.8.0-h9cfc8f7_4 ...\n",
            "installing: libtiff-4.0.9-h28f6b97_0 ...\n",
            "installing: libxml2-2.9.7-h26e45fe_0 ...\n",
            "installing: mpfr-3.1.5-h11a74b3_2 ...\n",
            "installing: pandoc-1.19.2.1-hea2e7c5_1 ...\n",
            "installing: readline-7.0-ha6073c6_4 ...\n",
            "installing: zeromq-4.2.2-hbedb6e5_2 ...\n",
            "installing: dbus-1.12.2-hc3f9b76_1 ...\n",
            "installing: freetype-2.8-hab7d2ae_1 ...\n",
            "installing: gstreamer-1.12.4-hb53b477_0 ...\n",
            "installing: libcurl-7.58.0-h1ad7b7a_0 ...\n",
            "installing: libxslt-1.1.32-h1312cb7_0 ...\n",
            "installing: mpc-1.0.3-hec55b23_5 ...\n",
            "installing: sqlite-3.22.0-h1bed415_0 ...\n",
            "installing: curl-7.58.0-h84994c4_0 ...\n",
            "installing: fontconfig-2.12.4-h88586e7_1 ...\n",
            "installing: gst-plugins-base-1.12.4-h33fb286_0 ...\n",
            "installing: alabaster-0.7.10-py36h306e16b_0 ...\n",
            "installing: asn1crypto-0.24.0-py36_0 ...\n",
            "installing: attrs-17.4.0-py36_0 ...\n",
            "installing: backports-1.0-py36hfa02d7e_1 ...\n",
            "installing: beautifulsoup4-4.6.0-py36h49b8c8c_1 ...\n",
            "installing: bitarray-0.8.1-py36h14c3975_1 ...\n",
            "installing: boto-2.48.0-py36h6e4cd66_1 ...\n",
            "installing: cairo-1.14.12-h77bcde2_0 ...\n",
            "installing: certifi-2018.1.18-py36_0 ...\n",
            "installing: chardet-3.0.4-py36h0f667ec_1 ...\n",
            "installing: click-6.7-py36h5253387_0 ...\n",
            "installing: cloudpickle-0.5.2-py36_1 ...\n",
            "installing: colorama-0.3.9-py36h489cec4_0 ...\n",
            "installing: contextlib2-0.5.5-py36h6c84a62_0 ...\n",
            "installing: dask-core-0.16.1-py36_0 ...\n",
            "installing: decorator-4.2.1-py36_0 ...\n",
            "installing: docutils-0.14-py36hb0f60f5_0 ...\n",
            "installing: entrypoints-0.2.3-py36h1aec115_2 ...\n",
            "installing: et_xmlfile-1.0.1-py36hd6bccc3_0 ...\n",
            "installing: fastcache-1.0.2-py36h14c3975_2 ...\n",
            "installing: filelock-2.0.13-py36h646ffb5_0 ...\n",
            "installing: glob2-0.6-py36he249c77_0 ...\n",
            "installing: gmpy2-2.0.8-py36hc8893dd_2 ...\n",
            "installing: greenlet-0.4.12-py36h2d503a6_0 ...\n",
            "installing: heapdict-1.0.0-py36_2 ...\n",
            "installing: idna-2.6-py36h82fb2a8_1 ...\n",
            "installing: imagesize-0.7.1-py36h52d8127_0 ...\n",
            "installing: ipython_genutils-0.2.0-py36hb52b0d5_0 ...\n",
            "installing: itsdangerous-0.24-py36h93cc618_1 ...\n",
            "installing: jdcal-1.3-py36h4c697fb_0 ...\n",
            "installing: lazy-object-proxy-1.3.1-py36h10fcdad_0 ...\n",
            "installing: llvmlite-0.21.0-py36ha241eea_0 ...\n",
            "installing: locket-0.2.0-py36h787c0ad_1 ...\n",
            "installing: lxml-4.1.1-py36hf71bdeb_1 ...\n",
            "installing: markupsafe-1.0-py36hd9260cd_1 ...\n",
            "installing: mccabe-0.6.1-py36h5ad9710_1 ...\n",
            "installing: mistune-0.8.3-py36_0 ...\n",
            "installing: mkl-service-1.1.2-py36h17a0993_4 ...\n",
            "installing: mpmath-1.0.0-py36hfeacd6b_2 ...\n",
            "installing: msgpack-python-0.5.1-py36h6bb024c_0 ...\n",
            "installing: multipledispatch-0.4.9-py36h41da3fb_0 ...\n",
            "installing: numpy-1.14.0-py36h3dfced4_1 ...\n",
            "installing: olefile-0.45.1-py36_0 ...\n",
            "installing: pandocfilters-1.4.2-py36ha6701b7_1 ...\n",
            "installing: parso-0.1.1-py36h35f843b_0 ...\n",
            "installing: path.py-10.5-py36h55ceabb_0 ...\n",
            "installing: pep8-1.7.1-py36_0 ...\n",
            "installing: pickleshare-0.7.4-py36h63277f8_0 ...\n",
            "installing: pkginfo-1.4.1-py36h215d178_1 ...\n",
            "installing: pluggy-0.6.0-py36hb689045_0 ...\n",
            "installing: ply-3.10-py36hed35086_0 ...\n",
            "installing: psutil-5.4.3-py36h14c3975_0 ...\n",
            "installing: ptyprocess-0.5.2-py36h69acd42_0 ...\n",
            "installing: py-1.5.2-py36h29bf505_0 ...\n",
            "installing: pycodestyle-2.3.1-py36hf609f19_0 ...\n",
            "installing: pycosat-0.6.3-py36h0a5515d_0 ...\n",
            "installing: pycparser-2.18-py36hf9f622e_1 ...\n",
            "installing: pycrypto-2.6.1-py36h14c3975_7 ...\n",
            "installing: pycurl-7.43.0.1-py36hb7f436b_0 ...\n",
            "installing: pyodbc-4.0.22-py36hf484d3e_0 ...\n",
            "installing: pyparsing-2.2.0-py36hee85983_1 ...\n",
            "installing: pysocks-1.6.7-py36hd97a5b1_1 ...\n",
            "installing: pytz-2017.3-py36h63b9c63_0 ...\n",
            "installing: pyyaml-3.12-py36hafb9ca4_1 ...\n",
            "installing: pyzmq-16.0.3-py36he2533c7_0 ...\n",
            "installing: qt-5.6.2-h974d657_12 ...\n",
            "installing: qtpy-1.3.1-py36h3691cc8_0 ...\n",
            "installing: rope-0.10.7-py36h147e2ec_0 ...\n",
            "installing: ruamel_yaml-0.15.35-py36h14c3975_1 ...\n",
            "installing: send2trash-1.4.2-py36_0 ...\n",
            "installing: simplegeneric-0.8.1-py36_2 ...\n",
            "installing: sip-4.18.1-py36h51ed4ed_2 ...\n",
            "installing: six-1.11.0-py36h372c433_1 ...\n",
            "installing: snowballstemmer-1.2.1-py36h6febd40_0 ...\n",
            "installing: sortedcontainers-1.5.9-py36_0 ...\n",
            "installing: sphinxcontrib-1.0-py36h6d0f590_1 ...\n",
            "installing: sqlalchemy-1.2.1-py36h14c3975_0 ...\n",
            "installing: tblib-1.3.2-py36h34cf8b6_0 ...\n",
            "installing: testpath-0.3.1-py36h8cadb63_0 ...\n",
            "installing: toolz-0.9.0-py36_0 ...\n",
            "installing: tornado-4.5.3-py36_0 ...\n",
            "installing: typing-3.6.2-py36h7da032a_0 ...\n",
            "installing: unicodecsv-0.14.1-py36ha668878_0 ...\n",
            "installing: wcwidth-0.1.7-py36hdf4376a_0 ...\n",
            "installing: webencodings-0.5.1-py36h800622e_1 ...\n",
            "installing: werkzeug-0.14.1-py36_0 ...\n",
            "installing: wrapt-1.10.11-py36h28b7045_0 ...\n",
            "installing: xlrd-1.1.0-py36h1db9f0c_1 ...\n",
            "installing: xlsxwriter-1.0.2-py36h3de1aca_0 ...\n",
            "installing: xlwt-1.3.0-py36h7b00a1f_0 ...\n",
            "installing: babel-2.5.3-py36_0 ...\n",
            "installing: backports.shutil_get_terminal_size-1.0.0-py36hfea85ff_2 ...\n",
            "installing: bottleneck-1.2.1-py36haac1ea0_0 ...\n",
            "installing: cffi-1.11.4-py36h9745a5d_0 ...\n",
            "installing: conda-verify-2.0.0-py36h98955d8_0 ...\n",
            "installing: cycler-0.10.0-py36h93f1223_0 ...\n",
            "installing: cytoolz-0.9.0-py36h14c3975_0 ...\n",
            "installing: h5py-2.7.1-py36h3585f63_0 ...\n",
            "installing: harfbuzz-1.7.4-hc5b324e_0 ...\n",
            "installing: html5lib-1.0.1-py36h2f9c1c0_0 ...\n",
            "installing: jedi-0.11.1-py36_0 ...\n",
            "installing: networkx-2.1-py36_0 ...\n",
            "installing: nltk-3.2.5-py36h7532b22_0 ...\n",
            "installing: numba-0.36.2-np114py36hc6662d5_0 ...\n",
            "installing: numexpr-2.6.4-py36hc4a3f9a_0 ...\n",
            "installing: openpyxl-2.4.10-py36_0 ...\n",
            "installing: packaging-16.8-py36ha668100_1 ...\n",
            "installing: partd-0.3.8-py36h36fd896_0 ...\n",
            "installing: pathlib2-2.3.0-py36h49efa8e_0 ...\n",
            "installing: pexpect-4.3.1-py36_0 ...\n",
            "installing: pillow-5.0.0-py36h3deb7b8_0 ...\n",
            "installing: pyqt-5.6.0-py36h0386399_5 ...\n",
            "installing: python-dateutil-2.6.1-py36h88d3b88_1 ...\n",
            "installing: pywavelets-0.5.2-py36he602eb0_0 ...\n",
            "installing: qtawesome-0.4.4-py36h609ed8c_0 ...\n",
            "installing: scipy-1.0.0-py36hbf646e7_0 ...\n",
            "installing: setuptools-38.4.0-py36_0 ...\n",
            "installing: singledispatch-3.4.0.3-py36h7a266c3_0 ...\n",
            "installing: sortedcollections-0.5.3-py36h3c761f9_0 ...\n",
            "installing: sphinxcontrib-websupport-1.0.1-py36hb5cb234_1 ...\n",
            "installing: sympy-1.1.1-py36hc6d1c1c_0 ...\n",
            "installing: terminado-0.8.1-py36_1 ...\n",
            "installing: traitlets-4.3.2-py36h674d592_0 ...\n",
            "installing: zict-0.1.3-py36h3a3bf81_0 ...\n",
            "installing: astroid-1.6.1-py36_0 ...\n",
            "installing: bleach-2.1.2-py36_0 ...\n",
            "installing: clyent-1.2.2-py36h7e57e65_1 ...\n",
            "installing: cryptography-2.1.4-py36hd09be54_0 ...\n",
            "installing: cython-0.27.3-py36h1860423_0 ...\n",
            "installing: datashape-0.5.4-py36h3ad6b5c_0 ...\n",
            "installing: distributed-1.20.2-py36_0 ...\n",
            "installing: get_terminal_size-1.0.0-haa9412d_0 ...\n",
            "installing: gevent-1.2.2-py36h2fe25dc_0 ...\n",
            "installing: imageio-2.2.0-py36he555465_0 ...\n",
            "installing: isort-4.2.15-py36had401c0_0 ...\n",
            "installing: jinja2-2.10-py36ha16c418_0 ...\n",
            "installing: jsonschema-2.6.0-py36h006f8b5_0 ...\n",
            "installing: jupyter_core-4.4.0-py36h7c827e3_0 ...\n",
            "installing: matplotlib-2.1.2-py36h0e671d2_0 ...\n",
            "installing: navigator-updater-0.1.0-py36h14770f7_0 ...\n",
            "installing: nose-1.3.7-py36hcdf7029_2 ...\n",
            "installing: pandas-0.22.0-py36hf484d3e_0 ...\n",
            "installing: pango-1.41.0-hd475d92_0 ...\n",
            "installing: patsy-0.5.0-py36_0 ...\n",
            "installing: pyflakes-1.6.0-py36h7bd6a15_0 ...\n",
            "installing: pygments-2.2.0-py36h0d3125c_0 ...\n",
            "installing: pytables-3.4.2-py36h3b5282a_2 ...\n",
            "installing: pytest-3.3.2-py36_0 ...\n",
            "installing: scikit-learn-0.19.1-py36h7aa7ec6_0 ...\n",
            "installing: wheel-0.30.0-py36hfd4bba0_1 ...\n",
            "installing: astropy-2.0.3-py36h14c3975_0 ...\n",
            "installing: bkcharts-0.2-py36h735825a_0 ...\n",
            "installing: bokeh-0.12.13-py36h2f9c1c0_0 ...\n",
            "installing: flask-0.12.2-py36hb24657c_0 ...\n",
            "installing: jupyter_client-5.2.2-py36_0 ...\n",
            "installing: nbformat-4.4.0-py36h31c9010_0 ...\n",
            "installing: pip-9.0.1-py36h6c6f9ce_4 ...\n",
            "installing: prompt_toolkit-1.0.15-py36h17d85b1_0 ...\n",
            "installing: pylint-1.8.2-py36_0 ...\n",
            "installing: pyopenssl-17.5.0-py36h20ba746_0 ...\n",
            "installing: statsmodels-0.8.0-py36h8533d0b_0 ...\n",
            "installing: dask-0.16.1-py36_0 ...\n",
            "installing: flask-cors-3.0.3-py36h2d857d3_0 ...\n",
            "installing: ipython-6.2.1-py36h88c514a_1 ...\n",
            "installing: nbconvert-5.3.1-py36hb41ffb7_0 ...\n",
            "installing: seaborn-0.8.1-py36hfad7ec4_0 ...\n",
            "installing: urllib3-1.22-py36hbe7ace6_0 ...\n",
            "installing: ipykernel-4.8.0-py36_0 ...\n",
            "installing: odo-0.5.1-py36h90ed295_0 ...\n",
            "installing: requests-2.18.4-py36he2e5f8d_1 ...\n",
            "installing: scikit-image-0.13.1-py36h14c3975_1 ...\n",
            "installing: anaconda-client-1.6.9-py36_0 ...\n",
            "installing: blaze-0.11.3-py36h4e06776_0 ...\n",
            "installing: jupyter_console-5.2.0-py36he59e554_1 ...\n",
            "installing: notebook-5.4.0-py36_0 ...\n",
            "installing: qtconsole-4.3.1-py36h8f73b5b_0 ...\n",
            "installing: sphinx-1.6.6-py36_0 ...\n",
            "installing: anaconda-project-0.8.2-py36h44fb852_0 ...\n",
            "installing: jupyterlab_launcher-0.10.2-py36_0 ...\n",
            "installing: numpydoc-0.7.0-py36h18f165f_0 ...\n",
            "installing: widgetsnbextension-3.1.0-py36_0 ...\n",
            "installing: anaconda-navigator-1.7.0-py36_0 ...\n",
            "installing: ipywidgets-7.1.1-py36_0 ...\n",
            "installing: jupyterlab-0.31.5-py36_0 ...\n",
            "installing: spyder-3.2.6-py36_0 ...\n",
            "installing: _ipyw_jlab_nb_ext_conf-0.1.0-py36he11e457_0 ...\n",
            "installing: jupyter-1.0.0-py36_4 ...\n",
            "installing: anaconda-5.1.0-py36_2 ...\n",
            "installing: conda-4.4.10-py36_0 ...\n",
            "installing: conda-build-3.4.1-py36_0 ...\n",
            "installation finished.\n",
            "WARNING:\n",
            "    You currently have a PYTHONPATH environment variable set. This may cause\n",
            "    unexpected behavior when running the Python interpreter in Anaconda3.\n",
            "    For best results, please verify that your PYTHONPATH only points to\n",
            "    directories of packages that are compatible with the Python interpreter\n",
            "    in Anaconda3: /usr/local\n",
            "Solving environment: - \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\bdone\n",
            "\n",
            "\n",
            "==> WARNING: A newer version of conda exists. <==\n",
            "  current version: 4.4.10\n",
            "  latest version: 4.8.3\n",
            "\n",
            "Please update conda by running\n",
            "\n",
            "    $ conda update -n base conda\n",
            "\n",
            "\n",
            "\n",
            "## Package Plan ##\n",
            "\n",
            "  environment location: /usr/local\n",
            "\n",
            "  added / updated specs: \n",
            "    - python=3.6\n",
            "\n",
            "\n",
            "The following packages will be downloaded:\n",
            "\n",
            "    package                    |            build\n",
            "    ---------------------------|-----------------\n",
            "    setuptools-46.4.0          |           py36_0         646 KB\n",
            "    python-3.6.10              |       h7579374_2        33.9 MB\n",
            "    certifi-2020.4.5.1         |           py36_0         159 KB\n",
            "    libffi-3.3                 |       he6710b0_1          54 KB\n",
            "    libgcc-ng-9.1.0            |       hdf63c60_0         8.1 MB\n",
            "    pip-20.0.2                 |           py36_3         1.9 MB\n",
            "    libedit-3.1.20181209       |       hc058e9b_0         188 KB\n",
            "    xz-5.2.5                   |       h7b6447c_0         438 KB\n",
            "    zlib-1.2.11                |       h7b6447c_3         120 KB\n",
            "    readline-8.0               |       h7b6447c_0         428 KB\n",
            "    ca-certificates-2020.1.1   |                0         132 KB\n",
            "    _libgcc_mutex-0.1          |             main           3 KB\n",
            "    ncurses-6.2                |       he6710b0_1         1.1 MB\n",
            "    openssl-1.1.1g             |       h7b6447c_0         3.8 MB\n",
            "    ld_impl_linux-64-2.33.1    |       h53a641e_7         645 KB\n",
            "    tk-8.6.8                   |       hbc83047_0         3.1 MB\n",
            "    sqlite-3.31.1              |       h62c20be_1         2.0 MB\n",
            "    wheel-0.34.2               |           py36_0          49 KB\n",
            "    libstdcxx-ng-9.1.0         |       hdf63c60_0         4.0 MB\n",
            "    ------------------------------------------------------------\n",
            "                                           Total:        60.8 MB\n",
            "\n",
            "The following NEW packages will be INSTALLED:\n",
            "\n",
            "    _libgcc_mutex:    0.1-main             \n",
            "    ld_impl_linux-64: 2.33.1-h53a641e_7    \n",
            "\n",
            "The following packages will be UPDATED:\n",
            "\n",
            "    ca-certificates:  2017.08.26-h1d4fec5_0 --> 2020.1.1-0             \n",
            "    certifi:          2018.1.18-py36_0      --> 2020.4.5.1-py36_0      \n",
            "    libedit:          3.1-heed3624_0        --> 3.1.20181209-hc058e9b_0\n",
            "    libffi:           3.2.1-hd88cf55_4      --> 3.3-he6710b0_1         \n",
            "    libgcc-ng:        7.2.0-h7cc24e2_2      --> 9.1.0-hdf63c60_0       \n",
            "    libstdcxx-ng:     7.2.0-h7a57d05_2      --> 9.1.0-hdf63c60_0       \n",
            "    ncurses:          6.0-h9df7e31_2        --> 6.2-he6710b0_1         \n",
            "    openssl:          1.0.2n-hb7f436b_0     --> 1.1.1g-h7b6447c_0      \n",
            "    pip:              9.0.1-py36h6c6f9ce_4  --> 20.0.2-py36_3          \n",
            "    python:           3.6.4-hc3d631a_1      --> 3.6.10-h7579374_2      \n",
            "    readline:         7.0-ha6073c6_4        --> 8.0-h7b6447c_0         \n",
            "    setuptools:       38.4.0-py36_0         --> 46.4.0-py36_0          \n",
            "    sqlite:           3.22.0-h1bed415_0     --> 3.31.1-h62c20be_1      \n",
            "    tk:               8.6.7-hc745277_3      --> 8.6.8-hbc83047_0       \n",
            "    wheel:            0.30.0-py36hfd4bba0_1 --> 0.34.2-py36_0          \n",
            "    xz:               5.2.3-h55aa19d_2      --> 5.2.5-h7b6447c_0       \n",
            "    zlib:             1.2.11-ha838bed_2     --> 1.2.11-h7b6447c_3      \n",
            "\n",
            "\n",
            "Downloading and Extracting Packages\n",
            "setuptools 46.4.0: 100% 1.0/1 [00:00<00:00,  3.05it/s]               \n",
            "python 3.6.10: 100% 1.0/1 [00:09<00:00,  9.14s/it]               \n",
            "certifi 2020.4.5.1: 100% 1.0/1 [00:00<00:00, 14.11it/s]\n",
            "libffi 3.3: 100% 1.0/1 [00:00<00:00, 22.06it/s]\n",
            "libgcc-ng 9.1.0: 100% 1.0/1 [00:02<00:00,  2.44s/it]               \n",
            "pip 20.0.2: 100% 1.0/1 [00:00<00:00,  1.11it/s]               \n",
            "libedit 3.1.20181209: 100% 1.0/1 [00:00<00:00, 10.10it/s]\n",
            "xz 5.2.5: 100% 1.0/1 [00:00<00:00,  5.03it/s]               \n",
            "zlib 1.2.11: 100% 1.0/1 [00:00<00:00, 13.86it/s]\n",
            "readline 8.0: 100% 1.0/1 [00:00<00:00,  5.36it/s]               \n",
            "ca-certificates 2020.1.1: 100% 1.0/1 [00:00<00:00, 16.65it/s]\n",
            "_libgcc_mutex 0.1: 100% 1.0/1 [00:00<00:00, 30.33it/s]\n",
            "ncurses 6.2: 100% 1.0/1 [00:00<00:00,  1.01it/s]               \n",
            "openssl 1.1.1g: 100% 1.0/1 [00:01<00:00,  1.26s/it]               \n",
            "ld_impl_linux-64 2.33.1: 100% 1.0/1 [00:00<00:00,  3.42it/s] \n",
            "tk 8.6.8: 100% 1.0/1 [00:01<00:00,  1.10s/it]               \n",
            "sqlite 3.31.1: 100% 1.0/1 [00:00<00:00,  1.61it/s]               \n",
            "wheel 0.34.2: 100% 1.0/1 [00:00<00:00, 18.99it/s]\n",
            "libstdcxx-ng 9.1.0: 100% 1.0/1 [00:01<00:00,  1.23s/it]               \n",
            "Preparing transaction: / \b\b- \b\bdone\n",
            "Verifying transaction: | \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\bdone\n",
            "Executing transaction: - \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\bdone\n",
            "Solving environment: - \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \n",
            "WARNING: The remote server could not find the noarch directory for the\n",
            "requested channel with url: https://conda.anaconda.org/pip\n",
            "\n",
            "It is possible you have given conda an invalid channel. Please double-check\n",
            "your conda configuration using `conda config --show`.\n",
            "\n",
            "If the requested url is in fact a valid conda channel, please request that the\n",
            "channel administrator create `noarch/repodata.json` and associated\n",
            "`noarch/repodata.json.bz2` files, even if `noarch/repodata.json` is empty.\n",
            "$ mkdir noarch\n",
            "$ echo '{}' > noarch/repodata.json\n",
            "$ bzip2 -k noarch/repodata.json\n",
            "\b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\bdone\n",
            "\n",
            "\n",
            "==> WARNING: A newer version of conda exists. <==\n",
            "  current version: 4.4.10\n",
            "  latest version: 4.8.3\n",
            "\n",
            "Please update conda by running\n",
            "\n",
            "    $ conda update -n base conda\n",
            "\n",
            "\n",
            "\n",
            "# All requested packages already installed.\n",
            "\n",
            "Collecting mordred\n",
            "\u001b[?25l  Downloading https://files.pythonhosted.org/packages/93/3d/26c908ece761adafcea06320bf8fe73f4de69979273fb164226dc6038c39/mordred-1.2.0.tar.gz (128kB)\n",
            "\u001b[K     |████████████████████████████████| 133kB 2.8MB/s \n",
            "\u001b[?25hRequirement already satisfied: six==1.* in /usr/local/lib/python3.6/dist-packages (from mordred) (1.12.0)\n",
            "Requirement already satisfied: numpy==1.* in /usr/local/lib/python3.6/dist-packages (from mordred) (1.18.4)\n",
            "Requirement already satisfied: networkx==2.* in /usr/local/lib/python3.6/dist-packages (from mordred) (2.4)\n",
            "Requirement already satisfied: decorator>=4.3.0 in /usr/local/lib/python3.6/dist-packages (from networkx==2.*->mordred) (4.4.2)\n",
            "Building wheels for collected packages: mordred\n",
            "  Building wheel for mordred (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for mordred: filename=mordred-1.2.0-cp36-none-any.whl size=176721 sha256=0a14c0a7c735000ac07c857f8108bc8d9db56f241d6cf46cd9e22ed045ccb7f3\n",
            "  Stored in directory: /root/.cache/pip/wheels/ac/74/3f/2fd81b1187013f2eadb15620434813f1824c4c03b7bd1f94f6\n",
            "Successfully built mordred\n",
            "Installing collected packages: mordred\n",
            "Successfully installed mordred-1.2.0\n",
            "Collecting oddt\n",
            "\u001b[?25l  Downloading https://files.pythonhosted.org/packages/5c/98/0e6e00bf226b2f005197e1913f1572e91382f882cb89dd574af85c42734b/oddt-0.7.tar.gz (2.3MB)\n",
            "\u001b[K     |████████████████████████████████| 2.3MB 2.8MB/s \n",
            "\u001b[?25hCollecting pubchempy\n",
            "  Downloading https://files.pythonhosted.org/packages/aa/fb/8de3aa9804b614dbc8dc5c16ed061d819cc360e0ddecda3dcd01c1552339/PubChemPy-1.0.4.tar.gz\n",
            "Collecting numpydoc\n",
            "\u001b[?25l  Downloading https://files.pythonhosted.org/packages/3a/43/2402fd1f63992a52f88e3b169d59674617013bf7f1ad0cd7d842eafd0c58/numpydoc-1.0.0-py3-none-any.whl (47kB)\n",
            "\u001b[K     |████████████████████████████████| 51kB 7.1MB/s \n",
            "\u001b[?25hRequirement already satisfied: six in /usr/local/lib/python3.6/dist-packages (from oddt) (1.12.0)\n",
            "Requirement already satisfied: numpy>=1.11 in /usr/local/lib/python3.6/dist-packages (from oddt) (1.18.4)\n",
            "Requirement already satisfied: scipy>=0.17 in /usr/local/lib/python3.6/dist-packages (from oddt) (1.4.1)\n",
            "Requirement already satisfied: scikit-learn>=0.18 in /usr/local/lib/python3.6/dist-packages (from oddt) (0.22.2.post1)\n",
            "Requirement already satisfied: joblib>=0.9.4 in /usr/local/lib/python3.6/dist-packages (from oddt) (0.15.1)\n",
            "Requirement already satisfied: pandas>=0.19.2 in /usr/local/lib/python3.6/dist-packages (from oddt) (1.0.4)\n",
            "Requirement already satisfied: sphinx>=1.6.5 in /usr/local/lib/python3.6/dist-packages (from numpydoc->oddt) (1.8.5)\n",
            "Requirement already satisfied: Jinja2>=2.3 in /usr/local/lib/python3.6/dist-packages (from numpydoc->oddt) (2.11.2)\n",
            "Requirement already satisfied: python-dateutil>=2.6.1 in /usr/local/lib/python3.6/dist-packages (from pandas>=0.19.2->oddt) (2.8.1)\n",
            "Requirement already satisfied: pytz>=2017.2 in /usr/local/lib/python3.6/dist-packages (from pandas>=0.19.2->oddt) (2018.9)\n",
            "Requirement already satisfied: snowballstemmer>=1.1 in /usr/local/lib/python3.6/dist-packages (from sphinx>=1.6.5->numpydoc->oddt) (2.0.0)\n",
            "Requirement already satisfied: docutils>=0.11 in /usr/local/lib/python3.6/dist-packages (from sphinx>=1.6.5->numpydoc->oddt) (0.15.2)\n",
            "Requirement already satisfied: setuptools in /usr/local/lib/python3.6/dist-packages (from sphinx>=1.6.5->numpydoc->oddt) (47.1.1)\n",
            "Requirement already satisfied: sphinxcontrib-websupport in /usr/local/lib/python3.6/dist-packages (from sphinx>=1.6.5->numpydoc->oddt) (1.2.2)\n",
            "Requirement already satisfied: alabaster<0.8,>=0.7 in /usr/local/lib/python3.6/dist-packages (from sphinx>=1.6.5->numpydoc->oddt) (0.7.12)\n",
            "Requirement already satisfied: packaging in /usr/local/lib/python3.6/dist-packages (from sphinx>=1.6.5->numpydoc->oddt) (20.4)\n",
            "Requirement already satisfied: babel!=2.0,>=1.3 in /usr/local/lib/python3.6/dist-packages (from sphinx>=1.6.5->numpydoc->oddt) (2.8.0)\n",
            "Requirement already satisfied: requests>=2.0.0 in /usr/local/lib/python3.6/dist-packages (from sphinx>=1.6.5->numpydoc->oddt) (2.23.0)\n",
            "Requirement already satisfied: imagesize in /usr/local/lib/python3.6/dist-packages (from sphinx>=1.6.5->numpydoc->oddt) (1.2.0)\n",
            "Requirement already satisfied: Pygments>=2.0 in /usr/local/lib/python3.6/dist-packages (from sphinx>=1.6.5->numpydoc->oddt) (2.1.3)\n",
            "Requirement already satisfied: MarkupSafe>=0.23 in /usr/local/lib/python3.6/dist-packages (from Jinja2>=2.3->numpydoc->oddt) (1.1.1)\n",
            "Requirement already satisfied: pyparsing>=2.0.2 in /usr/local/lib/python3.6/dist-packages (from packaging->sphinx>=1.6.5->numpydoc->oddt) (2.4.7)\n",
            "Requirement already satisfied: idna<3,>=2.5 in /usr/local/lib/python3.6/dist-packages (from requests>=2.0.0->sphinx>=1.6.5->numpydoc->oddt) (2.9)\n",
            "Requirement already satisfied: urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1 in /usr/local/lib/python3.6/dist-packages (from requests>=2.0.0->sphinx>=1.6.5->numpydoc->oddt) (1.24.3)\n",
            "Requirement already satisfied: chardet<4,>=3.0.2 in /usr/local/lib/python3.6/dist-packages (from requests>=2.0.0->sphinx>=1.6.5->numpydoc->oddt) (3.0.4)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.6/dist-packages (from requests>=2.0.0->sphinx>=1.6.5->numpydoc->oddt) (2020.4.5.1)\n",
            "Building wheels for collected packages: oddt, pubchempy\n",
            "  Building wheel for oddt (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for oddt: filename=oddt-0.7-py2.py3-none-any.whl size=2409545 sha256=739905b59438e21d9e6a8fdb1d775b9511fdbad661a6b0dce18437920b26cca7\n",
            "  Stored in directory: /root/.cache/pip/wheels/fe/66/9f/cf08a706b47cf8de6cca9b406801acdb3c3908ab481aa4a373\n",
            "  Building wheel for pubchempy (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for pubchempy: filename=PubChemPy-1.0.4-cp36-none-any.whl size=13825 sha256=4e3f800b9f94f5b75371d5343ef38296ce26af43dd92e1880333155eebe4f163\n",
            "  Stored in directory: /root/.cache/pip/wheels/10/4d/51/6b843681a9a5aef35f0d0fbce243de46f85080036e16118752\n",
            "Successfully built oddt pubchempy\n",
            "Installing collected packages: numpydoc, oddt, pubchempy\n",
            "Successfully installed numpydoc-1.0.0 oddt-0.7 pubchempy-1.0.4\n",
            "Collecting biopython\n",
            "\u001b[?25l  Downloading https://files.pythonhosted.org/packages/a8/66/134dbd5f885fc71493c61b6cf04c9ea08082da28da5ed07709b02857cbd0/biopython-1.77-cp36-cp36m-manylinux1_x86_64.whl (2.3MB)\n",
            "\u001b[K     |████████████████████████████████| 2.3MB 2.7MB/s \n",
            "\u001b[?25hRequirement already satisfied: numpy in /usr/local/lib/python3.6/dist-packages (from biopython) (1.18.4)\n",
            "Installing collected packages: biopython\n",
            "Successfully installed biopython-1.77\n",
            "Solving environment: - \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\bdone\n",
            "\n",
            "\n",
            "==> WARNING: A newer version of conda exists. <==\n",
            "  current version: 4.4.10\n",
            "  latest version: 4.8.3\n",
            "\n",
            "Please update conda by running\n",
            "\n",
            "    $ conda update -n base conda\n",
            "\n",
            "\n",
            "\n",
            "## Package Plan ##\n",
            "\n",
            "  environment location: /usr/local\n",
            "\n",
            "  added / updated specs: \n",
            "    - rdkit\n",
            "\n",
            "\n",
            "The following packages will be downloaded:\n",
            "\n",
            "    package                    |            build\n",
            "    ---------------------------|-----------------\n",
            "    pandas-1.0.4               |   py36h830a2c2_0        10.1 MB  conda-forge\n",
            "    numpy-1.18.4               |   py36h7314795_0         5.2 MB  conda-forge\n",
            "    libopenblas-0.3.7          |       h5ec1e0e_6         7.6 MB  conda-forge\n",
            "    pycairo-1.18.0             |py36h1b9232e_1000          74 KB  conda-forge\n",
            "    libcblas-3.8.0             |      14_openblas          10 KB  conda-forge\n",
            "    ca-certificates-2020.4.5.1 |       hecc5488_0         146 KB  conda-forge\n",
            "    certifi-2020.4.5.1         |   py36h9f0ad1d_0         151 KB  conda-forge\n",
            "    python_abi-3.6             |          1_cp36m           4 KB  conda-forge\n",
            "    libblas-3.8.0              |      14_openblas          10 KB  conda-forge\n",
            "    boost-1.66.0               |           py36_1         317 KB  conda-forge\n",
            "    boost-cpp-1.66.0           |                1        18.8 MB  conda-forge\n",
            "    liblapack-3.8.0            |      14_openblas          10 KB  conda-forge\n",
            "    openssl-1.1.1g             |       h516909a_0         2.1 MB  conda-forge\n",
            "    rdkit-2018.03.3            |   py36hf2429ea_1        19.4 MB  conda-forge\n",
            "    ------------------------------------------------------------\n",
            "                                           Total:        64.0 MB\n",
            "\n",
            "The following NEW packages will be INSTALLED:\n",
            "\n",
            "    boost:           1.66.0-py36_1            conda-forge\n",
            "    boost-cpp:       1.66.0-1                 conda-forge\n",
            "    libblas:         3.8.0-14_openblas        conda-forge\n",
            "    libcblas:        3.8.0-14_openblas        conda-forge\n",
            "    liblapack:       3.8.0-14_openblas        conda-forge\n",
            "    libopenblas:     0.3.7-h5ec1e0e_6         conda-forge\n",
            "    pycairo:         1.18.0-py36h1b9232e_1000 conda-forge\n",
            "    python_abi:      3.6-1_cp36m              conda-forge\n",
            "    rdkit:           2018.03.3-py36hf2429ea_1 conda-forge\n",
            "\n",
            "The following packages will be UPDATED:\n",
            "\n",
            "    ca-certificates: 2020.1.1-0                           --> 2020.4.5.1-hecc5488_0     conda-forge\n",
            "    certifi:         2020.4.5.1-py36_0                    --> 2020.4.5.1-py36h9f0ad1d_0 conda-forge\n",
            "    numpy:           1.14.0-py36h3dfced4_1                --> 1.18.4-py36h7314795_0     conda-forge\n",
            "    openssl:         1.1.1g-h7b6447c_0                    --> 1.1.1g-h516909a_0         conda-forge\n",
            "    pandas:          0.22.0-py36hf484d3e_0                --> 1.0.4-py36h830a2c2_0      conda-forge\n",
            "\n",
            "\n",
            "Downloading and Extracting Packages\n",
            "pandas 1.0.4: 100% 1.0/1 [00:04<00:00,  4.30s/it]               \n",
            "numpy 1.18.4: 100% 1.0/1 [00:02<00:00,  2.14s/it]               \n",
            "libopenblas 0.3.7: 100% 1.0/1 [00:02<00:00,  2.67s/it]               \n",
            "pycairo 1.18.0: 100% 1.0/1 [00:00<00:00,  2.51it/s]                \n",
            "libcblas 3.8.0: 100% 1.0/1 [00:00<00:00, 29.57it/s]\n",
            "ca-certificates 2020.4.5.1: 100% 1.0/1 [00:00<00:00, 15.80it/s]\n",
            "certifi 2020.4.5.1: 100% 1.0/1 [00:00<00:00, 14.70it/s]\n",
            "python_abi 3.6: 100% 1.0/1 [00:00<00:00, 24.79it/s]\n",
            "libblas 3.8.0: 100% 1.0/1 [00:00<00:00, 22.38it/s]\n",
            "boost 1.66.0: 100% 1.0/1 [00:00<00:00,  1.74it/s]               \n",
            "boost-cpp 1.66.0: 100% 1.0/1 [00:15<00:00, 15.07s/it]                \n",
            "liblapack 3.8.0: 100% 1.0/1 [00:00<00:00, 16.73it/s]\n",
            "openssl 1.1.1g: 100% 1.0/1 [00:00<00:00,  1.51it/s]               \n",
            "rdkit 2018.03.3: 100% 1.0/1 [00:08<00:00,  8.22s/it]              \n",
            "Preparing transaction: \\ \b\b| \b\b/ \b\bdone\n",
            "Verifying transaction: \\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\bdone\n",
            "Executing transaction: | \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\bdone\n",
            "Reading package lists... Done\n",
            "Building dependency tree       \n",
            "Reading state information... Done\n",
            "The following additional packages will be installed:\n",
            "  libopenbabel4v5\n",
            "The following NEW packages will be installed:\n",
            "  libopenbabel4v5 openbabel\n",
            "0 upgraded, 2 newly installed, 0 to remove and 32 not upgraded.\n",
            "Need to get 2,766 kB of archives.\n",
            "After this operation, 11.8 MB of additional disk space will be used.\n",
            "Get:1 http://archive.ubuntu.com/ubuntu bionic/universe amd64 libopenbabel4v5 amd64 2.3.2+dfsg-3build1 [2,664 kB]\n",
            "Get:2 http://archive.ubuntu.com/ubuntu bionic/universe amd64 openbabel amd64 2.3.2+dfsg-3build1 [102 kB]\n",
            "Fetched 2,766 kB in 3s (990 kB/s)\n",
            "Selecting previously unselected package libopenbabel4v5.\n",
            "(Reading database ... 144467 files and directories currently installed.)\n",
            "Preparing to unpack .../libopenbabel4v5_2.3.2+dfsg-3build1_amd64.deb ...\n",
            "Unpacking libopenbabel4v5 (2.3.2+dfsg-3build1) ...\n",
            "Selecting previously unselected package openbabel.\n",
            "Preparing to unpack .../openbabel_2.3.2+dfsg-3build1_amd64.deb ...\n",
            "Unpacking openbabel (2.3.2+dfsg-3build1) ...\n",
            "Setting up libopenbabel4v5 (2.3.2+dfsg-3build1) ...\n",
            "Setting up openbabel (2.3.2+dfsg-3build1) ...\n",
            "Processing triggers for man-db (2.8.3-2ubuntu0.1) ...\n",
            "Processing triggers for libc-bin (2.27-3ubuntu1) ...\n",
            "/sbin/ldconfig.real: /usr/local/lib/python3.6/dist-packages/ideep4py/lib/libmkldnn.so.0 is not a symbolic link\n",
            "\n",
            "Collecting biopython\n",
            "  Using cached biopython-1.77-cp36-cp36m-manylinux1_x86_64.whl (2.3 MB)\n",
            "Requirement already satisfied: numpy in /usr/local/lib/python3.6/site-packages (from biopython) (1.18.4)\n",
            "Installing collected packages: biopython\n",
            "Successfully installed biopython-1.77\n",
            "Solving environment: - \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\bdone\n",
            "\n",
            "\n",
            "==> WARNING: A newer version of conda exists. <==\n",
            "  current version: 4.4.10\n",
            "  latest version: 4.8.3\n",
            "\n",
            "Please update conda by running\n",
            "\n",
            "    $ conda update -n base conda\n",
            "\n",
            "\n",
            "\n",
            "## Package Plan ##\n",
            "\n",
            "  environment location: /usr/local\n",
            "\n",
            "  removed specs: \n",
            "    - sympy\n",
            "\n",
            "\n",
            "The following packages will be REMOVED:\n",
            "\n",
            "    anaconda:        5.1.0-py36_2                         \n",
            "    sympy:           1.1.1-py36hc6d1c1c_0                 \n",
            "\n",
            "The following packages will be UPDATED:\n",
            "\n",
            "    certifi:         2020.4.5.1-py36h9f0ad1d_0 conda-forge --> 2020.4.5.1-py36_0\n",
            "    openssl:         1.1.1g-h516909a_0         conda-forge --> 1.1.1g-h7b6447c_0\n",
            "\n",
            "The following packages will be DOWNGRADED:\n",
            "\n",
            "    ca-certificates: 2020.4.5.1-hecc5488_0     conda-forge --> 2020.1.1-0       \n",
            "\n",
            "Preparing transaction: - \b\bdone\n",
            "Verifying transaction: | \b\b/ \b\bdone\n",
            "Executing transaction: \\ \b\b| \b\bdone\n",
            "Collecting qiskit\n",
            "  Downloading qiskit-0.19.3.tar.gz (3.9 kB)\n",
            "Collecting qiskit-terra==0.14.1\n",
            "  Downloading qiskit_terra-0.14.1-cp36-cp36m-manylinux2010_x86_64.whl (6.7 MB)\n",
            "\u001b[K     |████████████████████████████████| 6.7 MB 2.6 MB/s \n",
            "\u001b[?25hCollecting qiskit-aer==0.5.2\n",
            "  Downloading qiskit_aer-0.5.2-cp36-cp36m-manylinux2010_x86_64.whl (23.3 MB)\n",
            "\u001b[K     |████████████████████████████████| 23.3 MB 1.4 MB/s \n",
            "\u001b[?25hCollecting qiskit-ibmq-provider==0.7.2\n",
            "  Downloading qiskit_ibmq_provider-0.7.2-py3-none-any.whl (155 kB)\n",
            "\u001b[K     |████████████████████████████████| 155 kB 62.8 MB/s \n",
            "\u001b[?25hCollecting qiskit-ignis==0.3.0\n",
            "  Downloading qiskit_ignis-0.3.0-py3-none-any.whl (172 kB)\n",
            "\u001b[K     |████████████████████████████████| 172 kB 54.3 MB/s \n",
            "\u001b[?25hCollecting qiskit-aqua==0.7.1\n",
            "  Downloading qiskit_aqua-0.7.1-py3-none-any.whl (1.9 MB)\n",
            "\u001b[K     |████████████████████████████████| 1.9 MB 50.8 MB/s \n",
            "\u001b[?25hCollecting scipy>=1.4\n",
            "  Downloading scipy-1.4.1-cp36-cp36m-manylinux1_x86_64.whl (26.1 MB)\n",
            "\u001b[K     |████████████████████████████████| 26.1 MB 24 kB/s \n",
            "\u001b[?25hRequirement already satisfied: psutil>=5 in /usr/local/lib/python3.6/site-packages (from qiskit-terra==0.14.1->qiskit) (5.4.3)\n",
            "Requirement already satisfied: jsonschema>=2.6 in /usr/local/lib/python3.6/site-packages (from qiskit-terra==0.14.1->qiskit) (2.6.0)\n",
            "Collecting marshmallow<4,>=3\n",
            "  Downloading marshmallow-3.6.1-py2.py3-none-any.whl (45 kB)\n",
            "\u001b[K     |████████████████████████████████| 45 kB 3.5 MB/s \n",
            "\u001b[?25hCollecting networkx>=2.2; python_version > \"3.5\"\n",
            "  Downloading networkx-2.4-py3-none-any.whl (1.6 MB)\n",
            "\u001b[K     |████████████████████████████████| 1.6 MB 56.9 MB/s \n",
            "\u001b[?25hCollecting python-dateutil>=2.8.0\n",
            "  Downloading python_dateutil-2.8.1-py2.py3-none-any.whl (227 kB)\n",
            "\u001b[K     |████████████████████████████████| 227 kB 63.5 MB/s \n",
            "\u001b[?25hCollecting fastjsonschema>=2.10\n",
            "  Downloading fastjsonschema-2.14.4-py3-none-any.whl (21 kB)\n",
            "Requirement already satisfied: numpy>=1.17 in /usr/local/lib/python3.6/site-packages (from qiskit-terra==0.14.1->qiskit) (1.18.4)\n",
            "Collecting retworkx>=0.3.2\n",
            "  Downloading retworkx-0.3.4-cp36-cp36m-manylinux2010_x86_64.whl (763 kB)\n",
            "\u001b[K     |████████████████████████████████| 763 kB 57.1 MB/s \n",
            "\u001b[?25hRequirement already satisfied: ply>=3.10 in /usr/local/lib/python3.6/site-packages (from qiskit-terra==0.14.1->qiskit) (3.10)\n",
            "Collecting sympy>=1.3\n",
            "  Downloading sympy-1.6-py3-none-any.whl (5.8 MB)\n",
            "\u001b[K     |████████████████████████████████| 5.8 MB 48.3 MB/s \n",
            "\u001b[?25hCollecting python-constraint>=1.4\n",
            "  Downloading python-constraint-1.4.0.tar.bz2 (18 kB)\n",
            "Collecting dill>=0.3\n",
            "  Downloading dill-0.3.1.1.tar.gz (151 kB)\n",
            "\u001b[K     |████████████████████████████████| 151 kB 57.5 MB/s \n",
            "\u001b[?25hCollecting marshmallow-polyfield<6,>=5.7\n",
            "  Downloading marshmallow-polyfield-5.9.tar.gz (8.5 kB)\n",
            "Requirement already satisfied: cython>=0.27.1 in /usr/local/lib/python3.6/site-packages (from qiskit-aer==0.5.2->qiskit) (0.27.3)\n",
            "Collecting pybind11>=2.4\n",
            "  Downloading pybind11-2.5.0-py2.py3-none-any.whl (296 kB)\n",
            "\u001b[K     |████████████████████████████████| 296 kB 55.6 MB/s \n",
            "\u001b[?25hCollecting requests-ntlm>=1.1.0\n",
            "  Downloading requests_ntlm-1.1.0-py2.py3-none-any.whl (5.7 kB)\n",
            "Collecting nest-asyncio!=1.1.0,>=1.0.0\n",
            "  Downloading nest_asyncio-1.3.3-py3-none-any.whl (4.7 kB)\n",
            "Collecting requests>=2.19\n",
            "  Downloading requests-2.23.0-py2.py3-none-any.whl (58 kB)\n",
            "\u001b[K     |████████████████████████████████| 58 kB 5.9 MB/s \n",
            "\u001b[?25hCollecting websockets<8,>=7\n",
            "  Downloading websockets-7.0-cp36-cp36m-manylinux1_x86_64.whl (63 kB)\n",
            "\u001b[K     |████████████████████████████████| 63 kB 2.8 MB/s \n",
            "\u001b[?25hRequirement already satisfied: urllib3>=1.21.1 in /usr/local/lib/python3.6/site-packages (from qiskit-ibmq-provider==0.7.2->qiskit) (1.22)\n",
            "Requirement already satisfied: setuptools>=40.1.0 in /usr/local/lib/python3.6/site-packages (from qiskit-ignis==0.3.0->qiskit) (46.4.0.post20200518)\n",
            "Collecting scikit-learn>=0.20.0\n",
            "  Downloading scikit_learn-0.23.1-cp36-cp36m-manylinux1_x86_64.whl (6.8 MB)\n",
            "\u001b[K     |████████████████████████████████| 6.8 MB 50.1 MB/s \n",
            "\u001b[?25hCollecting fastdtw\n",
            "  Downloading fastdtw-0.3.4.tar.gz (133 kB)\n",
            "\u001b[K     |████████████████████████████████| 133 kB 62.8 MB/s \n",
            "\u001b[?25hCollecting docplex\n",
            "  Downloading docplex-2.14.186.tar.gz (572 kB)\n",
            "\u001b[K     |████████████████████████████████| 572 kB 49.6 MB/s \n",
            "\u001b[?25hCollecting quandl\n",
            "  Downloading Quandl-3.5.0-py2.py3-none-any.whl (25 kB)\n",
            "Collecting pyscf; sys_platform != \"win32\"\n",
            "  Downloading pyscf-1.7.2.post2-cp36-cp36m-manylinux1_x86_64.whl (26.0 MB)\n",
            "\u001b[K     |████████████████████████████████| 26.0 MB 1.3 MB/s \n",
            "\u001b[?25hRequirement already satisfied: h5py in /usr/local/lib/python3.6/site-packages (from qiskit-aqua==0.7.1->qiskit) (2.7.1)\n",
            "Collecting dlx\n",
            "  Downloading dlx-1.0.4.tar.gz (5.5 kB)\n",
            "Collecting decorator>=4.3.0\n",
            "  Downloading decorator-4.4.2-py2.py3-none-any.whl (9.2 kB)\n",
            "Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.6/site-packages (from python-dateutil>=2.8.0->qiskit-terra==0.14.1->qiskit) (1.11.0)\n",
            "Requirement already satisfied: mpmath>=0.19 in /usr/local/lib/python3.6/site-packages (from sympy>=1.3->qiskit-terra==0.14.1->qiskit) (1.0.0)\n",
            "Requirement already satisfied: cryptography>=1.3 in /usr/local/lib/python3.6/site-packages (from requests-ntlm>=1.1.0->qiskit-ibmq-provider==0.7.2->qiskit) (2.1.4)\n",
            "Collecting ntlm-auth>=1.0.2\n",
            "  Downloading ntlm_auth-1.4.0-py2.py3-none-any.whl (29 kB)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.6/site-packages (from requests>=2.19->qiskit-ibmq-provider==0.7.2->qiskit) (2020.4.5.1)\n",
            "Requirement already satisfied: idna<3,>=2.5 in /usr/local/lib/python3.6/site-packages (from requests>=2.19->qiskit-ibmq-provider==0.7.2->qiskit) (2.6)\n",
            "Requirement already satisfied: chardet<4,>=3.0.2 in /usr/local/lib/python3.6/site-packages (from requests>=2.19->qiskit-ibmq-provider==0.7.2->qiskit) (3.0.4)\n",
            "Collecting threadpoolctl>=2.0.0\n",
            "  Downloading threadpoolctl-2.1.0-py3-none-any.whl (12 kB)\n",
            "Collecting joblib>=0.11\n",
            "  Downloading joblib-0.15.1-py3-none-any.whl (298 kB)\n",
            "\u001b[K     |████████████████████████████████| 298 kB 55.9 MB/s \n",
            "\u001b[?25hRequirement already satisfied: pandas>=0.14 in /usr/local/lib/python3.6/site-packages (from quandl->qiskit-aqua==0.7.1->qiskit) (1.0.4)\n",
            "Collecting more-itertools\n",
            "  Downloading more_itertools-8.3.0-py3-none-any.whl (44 kB)\n",
            "\u001b[K     |████████████████████████████████| 44 kB 3.1 MB/s \n",
            "\u001b[?25hCollecting inflection>=0.3.1\n",
            "  Downloading inflection-0.4.0-py2.py3-none-any.whl (5.8 kB)\n",
            "Requirement already satisfied: asn1crypto>=0.21.0 in /usr/local/lib/python3.6/site-packages (from cryptography>=1.3->requests-ntlm>=1.1.0->qiskit-ibmq-provider==0.7.2->qiskit) (0.24.0)\n",
            "Requirement already satisfied: cffi>=1.7 in /usr/local/lib/python3.6/site-packages (from cryptography>=1.3->requests-ntlm>=1.1.0->qiskit-ibmq-provider==0.7.2->qiskit) (1.11.4)\n",
            "Requirement already satisfied: pytz>=2017.2 in /usr/local/lib/python3.6/site-packages (from pandas>=0.14->quandl->qiskit-aqua==0.7.1->qiskit) (2017.3)\n",
            "Requirement already satisfied: pycparser in /usr/local/lib/python3.6/site-packages (from cffi>=1.7->cryptography>=1.3->requests-ntlm>=1.1.0->qiskit-ibmq-provider==0.7.2->qiskit) (2.18)\n",
            "Building wheels for collected packages: qiskit, python-constraint, dill, marshmallow-polyfield, fastdtw, docplex, dlx\n",
            "  Building wheel for qiskit (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for qiskit: filename=qiskit-0.19.3-py3-none-any.whl size=2898 sha256=adb94bf23ad6b4961c8a9f2a242df9464a2ee6f929d1be30903f5fbfc2f31fc0\n",
            "  Stored in directory: /root/.cache/pip/wheels/68/ba/b0/78177cd3f8cba974755148c5901a9abc2696660246187e0efd\n",
            "  Building wheel for python-constraint (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for python-constraint: filename=python_constraint-1.4.0-py2.py3-none-any.whl size=24080 sha256=abe7b3fb865b85bee58a743007d173a7dea755ddb8325f723a00002b017fe079\n",
            "  Stored in directory: /root/.cache/pip/wheels/c2/e2/65/6222764e3e5b677acdb04b592abab7e15140a70baace656d81\n",
            "  Building wheel for dill (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for dill: filename=dill-0.3.1.1-py3-none-any.whl size=78530 sha256=4d8e09b0880bc026478dc2b2074c87d87a277d424beea8c23fad0398ff3a3e23\n",
            "  Stored in directory: /root/.cache/pip/wheels/09/84/74/d2b4feb9ac9488bc83c475cb2cbe8e8b7d9cea8320d32f3787\n",
            "  Building wheel for marshmallow-polyfield (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for marshmallow-polyfield: filename=marshmallow_polyfield-5.9-py3-none-any.whl size=8535 sha256=253fc794a78dbb020d74e68a5d98a9dd5bab491dbd4364d5243abd14200d8277\n",
            "  Stored in directory: /root/.cache/pip/wheels/9a/54/a7/4e0d56296d25c6eae6b1f213db626d8a4828228406abb73918\n",
            "  Building wheel for fastdtw (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for fastdtw: filename=fastdtw-0.3.4-cp36-cp36m-linux_x86_64.whl size=391289 sha256=0bae1ea9b83d9e0ff1dea6fc1f27198534fbce1a7c1b6e91fbeac65178d51a7f\n",
            "  Stored in directory: /root/.cache/pip/wheels/50/66/7a/7c98f31ea8ecc3a04a5b20ec6b48fbdf02c899034e95cc1ede\n",
            "  Building wheel for docplex (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for docplex: filename=docplex-2.14.186-py3-none-any.whl size=634441 sha256=f1f5e133bf1e667200fdda9e5f27dc7c1414f76397059f8786c10888c5155f7f\n",
            "  Stored in directory: /root/.cache/pip/wheels/af/8a/77/21b87b28bbef8aeea189643cb281511d584d1f30e57378dd9c\n",
            "  Building wheel for dlx (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for dlx: filename=dlx-1.0.4-py3-none-any.whl size=5709 sha256=e1851de1220d4f9dba5bc72a64467a36995d70d867b0ee4fe83710dba3d6fef1\n",
            "  Stored in directory: /root/.cache/pip/wheels/ad/a3/52/859d389366d76e9e371e5a2dabefde5da7678c0032f008ca0a\n",
            "Successfully built qiskit python-constraint dill marshmallow-polyfield fastdtw docplex dlx\n",
            "Installing collected packages: scipy, marshmallow, decorator, networkx, python-dateutil, fastjsonschema, retworkx, sympy, python-constraint, dill, marshmallow-polyfield, qiskit-terra, pybind11, qiskit-aer, requests, ntlm-auth, requests-ntlm, nest-asyncio, websockets, qiskit-ibmq-provider, qiskit-ignis, threadpoolctl, joblib, scikit-learn, fastdtw, docplex, more-itertools, inflection, quandl, pyscf, dlx, qiskit-aqua, qiskit\n",
            "  Attempting uninstall: scipy\n",
            "    Found existing installation: scipy 1.0.0\n",
            "    Uninstalling scipy-1.0.0:\n",
            "      Successfully uninstalled scipy-1.0.0\n",
            "  Attempting uninstall: decorator\n",
            "    Found existing installation: decorator 4.2.1\n",
            "    Uninstalling decorator-4.2.1:\n",
            "      Successfully uninstalled decorator-4.2.1\n",
            "  Attempting uninstall: networkx\n",
            "    Found existing installation: networkx 2.1\n",
            "    Uninstalling networkx-2.1:\n",
            "      Successfully uninstalled networkx-2.1\n",
            "  Attempting uninstall: python-dateutil\n",
            "    Found existing installation: python-dateutil 2.6.1\n",
            "    Uninstalling python-dateutil-2.6.1:\n",
            "      Successfully uninstalled python-dateutil-2.6.1\n",
            "  Attempting uninstall: requests\n",
            "    Found existing installation: requests 2.18.4\n",
            "    Uninstalling requests-2.18.4:\n",
            "      Successfully uninstalled requests-2.18.4\n",
            "  Attempting uninstall: scikit-learn\n",
            "    Found existing installation: scikit-learn 0.19.1\n",
            "    Uninstalling scikit-learn-0.19.1:\n",
            "      Successfully uninstalled scikit-learn-0.19.1\n",
            "Successfully installed decorator-4.4.2 dill-0.3.1.1 dlx-1.0.4 docplex-2.14.186 fastdtw-0.3.4 fastjsonschema-2.14.4 inflection-0.4.0 joblib-0.15.1 marshmallow-3.6.1 marshmallow-polyfield-5.9 more-itertools-8.3.0 nest-asyncio-1.3.3 networkx-2.4 ntlm-auth-1.4.0 pybind11-2.5.0 pyscf-1.7.2.post2 python-constraint-1.4.0 python-dateutil-2.8.1 qiskit-0.19.3 qiskit-aer-0.5.2 qiskit-aqua-0.7.1 qiskit-ibmq-provider-0.7.2 qiskit-ignis-0.3.0 qiskit-terra-0.14.1 quandl-3.5.0 requests-2.23.0 requests-ntlm-1.1.0 retworkx-0.3.4 scikit-learn-0.23.1 scipy-1.4.1 sympy-1.6 threadpoolctl-2.1.0 websockets-7.0\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "display_data",
          "data": {
            "application/vnd.colab-display-data+json": {
              "pip_warning": {
                "packages": [
                  "dateutil",
                  "decorator",
                  "requests"
                ]
              }
            }
          },
          "metadata": {
            "tags": []
          }
        },
        {
          "output_type": "stream",
          "text": [
            "Collecting pennylane\n",
            "  Downloading PennyLane-0.9.0-py3-none-any.whl (239 kB)\n",
            "\u001b[?25l\r\u001b[K     |█▍                              | 10 kB 23.3 MB/s eta 0:00:01\r\u001b[K     |██▊                             | 20 kB 1.5 MB/s eta 0:00:01\r\u001b[K     |████                            | 30 kB 2.1 MB/s eta 0:00:01\r\u001b[K     |█████▌                          | 40 kB 2.4 MB/s eta 0:00:01\r\u001b[K     |██████▉                         | 51 kB 2.0 MB/s eta 0:00:01\r\u001b[K     |████████▏                       | 61 kB 2.2 MB/s eta 0:00:01\r\u001b[K     |█████████▋                      | 71 kB 2.4 MB/s eta 0:00:01\r\u001b[K     |███████████                     | 81 kB 2.7 MB/s eta 0:00:01\r\u001b[K     |████████████▎                   | 92 kB 2.8 MB/s eta 0:00:01\r\u001b[K     |█████████████▊                  | 102 kB 2.7 MB/s eta 0:00:01\r\u001b[K     |███████████████                 | 112 kB 2.7 MB/s eta 0:00:01\r\u001b[K     |████████████████▍               | 122 kB 2.7 MB/s eta 0:00:01\r\u001b[K     |█████████████████▊              | 133 kB 2.7 MB/s eta 0:00:01\r\u001b[K     |███████████████████▏            | 143 kB 2.7 MB/s eta 0:00:01\r\u001b[K     |████████████████████▌           | 153 kB 2.7 MB/s eta 0:00:01\r\u001b[K     |█████████████████████▉          | 163 kB 2.7 MB/s eta 0:00:01\r\u001b[K     |███████████████████████▎        | 174 kB 2.7 MB/s eta 0:00:01\r\u001b[K     |████████████████████████▋       | 184 kB 2.7 MB/s eta 0:00:01\r\u001b[K     |██████████████████████████      | 194 kB 2.7 MB/s eta 0:00:01\r\u001b[K     |███████████████████████████▍    | 204 kB 2.7 MB/s eta 0:00:01\r\u001b[K     |████████████████████████████▊   | 215 kB 2.7 MB/s eta 0:00:01\r\u001b[K     |██████████████████████████████  | 225 kB 2.7 MB/s eta 0:00:01\r\u001b[K     |███████████████████████████████▍| 235 kB 2.7 MB/s eta 0:00:01\r\u001b[K     |████████████████████████████████| 239 kB 2.7 MB/s \n",
            "\u001b[?25hRequirement already satisfied: numpy in /usr/local/lib/python3.6/site-packages (from pennylane) (1.18.4)\n",
            "Collecting semantic-version==2.6\n",
            "  Downloading semantic_version-2.6.0-py3-none-any.whl (14 kB)\n",
            "Requirement already satisfied: scipy in /usr/local/lib/python3.6/site-packages (from pennylane) (1.4.1)\n",
            "Collecting autograd\n",
            "  Downloading autograd-1.3.tar.gz (38 kB)\n",
            "Collecting appdirs\n",
            "  Downloading appdirs-1.4.4-py2.py3-none-any.whl (9.6 kB)\n",
            "Collecting toml\n",
            "  Downloading toml-0.10.1-py2.py3-none-any.whl (19 kB)\n",
            "Requirement already satisfied: networkx in /usr/local/lib/python3.6/site-packages (from pennylane) (2.4)\n",
            "Collecting future>=0.15.2\n",
            "  Downloading future-0.18.2.tar.gz (829 kB)\n",
            "\u001b[K     |████████████████████████████████| 829 kB 8.4 MB/s \n",
            "\u001b[?25hRequirement already satisfied: decorator>=4.3.0 in /usr/local/lib/python3.6/site-packages (from networkx->pennylane) (4.4.2)\n",
            "Building wheels for collected packages: autograd, future\n",
            "  Building wheel for autograd (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for autograd: filename=autograd-1.3-py3-none-any.whl size=47990 sha256=8f373a6540801bb11299cac7b559337d67a65056e4224bf375174d17efe8557f\n",
            "  Stored in directory: /root/.cache/pip/wheels/b2/a5/3c/929b91003a3b75a175ccba7cef35200a1890cdb46903f18072\n",
            "  Building wheel for future (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for future: filename=future-0.18.2-py3-none-any.whl size=491058 sha256=31ddd924ca6f0c92f25ad2c2c30c48ae0d3c4068bc0cc90bf8ebb64b0c4e02f0\n",
            "  Stored in directory: /root/.cache/pip/wheels/6e/9c/ed/4499c9865ac1002697793e0ae05ba6be33553d098f3347fb94\n",
            "Successfully built autograd future\n",
            "Installing collected packages: semantic-version, future, autograd, appdirs, toml, pennylane\n",
            "Successfully installed appdirs-1.4.4 autograd-1.3 future-0.18.2 pennylane-0.9.0 semantic-version-2.6.0 toml-0.10.1\n",
            "CPU times: user 1.84 s, sys: 416 ms, total: 2.26 s\n",
            "Wall time: 6min 20s\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "C-nWJ6XGXlGY",
        "colab_type": "code",
        "outputId": "6828d4c7-d690-472c-e811-a4e9515ad228",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        }
      },
      "source": [
        "!pip install gdown\n",
        "!gdown https://drive.google.com/uc?id=1n7x4fP4A85rDDUcQ9nrnS9l2lu9pZWPm\n",
        "!gdown https://drive.google.com/uc?id=1MKQnbbzw2l92TrImG8F0qpQuCa4luZG8\n",
        "!unzip data.zip\n",
        "!unzip PyBioMed-1.0.zip\n",
        "%cd PyBioMed-1.0\n",
        "!python3 setup.py install\n",
        "%cd .."
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Collecting gdown\n",
            "  Downloading gdown-3.11.0.tar.gz (8.6 kB)\n",
            "  Installing build dependencies ... \u001b[?25l\u001b[?25hdone\n",
            "  Getting requirements to build wheel ... \u001b[?25l\u001b[?25hdone\n",
            "    Preparing wheel metadata ... \u001b[?25l\u001b[?25hdone\n",
            "Requirement already satisfied: six in /usr/local/lib/python3.6/site-packages (from gdown) (1.11.0)\n",
            "Collecting tqdm\n",
            "  Downloading tqdm-4.46.0-py2.py3-none-any.whl (63 kB)\n",
            "\u001b[K     |████████████████████████████████| 63 kB 1.5 MB/s \n",
            "\u001b[?25hRequirement already satisfied: requests[socks] in /usr/local/lib/python3.6/site-packages (from gdown) (2.23.0)\n",
            "Requirement already satisfied: filelock in /usr/local/lib/python3.6/site-packages (from gdown) (2.0.13)\n",
            "Requirement already satisfied: urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1 in /usr/local/lib/python3.6/site-packages (from requests[socks]->gdown) (1.22)\n",
            "Requirement already satisfied: idna<3,>=2.5 in /usr/local/lib/python3.6/site-packages (from requests[socks]->gdown) (2.6)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.6/site-packages (from requests[socks]->gdown) (2020.4.5.1)\n",
            "Requirement already satisfied: chardet<4,>=3.0.2 in /usr/local/lib/python3.6/site-packages (from requests[socks]->gdown) (3.0.4)\n",
            "Requirement already satisfied: PySocks!=1.5.7,>=1.5.6; extra == \"socks\" in /usr/local/lib/python3.6/site-packages (from requests[socks]->gdown) (1.6.7)\n",
            "Building wheels for collected packages: gdown\n",
            "  Building wheel for gdown (PEP 517) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for gdown: filename=gdown-3.11.0-py3-none-any.whl size=9619 sha256=60c7828301a0e0ae0113c36a95268cdb6d3df730f0c9eaa70e25d9b834d27913\n",
            "  Stored in directory: /root/.cache/pip/wheels/38/fe/03/8ea0c23ae5185b924065e9e5cf7b054268d4f73ab7c06946aa\n",
            "Successfully built gdown\n",
            "Installing collected packages: tqdm, gdown\n",
            "Successfully installed gdown-3.11.0 tqdm-4.46.0\n",
            "Downloading...\n",
            "From: https://drive.google.com/uc?id=1n7x4fP4A85rDDUcQ9nrnS9l2lu9pZWPm\n",
            "To: /content/data.zip\n",
            "141MB [00:02, 51.4MB/s]\n",
            "Downloading...\n",
            "From: https://drive.google.com/uc?id=1MKQnbbzw2l92TrImG8F0qpQuCa4luZG8\n",
            "To: /content/PyBioMed-1.0.zip\n",
            "22.4MB [00:00, 114MB/s] \n",
            "Archive:  data.zip\n",
            "  inflating: positive_pair.csv       \n",
            "   creating: __MACOSX/\n",
            "  inflating: __MACOSX/._positive_pair.csv  \n",
            "  inflating: protein_descriptor.csv  \n",
            "  inflating: __MACOSX/._protein_descriptor.csv  \n",
            "  inflating: molecule_descriptor_new.csv  \n",
            "  inflating: __MACOSX/._molecule_descriptor_new.csv  \n",
            "  inflating: negative_pair.csv       \n",
            "  inflating: __MACOSX/._negative_pair.csv  \n",
            "Archive:  PyBioMed-1.0.zip\n",
            "   creating: PyBioMed-1.0/\n",
            "   creating: PyBioMed-1.0/build/\n",
            "   creating: PyBioMed-1.0/build/lib/\n",
            "   creating: PyBioMed-1.0/build/lib/PyBioMed/\n",
            "   creating: PyBioMed-1.0/build/lib/PyBioMed/doc/\n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/doc/1atp (1).pdb  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/doc/1atp.pdb  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/doc/1efz.pdb  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/doc/1f88 (1).ent.gz  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/doc/1f88.ent.gz  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/doc/1f88.pdb  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/doc/application (1).rst  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/doc/application.rst  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/doc/conf.py  \n",
            "   creating: PyBioMed-1.0/build/lib/PyBioMed/doc/Descriptor/\n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/doc/Descriptor/Documentation.pdf  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/doc/Descriptor/PyBioMed Chem.pdf  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/doc/Descriptor/PyBioMed DNA.pdf  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/doc/Descriptor/PyBioMed Interaction.pdf  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/doc/Descriptor/PyBioMed Protein.pdf  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/doc/download.rst  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/doc/index.rst  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/doc/make.bat  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/doc/Makefile  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/doc/modules (1).rst  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/doc/modules.rst  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/doc/overview.rst  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/doc/test.rst  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/doc/User_guide.rst  \n",
            "   creating: PyBioMed-1.0/build/lib/PyBioMed/example/\n",
            "   creating: PyBioMed-1.0/build/lib/PyBioMed/example/caco2/\n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/example/caco2/caco2.xlsx  \n",
            "   creating: PyBioMed-1.0/build/lib/PyBioMed/example/dna/\n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/example/dna/H_sapiens_neg.fasta  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/example/dna/H_sapiens_pos.fasta  \n",
            "   creating: PyBioMed-1.0/build/lib/PyBioMed/example/dpi/\n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/example/dpi/DPI_SMIs.xlsx  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/example/dpi/Enzyme.xls  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/example/dpi/hsa_seqs_all.tsv  \n",
            "   creating: PyBioMed-1.0/build/lib/PyBioMed/example/solubility/\n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/example/solubility/Solubility-total.xlsx  \n",
            "   creating: PyBioMed-1.0/build/lib/PyBioMed/example/subcell/\n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/example/subcell/Cytoplasm_seq.txt  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/example/subcell/Nuclear_seq.txt  \n",
            "   creating: PyBioMed-1.0/build/lib/PyBioMed/PyDNA/\n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/Pydna.py  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyDNA/1atp.pdb  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyDNA/1efz (1).pdb  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyDNA/1efz.pdb  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyDNA/1f88.pdb  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyDNA/mmc3 (1).data  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyDNA/mmc3.data  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyDNA/mmc4 (1).data  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyDNA/mmc4.data  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyDNA/PyDNA (1).py  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyDNA/PyDNA.py  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyDNA/PyDNAac (1).py  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyDNA/PyDNAac.py  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyDNA/PyDNAacutil.py  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyDNA/PyDNAnac.py  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyDNA/PyDNAnacutil (1).py  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyDNA/PyDNAnacutil.py  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyDNA/PyDNApsenac (1).py  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyDNA/PyDNApsenac.py  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyDNA/PyDNApsenacutil.py  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyDNA/PyDNAutil (1).py  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyDNA/PyDNAutil.py  \n",
            " extracting: PyBioMed-1.0/build/lib/PyBioMed/PyDNA/__init__ (1).py  \n",
            " extracting: PyBioMed-1.0/build/lib/PyBioMed/PyDNA/__init__.py  \n",
            "   creating: PyBioMed-1.0/build/lib/PyBioMed/PyGetMol/\n",
            " extracting: PyBioMed-1.0/build/lib/PyBioMed/PyGetMol/1atp.ent.gz  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyGetMol/1atp.pdb  \n",
            " extracting: PyBioMed-1.0/build/lib/PyBioMed/PyGetMol/1efz.ent.gz  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyGetMol/1efz.pdb  \n",
            " extracting: PyBioMed-1.0/build/lib/PyBioMed/PyGetMol/1f88.ent.gz  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyGetMol/1f88.pdb  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyGetMol/GetDNA.py  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyGetMol/Getmol.py  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyGetMol/GetProtein.py  \n",
            " extracting: PyBioMed-1.0/build/lib/PyBioMed/PyGetMol/__init__.py  \n",
            "   creating: PyBioMed-1.0/build/lib/PyBioMed/PyInteraction/\n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/Pyinteraction.py  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyInteraction/PyInteraction.py  \n",
            " extracting: PyBioMed-1.0/build/lib/PyBioMed/PyInteraction/__init__.py  \n",
            "   creating: PyBioMed-1.0/build/lib/PyBioMed/PyMolecule/\n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/Pymolecule.py  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyMolecule/AtomProperty.py  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyMolecule/AtomTypes.py  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyMolecule/basak.py  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyMolecule/bcut.py  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyMolecule/cats2d.py  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyMolecule/charge.py  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyMolecule/connectivity.py  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyMolecule/constitution.py  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyMolecule/Crippen.txt  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyMolecule/estate.py  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyMolecule/fingerprint.py  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyMolecule/geary.py  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyMolecule/ghosecrippen.py  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyMolecule/kappa.py  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyMolecule/MinimalFeatures.fdef  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyMolecule/moe.py  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyMolecule/molproperty.py  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyMolecule/moran.py  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyMolecule/moreaubroto.py  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyMolecule/PubChemFingerprints.py  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyMolecule/Scaffolds.py  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyMolecule/topology.py  \n",
            " extracting: PyBioMed-1.0/build/lib/PyBioMed/PyMolecule/__init__.py  \n",
            "   creating: PyBioMed-1.0/build/lib/PyBioMed/PyPretreat/\n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyPretreat/PyDNAutil.py  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyPretreat/PyPreTools (1).py  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyPretreat/PyPreTools.py  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyPretreat/PyPretreatDNA.py  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyPretreat/PyPretreatMol.py  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyPretreat/PyPretreatMolutil.py  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyPretreat/PyPretreatPro.py  \n",
            " extracting: PyBioMed-1.0/build/lib/PyBioMed/PyPretreat/__init__.py  \n",
            "   creating: PyBioMed-1.0/build/lib/PyBioMed/PyProtein/\n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/Pyprotein.py  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyProtein/AAComposition.py  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyProtein/AAIndex.py  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyProtein/aaindex1  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyProtein/aaindex2  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyProtein/aaindex3  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyProtein/Autocorrelation.py  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyProtein/ConjointTriad.py  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyProtein/CTD.py  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyProtein/GetProteinFromUniprot.py  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyProtein/GetSubSeq.py  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyProtein/ProCheck.py  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyProtein/PseudoAAC.py  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyProtein/PyProtein.py  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyProtein/PyProteinAAComposition.py  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyProtein/PyProteinAAIndex.py  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyProtein/QuasiSequenceOrder.py  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/PyProtein/__init__.py  \n",
            "   creating: PyBioMed-1.0/build/lib/PyBioMed/test/\n",
            " extracting: PyBioMed-1.0/build/lib/PyBioMed/test/1atp.ent.gz  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/test/1atp.pdb  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/test/1efz.pdb  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/test/1f88.pdb  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/test/test.py  \n",
            "   creating: PyBioMed-1.0/build/lib/PyBioMed/test/test_data/\n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/test/test_data/drug.inchi  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/test/test_data/drug.mol  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/test/test_data/drug.sdf  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/test/test_data/drug.smi  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/test/test_data/example.fasta  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/test/test_data/hs.fasta  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/test/test_data/neg.fasta  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/test/test_data/pos.fasta  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/test/test_data/protein.fasta  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/test/test_data/target.txt  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/test/test_data/test.mol  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/test/test_PyBioMed.py  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/test/test_PyDNA.py  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/test/test_PyGetMol.py  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/test/test_PyInteration.py  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/test/test_PyMolecule.py  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/test/test_PyPretreat.py  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/test/test_PyProtein.py  \n",
            " extracting: PyBioMed-1.0/build/lib/PyBioMed/test/__init__.py  \n",
            "  inflating: PyBioMed-1.0/build/lib/PyBioMed/__init__.py  \n",
            "  inflating: PyBioMed-1.0/PKG-INFO   \n",
            "   creating: PyBioMed-1.0/PyBioMed/\n",
            "   creating: PyBioMed-1.0/PyBioMed/.vscode/\n",
            "  inflating: PyBioMed-1.0/PyBioMed/.vscode/settings.json  \n",
            "   creating: PyBioMed-1.0/PyBioMed/doc/\n",
            "  inflating: PyBioMed-1.0/PyBioMed/doc/1atp(1).pdb  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/doc/1atp.pdb  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/doc/1efz.pdb  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/doc/1f88.ent(1).gz  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/doc/1f88.ent.gz  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/doc/1f88.pdb  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/doc/application(1).rst  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/doc/application.rst  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/doc/conf.py  \n",
            "   creating: PyBioMed-1.0/PyBioMed/doc/Descriptor/\n",
            "  inflating: PyBioMed-1.0/PyBioMed/doc/Descriptor/Documentation.pdf  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/doc/Descriptor/PyBioMed Chem.pdf  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/doc/Descriptor/PyBioMed DNA.pdf  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/doc/Descriptor/PyBioMed Interaction.pdf  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/doc/Descriptor/PyBioMed Protein.pdf  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/doc/download.rst  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/doc/index.rst  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/doc/make.bat  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/doc/Makefile  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/doc/modules(1).rst  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/doc/modules.rst  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/doc/overview.rst  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/doc/test.rst  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/doc/User_guide.rst  \n",
            "   creating: PyBioMed-1.0/PyBioMed/example/\n",
            "   creating: PyBioMed-1.0/PyBioMed/example/caco2/\n",
            "  inflating: PyBioMed-1.0/PyBioMed/example/caco2/caco2.xlsx  \n",
            "   creating: PyBioMed-1.0/PyBioMed/example/dna/\n",
            "  inflating: PyBioMed-1.0/PyBioMed/example/dna/H_sapiens_neg.fasta  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/example/dna/H_sapiens_pos.fasta  \n",
            "   creating: PyBioMed-1.0/PyBioMed/example/dpi/\n",
            "  inflating: PyBioMed-1.0/PyBioMed/example/dpi/DPI_SMIs.xlsx  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/example/dpi/Enzyme.xls  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/example/dpi/hsa_seqs_all.tsv  \n",
            "   creating: PyBioMed-1.0/PyBioMed/example/solubility/\n",
            "  inflating: PyBioMed-1.0/PyBioMed/example/solubility/Solubility-total.xlsx  \n",
            "   creating: PyBioMed-1.0/PyBioMed/example/subcell/\n",
            "  inflating: PyBioMed-1.0/PyBioMed/example/subcell/Cytoplasm_seq.txt  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/example/subcell/Nuclear_seq.txt  \n",
            "   creating: PyBioMed-1.0/PyBioMed/PyDNA/\n",
            "  inflating: PyBioMed-1.0/PyBioMed/Pydna.py  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyDNA/1atp.pdb  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyDNA/1efz(1).pdb  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyDNA/1efz.pdb  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyDNA/1f88.pdb  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyDNA/mmc3(1).data  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyDNA/mmc3.data  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyDNA/mmc4(1).data  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyDNA/mmc4.data  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyDNA/PyDNA(1).py  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyDNA/PyDNA.py  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyDNA/PyDNAac(1).py  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyDNA/PyDNAac.py  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyDNA/PyDNAacutil.py  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyDNA/PyDNAnac.py  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyDNA/PyDNAnacutil(1).py  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyDNA/PyDNAnacutil.py  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyDNA/PyDNApsenac(1).py  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyDNA/PyDNApsenac.py  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyDNA/PyDNApsenacutil.py  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyDNA/PyDNAutil(1).py  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyDNA/PyDNAutil.py  \n",
            " extracting: PyBioMed-1.0/PyBioMed/PyDNA/__init__(1).py  \n",
            " extracting: PyBioMed-1.0/PyBioMed/PyDNA/__init__.py  \n",
            "   creating: PyBioMed-1.0/PyBioMed/PyGetMol/\n",
            " extracting: PyBioMed-1.0/PyBioMed/PyGetMol/1atp.ent.gz  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyGetMol/1atp.pdb  \n",
            " extracting: PyBioMed-1.0/PyBioMed/PyGetMol/1efz.ent.gz  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyGetMol/1efz.pdb  \n",
            " extracting: PyBioMed-1.0/PyBioMed/PyGetMol/1f88.ent.gz  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyGetMol/1f88.pdb  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyGetMol/GetDNA.py  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyGetMol/Getmol.py  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyGetMol/GetProtein.py  \n",
            " extracting: PyBioMed-1.0/PyBioMed/PyGetMol/__init__.py  \n",
            "   creating: PyBioMed-1.0/PyBioMed/PyGetMol/__pycache__/\n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyGetMol/__pycache__/Getmol.cpython-36.pyc  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyGetMol/__pycache__/__init__.cpython-36.pyc  \n",
            "   creating: PyBioMed-1.0/PyBioMed/PyInteraction/\n",
            "  inflating: PyBioMed-1.0/PyBioMed/Pyinteraction.py  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyInteraction/PyInteraction.py  \n",
            " extracting: PyBioMed-1.0/PyBioMed/PyInteraction/__init__.py  \n",
            "   creating: PyBioMed-1.0/PyBioMed/PyMolecule/\n",
            "  inflating: PyBioMed-1.0/PyBioMed/Pymolecule.py  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyMolecule/AtomProperty.py  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyMolecule/AtomTypes.py  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyMolecule/basak.py  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyMolecule/bcut.py  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyMolecule/cats2d.py  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyMolecule/charge.py  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyMolecule/connectivity.py  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyMolecule/constitution.py  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyMolecule/Crippen.txt  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyMolecule/estate.py  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyMolecule/fingerprint.py  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyMolecule/geary.py  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyMolecule/ghosecrippen.py  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyMolecule/kappa.py  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyMolecule/MinimalFeatures.fdef  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyMolecule/moe.py  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyMolecule/molproperty.py  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyMolecule/moran.py  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyMolecule/moreaubroto.py  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyMolecule/PubChemFingerprints.py  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyMolecule/Scaffolds.py  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyMolecule/topology.py  \n",
            " extracting: PyBioMed-1.0/PyBioMed/PyMolecule/__init__.py  \n",
            "   creating: PyBioMed-1.0/PyBioMed/PyMolecule/__pycache__/\n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyMolecule/__pycache__/AtomProperty.cpython-36.pyc  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyMolecule/__pycache__/AtomTypes.cpython-36.pyc  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyMolecule/__pycache__/basak.cpython-36.pyc  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyMolecule/__pycache__/bcut.cpython-36.pyc  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyMolecule/__pycache__/cats2d.cpython-36.pyc  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyMolecule/__pycache__/charge.cpython-36.pyc  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyMolecule/__pycache__/connectivity.cpython-36.pyc  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyMolecule/__pycache__/constitution.cpython-36.pyc  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyMolecule/__pycache__/estate.cpython-36.pyc  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyMolecule/__pycache__/kappa.cpython-36.pyc  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyMolecule/__pycache__/molproperty.cpython-36.pyc  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyMolecule/__pycache__/moreaubroto.cpython-36.pyc  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyMolecule/__pycache__/PubChemFingerprints.cpython-36.pyc  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyMolecule/__pycache__/Scaffolds.cpython-36.pyc  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyMolecule/__pycache__/__init__.cpython-36.pyc  \n",
            "   creating: PyBioMed-1.0/PyBioMed/PyPretreat/\n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyPretreat/PyDNAutil.py  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyPretreat/PyPreTools(1).py  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyPretreat/PyPreTools.py  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyPretreat/PyPretreatDNA.py  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyPretreat/PyPretreatMol.py  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyPretreat/PyPretreatMolutil.py  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyPretreat/PyPretreatPro.py  \n",
            " extracting: PyBioMed-1.0/PyBioMed/PyPretreat/__init__.py  \n",
            "   creating: PyBioMed-1.0/PyBioMed/PyProtein/\n",
            "  inflating: PyBioMed-1.0/PyBioMed/Pyprotein.py  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyProtein/AAComposition.py  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyProtein/AAIndex.py  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyProtein/aaindex1  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyProtein/aaindex2  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyProtein/aaindex3  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyProtein/Autocorrelation.py  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyProtein/ConjointTriad.py  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyProtein/CTD.py  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyProtein/GetProteinFromUniprot.py  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyProtein/GetSubSeq.py  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyProtein/ProCheck.py  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyProtein/PseudoAAC.py  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyProtein/PyProtein.py  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyProtein/PyProteinAAComposition.py  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyProtein/PyProteinAAIndex.py  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyProtein/QuasiSequenceOrder.py  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyProtein/__init__.py  \n",
            "   creating: PyBioMed-1.0/PyBioMed/PyProtein/__pycache__/\n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyProtein/__pycache__/AAComposition.cpython-36.pyc  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyProtein/__pycache__/AAIndex.cpython-36.pyc  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyProtein/__pycache__/Autocorrelation.cpython-36.pyc  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyProtein/__pycache__/ConjointTriad.cpython-36.pyc  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyProtein/__pycache__/CTD.cpython-36.pyc  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyProtein/__pycache__/PyProtein.cpython-36.pyc  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/PyProtein/__pycache__/__init__.cpython-36.pyc  \n",
            "   creating: PyBioMed-1.0/PyBioMed/test/\n",
            " extracting: PyBioMed-1.0/PyBioMed/test/1atp.ent.gz  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/test/1atp.pdb  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/test/1efz.pdb  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/test/1f88.pdb  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/test/test.py  \n",
            "   creating: PyBioMed-1.0/PyBioMed/test/test_data/\n",
            "  inflating: PyBioMed-1.0/PyBioMed/test/test_data/drug.inchi  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/test/test_data/drug.mol  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/test/test_data/drug.sdf  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/test/test_data/drug.smi  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/test/test_data/example.fasta  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/test/test_data/hs.fasta  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/test/test_data/neg.fasta  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/test/test_data/pos.fasta  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/test/test_data/protein.fasta  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/test/test_data/target.txt  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/test/test_data/test.mol  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/test/test_PyBioMed.py  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/test/test_PyDNA.py  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/test/test_PyGetMol.py  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/test/test_PyInteration.py  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/test/test_PyMolecule.py  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/test/test_PyPretreat.py  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/test/test_PyProtein.py  \n",
            " extracting: PyBioMed-1.0/PyBioMed/test/__init__.py  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/__init__.py  \n",
            "   creating: PyBioMed-1.0/PyBioMed/__pycache__/\n",
            "  inflating: PyBioMed-1.0/PyBioMed/__pycache__/Pymolecule.cpython-36.pyc  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/__pycache__/Pyprotein.cpython-36.pyc  \n",
            "  inflating: PyBioMed-1.0/PyBioMed/__pycache__/__init__.cpython-36.pyc  \n",
            "  inflating: PyBioMed-1.0/setup.py   \n",
            "/content/PyBioMed-1.0\n",
            "/usr/local/lib/python3.6/distutils/dist.py:261: UserWarning: Unknown distribution option: 'zip_safe'\n",
            "  warnings.warn(msg)\n",
            "running install\n",
            "running build\n",
            "running build_py\n",
            "copying PyBioMed/PyDNA/PyDNApsenac(1).py -> build/lib/PyBioMed/PyDNA\n",
            "copying PyBioMed/PyDNA/PyDNAac(1).py -> build/lib/PyBioMed/PyDNA\n",
            "copying PyBioMed/PyDNA/PyDNAutil(1).py -> build/lib/PyBioMed/PyDNA\n",
            "copying PyBioMed/PyDNA/mmc3(1).data -> build/lib/PyBioMed/PyDNA\n",
            "copying PyBioMed/PyDNA/PyDNAnacutil(1).py -> build/lib/PyBioMed/PyDNA\n",
            "copying PyBioMed/PyDNA/1efz(1).pdb -> build/lib/PyBioMed/PyDNA\n",
            "copying PyBioMed/PyDNA/mmc4(1).data -> build/lib/PyBioMed/PyDNA\n",
            "copying PyBioMed/PyDNA/__init__(1).py -> build/lib/PyBioMed/PyDNA\n",
            "copying PyBioMed/PyDNA/PyDNA(1).py -> build/lib/PyBioMed/PyDNA\n",
            "copying PyBioMed/PyPretreat/PyPreTools(1).py -> build/lib/PyBioMed/PyPretreat\n",
            "copying PyBioMed/PyPretreat/PyPreTools.py -> build/lib/PyBioMed/PyPretreat\n",
            "copying PyBioMed/doc/1f88.ent(1).gz -> build/lib/PyBioMed/doc\n",
            "copying PyBioMed/doc/modules(1).rst -> build/lib/PyBioMed/doc\n",
            "copying PyBioMed/doc/1atp(1).pdb -> build/lib/PyBioMed/doc\n",
            "copying PyBioMed/doc/application(1).rst -> build/lib/PyBioMed/doc\n",
            "running install_lib\n",
            "creating /usr/local/lib/python3.6/site-packages/PyBioMed\n",
            "creating /usr/local/lib/python3.6/site-packages/PyBioMed/PyDNA\n",
            "copying build/lib/PyBioMed/PyDNA/PyDNA (1).py -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyDNA\n",
            "copying build/lib/PyBioMed/PyDNA/PyDNAutil (1).py -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyDNA\n",
            "copying build/lib/PyBioMed/PyDNA/PyDNAacutil.py -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyDNA\n",
            "copying build/lib/PyBioMed/PyDNA/PyDNAnacutil.py -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyDNA\n",
            "copying build/lib/PyBioMed/PyDNA/PyDNApsenac(1).py -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyDNA\n",
            "copying build/lib/PyBioMed/PyDNA/PyDNApsenacutil.py -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyDNA\n",
            "copying build/lib/PyBioMed/PyDNA/PyDNApsenac (1).py -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyDNA\n",
            "copying build/lib/PyBioMed/PyDNA/mmc4 (1).data -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyDNA\n",
            "copying build/lib/PyBioMed/PyDNA/PyDNA.py -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyDNA\n",
            "copying build/lib/PyBioMed/PyDNA/1efz (1).pdb -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyDNA\n",
            "copying build/lib/PyBioMed/PyDNA/PyDNAac.py -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyDNA\n",
            "copying build/lib/PyBioMed/PyDNA/PyDNAutil.py -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyDNA\n",
            "copying build/lib/PyBioMed/PyDNA/1atp.pdb -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyDNA\n",
            "copying build/lib/PyBioMed/PyDNA/PyDNAac (1).py -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyDNA\n",
            "copying build/lib/PyBioMed/PyDNA/PyDNAnac.py -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyDNA\n",
            "copying build/lib/PyBioMed/PyDNA/PyDNAnacutil (1).py -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyDNA\n",
            "copying build/lib/PyBioMed/PyDNA/__init__ (1).py -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyDNA\n",
            "copying build/lib/PyBioMed/PyDNA/PyDNAac(1).py -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyDNA\n",
            "copying build/lib/PyBioMed/PyDNA/PyDNAutil(1).py -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyDNA\n",
            "copying build/lib/PyBioMed/PyDNA/__init__.py -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyDNA\n",
            "copying build/lib/PyBioMed/PyDNA/mmc3(1).data -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyDNA\n",
            "copying build/lib/PyBioMed/PyDNA/PyDNAnacutil(1).py -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyDNA\n",
            "copying build/lib/PyBioMed/PyDNA/1efz(1).pdb -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyDNA\n",
            "copying build/lib/PyBioMed/PyDNA/mmc4(1).data -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyDNA\n",
            "copying build/lib/PyBioMed/PyDNA/mmc3.data -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyDNA\n",
            "copying build/lib/PyBioMed/PyDNA/PyDNApsenac.py -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyDNA\n",
            "copying build/lib/PyBioMed/PyDNA/__init__(1).py -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyDNA\n",
            "copying build/lib/PyBioMed/PyDNA/mmc3 (1).data -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyDNA\n",
            "copying build/lib/PyBioMed/PyDNA/1f88.pdb -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyDNA\n",
            "copying build/lib/PyBioMed/PyDNA/1efz.pdb -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyDNA\n",
            "copying build/lib/PyBioMed/PyDNA/PyDNA(1).py -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyDNA\n",
            "copying build/lib/PyBioMed/PyDNA/mmc4.data -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyDNA\n",
            "copying build/lib/PyBioMed/Pymolecule.py -> /usr/local/lib/python3.6/site-packages/PyBioMed\n",
            "copying build/lib/PyBioMed/Pyinteraction.py -> /usr/local/lib/python3.6/site-packages/PyBioMed\n",
            "copying build/lib/PyBioMed/Pydna.py -> /usr/local/lib/python3.6/site-packages/PyBioMed\n",
            "creating /usr/local/lib/python3.6/site-packages/PyBioMed/PyPretreat\n",
            "copying build/lib/PyBioMed/PyPretreat/PyPretreatMolutil.py -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyPretreat\n",
            "copying build/lib/PyBioMed/PyPretreat/PyPretreatPro.py -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyPretreat\n",
            "copying build/lib/PyBioMed/PyPretreat/PyDNAutil.py -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyPretreat\n",
            "copying build/lib/PyBioMed/PyPretreat/PyPretreatDNA.py -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyPretreat\n",
            "copying build/lib/PyBioMed/PyPretreat/__init__.py -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyPretreat\n",
            "copying build/lib/PyBioMed/PyPretreat/PyPreTools (1).py -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyPretreat\n",
            "copying build/lib/PyBioMed/PyPretreat/PyPreTools(1).py -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyPretreat\n",
            "copying build/lib/PyBioMed/PyPretreat/PyPretreatMol.py -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyPretreat\n",
            "copying build/lib/PyBioMed/PyPretreat/PyPreTools.py -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyPretreat\n",
            "creating /usr/local/lib/python3.6/site-packages/PyBioMed/PyProtein\n",
            "copying build/lib/PyBioMed/PyProtein/PyProteinAAComposition.py -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyProtein\n",
            "copying build/lib/PyBioMed/PyProtein/ConjointTriad.py -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyProtein\n",
            "copying build/lib/PyBioMed/PyProtein/aaindex1 -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyProtein\n",
            "copying build/lib/PyBioMed/PyProtein/GetProteinFromUniprot.py -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyProtein\n",
            "copying build/lib/PyBioMed/PyProtein/PyProteinAAIndex.py -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyProtein\n",
            "copying build/lib/PyBioMed/PyProtein/PseudoAAC.py -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyProtein\n",
            "copying build/lib/PyBioMed/PyProtein/QuasiSequenceOrder.py -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyProtein\n",
            "copying build/lib/PyBioMed/PyProtein/__init__.py -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyProtein\n",
            "copying build/lib/PyBioMed/PyProtein/AAComposition.py -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyProtein\n",
            "copying build/lib/PyBioMed/PyProtein/ProCheck.py -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyProtein\n",
            "copying build/lib/PyBioMed/PyProtein/GetSubSeq.py -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyProtein\n",
            "copying build/lib/PyBioMed/PyProtein/aaindex2 -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyProtein\n",
            "copying build/lib/PyBioMed/PyProtein/Autocorrelation.py -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyProtein\n",
            "copying build/lib/PyBioMed/PyProtein/CTD.py -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyProtein\n",
            "copying build/lib/PyBioMed/PyProtein/aaindex3 -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyProtein\n",
            "copying build/lib/PyBioMed/PyProtein/AAIndex.py -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyProtein\n",
            "copying build/lib/PyBioMed/PyProtein/PyProtein.py -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyProtein\n",
            "creating /usr/local/lib/python3.6/site-packages/PyBioMed/example\n",
            "creating /usr/local/lib/python3.6/site-packages/PyBioMed/example/dpi\n",
            "copying build/lib/PyBioMed/example/dpi/DPI_SMIs.xlsx -> /usr/local/lib/python3.6/site-packages/PyBioMed/example/dpi\n",
            "copying build/lib/PyBioMed/example/dpi/Enzyme.xls -> /usr/local/lib/python3.6/site-packages/PyBioMed/example/dpi\n",
            "copying build/lib/PyBioMed/example/dpi/hsa_seqs_all.tsv -> /usr/local/lib/python3.6/site-packages/PyBioMed/example/dpi\n",
            "creating /usr/local/lib/python3.6/site-packages/PyBioMed/example/dna\n",
            "copying build/lib/PyBioMed/example/dna/H_sapiens_neg.fasta -> /usr/local/lib/python3.6/site-packages/PyBioMed/example/dna\n",
            "copying build/lib/PyBioMed/example/dna/H_sapiens_pos.fasta -> /usr/local/lib/python3.6/site-packages/PyBioMed/example/dna\n",
            "creating /usr/local/lib/python3.6/site-packages/PyBioMed/example/solubility\n",
            "copying build/lib/PyBioMed/example/solubility/Solubility-total.xlsx -> /usr/local/lib/python3.6/site-packages/PyBioMed/example/solubility\n",
            "creating /usr/local/lib/python3.6/site-packages/PyBioMed/example/caco2\n",
            "copying build/lib/PyBioMed/example/caco2/caco2.xlsx -> /usr/local/lib/python3.6/site-packages/PyBioMed/example/caco2\n",
            "creating /usr/local/lib/python3.6/site-packages/PyBioMed/example/subcell\n",
            "copying build/lib/PyBioMed/example/subcell/Cytoplasm_seq.txt -> /usr/local/lib/python3.6/site-packages/PyBioMed/example/subcell\n",
            "copying build/lib/PyBioMed/example/subcell/Nuclear_seq.txt -> /usr/local/lib/python3.6/site-packages/PyBioMed/example/subcell\n",
            "creating /usr/local/lib/python3.6/site-packages/PyBioMed/doc\n",
            "copying build/lib/PyBioMed/doc/1f88.ent.gz -> /usr/local/lib/python3.6/site-packages/PyBioMed/doc\n",
            "copying build/lib/PyBioMed/doc/1f88.ent(1).gz -> /usr/local/lib/python3.6/site-packages/PyBioMed/doc\n",
            "copying build/lib/PyBioMed/doc/1atp (1).pdb -> /usr/local/lib/python3.6/site-packages/PyBioMed/doc\n",
            "copying build/lib/PyBioMed/doc/download.rst -> /usr/local/lib/python3.6/site-packages/PyBioMed/doc\n",
            "copying build/lib/PyBioMed/doc/1f88 (1).ent.gz -> /usr/local/lib/python3.6/site-packages/PyBioMed/doc\n",
            "copying build/lib/PyBioMed/doc/overview.rst -> /usr/local/lib/python3.6/site-packages/PyBioMed/doc\n",
            "copying build/lib/PyBioMed/doc/modules(1).rst -> /usr/local/lib/python3.6/site-packages/PyBioMed/doc\n",
            "copying build/lib/PyBioMed/doc/application.rst -> /usr/local/lib/python3.6/site-packages/PyBioMed/doc\n",
            "copying build/lib/PyBioMed/doc/User_guide.rst -> /usr/local/lib/python3.6/site-packages/PyBioMed/doc\n",
            "copying build/lib/PyBioMed/doc/1atp.pdb -> /usr/local/lib/python3.6/site-packages/PyBioMed/doc\n",
            "copying build/lib/PyBioMed/doc/1atp(1).pdb -> /usr/local/lib/python3.6/site-packages/PyBioMed/doc\n",
            "copying build/lib/PyBioMed/doc/Makefile -> /usr/local/lib/python3.6/site-packages/PyBioMed/doc\n",
            "copying build/lib/PyBioMed/doc/application(1).rst -> /usr/local/lib/python3.6/site-packages/PyBioMed/doc\n",
            "copying build/lib/PyBioMed/doc/modules (1).rst -> /usr/local/lib/python3.6/site-packages/PyBioMed/doc\n",
            "creating /usr/local/lib/python3.6/site-packages/PyBioMed/doc/Descriptor\n",
            "copying build/lib/PyBioMed/doc/Descriptor/Documentation.pdf -> /usr/local/lib/python3.6/site-packages/PyBioMed/doc/Descriptor\n",
            "copying build/lib/PyBioMed/doc/Descriptor/PyBioMed Chem.pdf -> /usr/local/lib/python3.6/site-packages/PyBioMed/doc/Descriptor\n",
            "copying build/lib/PyBioMed/doc/Descriptor/PyBioMed Protein.pdf -> /usr/local/lib/python3.6/site-packages/PyBioMed/doc/Descriptor\n",
            "copying build/lib/PyBioMed/doc/Descriptor/PyBioMed DNA.pdf -> /usr/local/lib/python3.6/site-packages/PyBioMed/doc/Descriptor\n",
            "copying build/lib/PyBioMed/doc/Descriptor/PyBioMed Interaction.pdf -> /usr/local/lib/python3.6/site-packages/PyBioMed/doc/Descriptor\n",
            "copying build/lib/PyBioMed/doc/application (1).rst -> /usr/local/lib/python3.6/site-packages/PyBioMed/doc\n",
            "copying build/lib/PyBioMed/doc/conf.py -> /usr/local/lib/python3.6/site-packages/PyBioMed/doc\n",
            "copying build/lib/PyBioMed/doc/index.rst -> /usr/local/lib/python3.6/site-packages/PyBioMed/doc\n",
            "copying build/lib/PyBioMed/doc/test.rst -> /usr/local/lib/python3.6/site-packages/PyBioMed/doc\n",
            "copying build/lib/PyBioMed/doc/modules.rst -> /usr/local/lib/python3.6/site-packages/PyBioMed/doc\n",
            "copying build/lib/PyBioMed/doc/make.bat -> /usr/local/lib/python3.6/site-packages/PyBioMed/doc\n",
            "copying build/lib/PyBioMed/doc/1f88.pdb -> /usr/local/lib/python3.6/site-packages/PyBioMed/doc\n",
            "copying build/lib/PyBioMed/doc/1efz.pdb -> /usr/local/lib/python3.6/site-packages/PyBioMed/doc\n",
            "copying build/lib/PyBioMed/Pyprotein.py -> /usr/local/lib/python3.6/site-packages/PyBioMed\n",
            "creating /usr/local/lib/python3.6/site-packages/PyBioMed/test\n",
            "copying build/lib/PyBioMed/test/test.py -> /usr/local/lib/python3.6/site-packages/PyBioMed/test\n",
            "copying build/lib/PyBioMed/test/test_PyProtein.py -> /usr/local/lib/python3.6/site-packages/PyBioMed/test\n",
            "copying build/lib/PyBioMed/test/test_PyBioMed.py -> /usr/local/lib/python3.6/site-packages/PyBioMed/test\n",
            "creating /usr/local/lib/python3.6/site-packages/PyBioMed/test/test_data\n",
            "copying build/lib/PyBioMed/test/test_data/neg.fasta -> /usr/local/lib/python3.6/site-packages/PyBioMed/test/test_data\n",
            "copying build/lib/PyBioMed/test/test_data/hs.fasta -> /usr/local/lib/python3.6/site-packages/PyBioMed/test/test_data\n",
            "copying build/lib/PyBioMed/test/test_data/drug.sdf -> /usr/local/lib/python3.6/site-packages/PyBioMed/test/test_data\n",
            "copying build/lib/PyBioMed/test/test_data/protein.fasta -> /usr/local/lib/python3.6/site-packages/PyBioMed/test/test_data\n",
            "copying build/lib/PyBioMed/test/test_data/example.fasta -> /usr/local/lib/python3.6/site-packages/PyBioMed/test/test_data\n",
            "copying build/lib/PyBioMed/test/test_data/pos.fasta -> /usr/local/lib/python3.6/site-packages/PyBioMed/test/test_data\n",
            "copying build/lib/PyBioMed/test/test_data/drug.smi -> /usr/local/lib/python3.6/site-packages/PyBioMed/test/test_data\n",
            "copying build/lib/PyBioMed/test/test_data/drug.mol -> /usr/local/lib/python3.6/site-packages/PyBioMed/test/test_data\n",
            "copying build/lib/PyBioMed/test/test_data/target.txt -> /usr/local/lib/python3.6/site-packages/PyBioMed/test/test_data\n",
            "copying build/lib/PyBioMed/test/test_data/drug.inchi -> /usr/local/lib/python3.6/site-packages/PyBioMed/test/test_data\n",
            "copying build/lib/PyBioMed/test/test_data/test.mol -> /usr/local/lib/python3.6/site-packages/PyBioMed/test/test_data\n",
            "copying build/lib/PyBioMed/test/test_PyPretreat.py -> /usr/local/lib/python3.6/site-packages/PyBioMed/test\n",
            "copying build/lib/PyBioMed/test/1atp.pdb -> /usr/local/lib/python3.6/site-packages/PyBioMed/test\n",
            "copying build/lib/PyBioMed/test/test_PyInteration.py -> /usr/local/lib/python3.6/site-packages/PyBioMed/test\n",
            "copying build/lib/PyBioMed/test/__init__.py -> /usr/local/lib/python3.6/site-packages/PyBioMed/test\n",
            "copying build/lib/PyBioMed/test/1atp.ent.gz -> /usr/local/lib/python3.6/site-packages/PyBioMed/test\n",
            "copying build/lib/PyBioMed/test/test_PyDNA.py -> /usr/local/lib/python3.6/site-packages/PyBioMed/test\n",
            "copying build/lib/PyBioMed/test/test_PyMolecule.py -> /usr/local/lib/python3.6/site-packages/PyBioMed/test\n",
            "copying build/lib/PyBioMed/test/test_PyGetMol.py -> /usr/local/lib/python3.6/site-packages/PyBioMed/test\n",
            "copying build/lib/PyBioMed/test/1f88.pdb -> /usr/local/lib/python3.6/site-packages/PyBioMed/test\n",
            "copying build/lib/PyBioMed/test/1efz.pdb -> /usr/local/lib/python3.6/site-packages/PyBioMed/test\n",
            "copying build/lib/PyBioMed/__init__.py -> /usr/local/lib/python3.6/site-packages/PyBioMed\n",
            "creating /usr/local/lib/python3.6/site-packages/PyBioMed/PyGetMol\n",
            "copying build/lib/PyBioMed/PyGetMol/1f88.ent.gz -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyGetMol\n",
            "copying build/lib/PyBioMed/PyGetMol/1atp.pdb -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyGetMol\n",
            "copying build/lib/PyBioMed/PyGetMol/GetDNA.py -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyGetMol\n",
            "copying build/lib/PyBioMed/PyGetMol/Getmol.py -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyGetMol\n",
            "copying build/lib/PyBioMed/PyGetMol/__init__.py -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyGetMol\n",
            "copying build/lib/PyBioMed/PyGetMol/1atp.ent.gz -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyGetMol\n",
            "copying build/lib/PyBioMed/PyGetMol/GetProtein.py -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyGetMol\n",
            "copying build/lib/PyBioMed/PyGetMol/1efz.ent.gz -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyGetMol\n",
            "copying build/lib/PyBioMed/PyGetMol/1f88.pdb -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyGetMol\n",
            "copying build/lib/PyBioMed/PyGetMol/1efz.pdb -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyGetMol\n",
            "creating /usr/local/lib/python3.6/site-packages/PyBioMed/PyInteraction\n",
            "copying build/lib/PyBioMed/PyInteraction/PyInteraction.py -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyInteraction\n",
            "copying build/lib/PyBioMed/PyInteraction/__init__.py -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyInteraction\n",
            "creating /usr/local/lib/python3.6/site-packages/PyBioMed/PyMolecule\n",
            "copying build/lib/PyBioMed/PyMolecule/topology.py -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyMolecule\n",
            "copying build/lib/PyBioMed/PyMolecule/basak.py -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyMolecule\n",
            "copying build/lib/PyBioMed/PyMolecule/Scaffolds.py -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyMolecule\n",
            "copying build/lib/PyBioMed/PyMolecule/connectivity.py -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyMolecule\n",
            "copying build/lib/PyBioMed/PyMolecule/bcut.py -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyMolecule\n",
            "copying build/lib/PyBioMed/PyMolecule/kappa.py -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyMolecule\n",
            "copying build/lib/PyBioMed/PyMolecule/cats2d.py -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyMolecule\n",
            "copying build/lib/PyBioMed/PyMolecule/AtomProperty.py -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyMolecule\n",
            "copying build/lib/PyBioMed/PyMolecule/moe.py -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyMolecule\n",
            "copying build/lib/PyBioMed/PyMolecule/MinimalFeatures.fdef -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyMolecule\n",
            "copying build/lib/PyBioMed/PyMolecule/molproperty.py -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyMolecule\n",
            "copying build/lib/PyBioMed/PyMolecule/__init__.py -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyMolecule\n",
            "copying build/lib/PyBioMed/PyMolecule/PubChemFingerprints.py -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyMolecule\n",
            "copying build/lib/PyBioMed/PyMolecule/charge.py -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyMolecule\n",
            "copying build/lib/PyBioMed/PyMolecule/AtomTypes.py -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyMolecule\n",
            "copying build/lib/PyBioMed/PyMolecule/ghosecrippen.py -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyMolecule\n",
            "copying build/lib/PyBioMed/PyMolecule/estate.py -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyMolecule\n",
            "copying build/lib/PyBioMed/PyMolecule/geary.py -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyMolecule\n",
            "copying build/lib/PyBioMed/PyMolecule/fingerprint.py -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyMolecule\n",
            "copying build/lib/PyBioMed/PyMolecule/moreaubroto.py -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyMolecule\n",
            "copying build/lib/PyBioMed/PyMolecule/Crippen.txt -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyMolecule\n",
            "copying build/lib/PyBioMed/PyMolecule/constitution.py -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyMolecule\n",
            "copying build/lib/PyBioMed/PyMolecule/moran.py -> /usr/local/lib/python3.6/site-packages/PyBioMed/PyMolecule\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/PyDNA/PyDNA (1).py to PyDNA (1).cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/PyDNA/PyDNAutil (1).py to PyDNAutil (1).cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/PyDNA/PyDNAacutil.py to PyDNAacutil.cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/PyDNA/PyDNAnacutil.py to PyDNAnacutil.cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/PyDNA/PyDNApsenac(1).py to PyDNApsenac(1).cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/PyDNA/PyDNApsenacutil.py to PyDNApsenacutil.cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/PyDNA/PyDNApsenac (1).py to PyDNApsenac (1).cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/PyDNA/PyDNA.py to PyDNA.cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/PyDNA/PyDNAac.py to PyDNAac.cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/PyDNA/PyDNAutil.py to PyDNAutil.cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/PyDNA/PyDNAac (1).py to PyDNAac (1).cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/PyDNA/PyDNAnac.py to PyDNAnac.cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/PyDNA/PyDNAnacutil (1).py to PyDNAnacutil (1).cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/PyDNA/__init__ (1).py to __init__ (1).cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/PyDNA/PyDNAac(1).py to PyDNAac(1).cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/PyDNA/PyDNAutil(1).py to PyDNAutil(1).cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/PyDNA/__init__.py to __init__.cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/PyDNA/PyDNAnacutil(1).py to PyDNAnacutil(1).cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/PyDNA/PyDNApsenac.py to PyDNApsenac.cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/PyDNA/__init__(1).py to __init__(1).cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/PyDNA/PyDNA(1).py to PyDNA(1).cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/Pymolecule.py to Pymolecule.cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/Pyinteraction.py to Pyinteraction.cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/Pydna.py to Pydna.cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/PyPretreat/PyPretreatMolutil.py to PyPretreatMolutil.cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/PyPretreat/PyPretreatPro.py to PyPretreatPro.cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/PyPretreat/PyDNAutil.py to PyDNAutil.cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/PyPretreat/PyPretreatDNA.py to PyPretreatDNA.cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/PyPretreat/__init__.py to __init__.cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/PyPretreat/PyPreTools (1).py to PyPreTools (1).cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/PyPretreat/PyPreTools(1).py to PyPreTools(1).cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/PyPretreat/PyPretreatMol.py to PyPretreatMol.cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/PyPretreat/PyPreTools.py to PyPreTools.cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/PyProtein/PyProteinAAComposition.py to PyProteinAAComposition.cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/PyProtein/ConjointTriad.py to ConjointTriad.cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/PyProtein/GetProteinFromUniprot.py to GetProteinFromUniprot.cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/PyProtein/PyProteinAAIndex.py to PyProteinAAIndex.cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/PyProtein/PseudoAAC.py to PseudoAAC.cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/PyProtein/QuasiSequenceOrder.py to QuasiSequenceOrder.cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/PyProtein/__init__.py to __init__.cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/PyProtein/AAComposition.py to AAComposition.cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/PyProtein/ProCheck.py to ProCheck.cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/PyProtein/GetSubSeq.py to GetSubSeq.cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/PyProtein/Autocorrelation.py to Autocorrelation.cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/PyProtein/CTD.py to CTD.cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/PyProtein/AAIndex.py to AAIndex.cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/PyProtein/PyProtein.py to PyProtein.cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/doc/conf.py to conf.cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/Pyprotein.py to Pyprotein.cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/test/test.py to test.cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/test/test_PyProtein.py to test_PyProtein.cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/test/test_PyBioMed.py to test_PyBioMed.cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/test/test_PyPretreat.py to test_PyPretreat.cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/test/test_PyInteration.py to test_PyInteration.cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/test/__init__.py to __init__.cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/test/test_PyDNA.py to test_PyDNA.cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/test/test_PyMolecule.py to test_PyMolecule.cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/test/test_PyGetMol.py to test_PyGetMol.cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/__init__.py to __init__.cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/PyGetMol/GetDNA.py to GetDNA.cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/PyGetMol/Getmol.py to Getmol.cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/PyGetMol/__init__.py to __init__.cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/PyGetMol/GetProtein.py to GetProtein.cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/PyInteraction/PyInteraction.py to PyInteraction.cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/PyInteraction/__init__.py to __init__.cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/PyMolecule/topology.py to topology.cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/PyMolecule/basak.py to basak.cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/PyMolecule/Scaffolds.py to Scaffolds.cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/PyMolecule/connectivity.py to connectivity.cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/PyMolecule/bcut.py to bcut.cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/PyMolecule/kappa.py to kappa.cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/PyMolecule/cats2d.py to cats2d.cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/PyMolecule/AtomProperty.py to AtomProperty.cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/PyMolecule/moe.py to moe.cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/PyMolecule/molproperty.py to molproperty.cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/PyMolecule/__init__.py to __init__.cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/PyMolecule/PubChemFingerprints.py to PubChemFingerprints.cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/PyMolecule/charge.py to charge.cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/PyMolecule/AtomTypes.py to AtomTypes.cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/PyMolecule/ghosecrippen.py to ghosecrippen.cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/PyMolecule/estate.py to estate.cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/PyMolecule/geary.py to geary.cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/PyMolecule/fingerprint.py to fingerprint.cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/PyMolecule/moreaubroto.py to moreaubroto.cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/PyMolecule/constitution.py to constitution.cpython-36.pyc\n",
            "byte-compiling /usr/local/lib/python3.6/site-packages/PyBioMed/PyMolecule/moran.py to moran.cpython-36.pyc\n",
            "running install_egg_info\n",
            "Writing /usr/local/lib/python3.6/site-packages/PyBioMed-1.0-py3.6.egg-info\n",
            "/content\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "_nOemLT-uGyY",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "# from google.colab import drive\n",
        "# drive.mount('/content/drive')"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "hccJpfWGuYj2",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "# %cd /content/drive/My Drive/PyBioMed-1.0\n",
        "# %pwd\n",
        "# !python3 setup.py install\n",
        "# %cd .."
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "c__Q6CXN6aNg",
        "colab_type": "text"
      },
      "source": [
        "Now you must **Restart the** **Runtime**"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "v02WyYYsG8sV",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "import pandas as pd\n",
        "import numpy as np\n",
        "import tensorflow as tf\n",
        "from google.colab import drive\n",
        "from google.colab import files\n",
        "import os\n",
        "from rdkit import Chem\n",
        "from Bio import SeqIO\n",
        "from PyBioMed.PyProtein import AAComposition\n",
        "from PyBioMed.PyProtein import CTD\n",
        "from PyBioMed.PyProtein import ConjointTriad,Autocorrelation\n",
        "from mordred import Calculator,descriptors,Weight\n",
        "import tqdm\n",
        "import random\n",
        "from sklearn.model_selection import train_test_split\n",
        "from sklearn.svm import SVC\n",
        "from sklearn.ensemble import RandomForestClassifier\n",
        "import random\n",
        "from sklearn.model_selection import train_test_split\n",
        "from sklearn.ensemble import RandomForestClassifier\n",
        "from sklearn.svm import SVC\n",
        "from mordred import Calculator,descriptors,Weight\n",
        "from tqdm import tqdm\n",
        "import time\n",
        "import pennylane as qml\n",
        "from pennylane import numpy as np\n",
        "import pennylane.optimize\n",
        "import os\n",
        "import time\n",
        "import random\n",
        "import pickle\n",
        "from tqdm import tqdm\n",
        "import matplotlib.pyplot as plt\n",
        "import torch\n",
        "import torch.nn as nn\n",
        "import torch.optim as optim\n",
        "from torch.autograd import Variable\n",
        "from torchsummary import summary\n",
        "import torchvision.datasets as dset\n",
        "import torchvision.transforms as transforms\n",
        "import math\n",
        "from qiskit import BasicAer\n",
        "from qiskit.aqua.algorithms import QSVM\n",
        "from qiskit.aqua import run_algorithm, QuantumInstance\n",
        "from qiskit.aqua.components.feature_maps import SecondOrderExpansion\n",
        "from qiskit import IBMQ\n",
        "import scipy\n",
        "from scipy.linalg import expm\n",
        "import matplotlib.pyplot as plt\n",
        "from mpl_toolkits.mplot3d import Axes3D\n",
        "from sklearn import datasets\n",
        "from sklearn.model_selection import train_test_split\n",
        "from sklearn.preprocessing import StandardScaler, MinMaxScaler\n",
        "from sklearn.decomposition import PCA"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "xT9teSiX7DQp",
        "colab_type": "text"
      },
      "source": [
        "## Data Download and Preprocessing"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "ctyylfNXMZEB",
        "colab_type": "text"
      },
      "source": [
        "### dataPrep():\n",
        "    This function can be used to generate the molecular descriptors from the drugBank dataset and store them as CSV files, which are later accessed by the preprocess function. The csv files used in the dataPrep function were obtained from the drugbank dataset. New csv files will be saved in the environment which will be used in the preprocess() function.\n",
        "\n",
        "### preprocess():\n",
        "    This function forms the training and test dataset from the csv files saved using the dataPrep() function.\n",
        "\n",
        "It is not required to run the dataPrep() function as we have already provided the csv files with the descriptors calulated."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "PIn-hg7OMYga",
        "colab_type": "text"
      },
      "source": [
        ""
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "IUKL0DZT6k24",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "config = {\n",
        "    'records': 100,\n",
        "    'qbits': 10\n",
        "}\n",
        "def getProteinDescriptor(protSeq):\n",
        "    res = {}\n",
        "    res.update(AAComposition.CalculateAAComposition(protSeq))\n",
        "    res.update(AAComposition.CalculateDipeptideComposition(protSeq))\n",
        "    res.update(Autocorrelation.CalculateNormalizedMoreauBrotoAutoTotal(protSeq))\n",
        "    res.update(Autocorrelation.CalculateMoranAutoTotal(protSeq))\n",
        "    res.update(ConjointTriad.CalculateConjointTriad(protSeq))\n",
        "    return res\n",
        "\n",
        "def preprocess():\n",
        "    proteinDescriptorsDF = pd.read_csv('/content/protein_descriptor.csv')\n",
        "    drugTargetPair_positive = pd.read_csv('/content/positive_pair.csv')\n",
        "    drugTargetPair_negative = pd.read_csv('/content/negative_pair.csv')\n",
        "    drugList_positive = drugTargetPair_positive['Drug'].tolist()\n",
        "    targetList_positive = drugTargetPair_positive['Target'].tolist()\n",
        "\n",
        "    drugList_negative = drugTargetPair_negative['Drug'].tolist()\n",
        "    targetList_negative = drugTargetPair_negative['Target'].tolist()\n",
        "\n",
        "    drugDescriptors = pd.read_csv('/content/molecule_descriptor_new.csv')\n",
        "\n",
        "    drugDescriptorsNumeric = drugDescriptors.apply(pd.to_numeric,errors='coerce')\n",
        "    drugDescriptorsNumeric = drugDescriptorsNumeric.dropna(axis = 1)\n",
        "    proteinDescriptorsDFNumeric = proteinDescriptorsDF.apply(pd.to_numeric,errors='coerce')\n",
        "    proteinDescriptorsDFNumeric = proteinDescriptorsDFNumeric.dropna(axis=1)\n",
        "\n",
        "    drugDescriptorID = drugDescriptors['ID'].tolist()\n",
        "\n",
        "    finalDrugTargetPair_positive = []\n",
        "    targetDescriptorID = proteinDescriptorsDF['Protein ID'].tolist()\n",
        "    Y_positive = []\n",
        "    for i in tqdm(range(len(targetList_positive))):\n",
        "        targetID = targetList_positive[i]\n",
        "        drugID = drugList_positive[i]\n",
        "        if targetID in targetDescriptorID and drugID in drugDescriptorID:\n",
        "            indexTarget = targetDescriptorID.index(targetID)\n",
        "            indexDrug = drugDescriptorID.index(drugID)\n",
        "            proteinDescriptorsArray = proteinDescriptorsDFNumeric.iloc[indexTarget].to_numpy().astype(np.float)\n",
        "            drugDescriptorArray = drugDescriptorsNumeric.iloc[indexDrug].to_numpy().astype(np.float)\n",
        "            pairDescriptor = np.concatenate([proteinDescriptorsArray,drugDescriptorArray])\n",
        "            finalDrugTargetPair_positive.append(([drugID,targetID],pairDescriptor))\n",
        "        else:\n",
        "            continue\n",
        "    print(len(finalDrugTargetPair_positive))\n",
        "    print(len(finalDrugTargetPair_positive[0][1]))\n",
        "\n",
        "    # Cell number 2\n",
        "\n",
        "    Y_positive = [1] * len(finalDrugTargetPair_positive)\n",
        "\n",
        "    finalDrugTargetPair_negative = []\n",
        "    for i in tqdm(range(len(targetList_positive))):\n",
        "        targetID = targetList_negative[i]\n",
        "        drugID = drugList_negative[i]\n",
        "        if targetID in targetDescriptorID and drugID in drugDescriptorID:\n",
        "            indexTarget = targetDescriptorID.index(targetID)\n",
        "            indexDrug = drugDescriptorID.index(drugID)\n",
        "            proteinDescriptorsArray = proteinDescriptorsDFNumeric.iloc[indexTarget].to_numpy().astype(np.float)\n",
        "            drugDescriptorArray = drugDescriptorsNumeric.iloc[indexDrug].to_numpy().astype(np.float)\n",
        "            pairDescriptor = np.concatenate([proteinDescriptorsArray,drugDescriptorArray])\n",
        "            finalDrugTargetPair_negative.append(([drugID,targetID],pairDescriptor))\n",
        "        else:\n",
        "            continue\n",
        "    Y_negative = [0] * len(finalDrugTargetPair_negative)\n",
        "\n",
        "    #Cell number 3\n",
        "\n",
        "    X = np.empty((len(finalDrugTargetPair_positive)+len(finalDrugTargetPair_negative),len(finalDrugTargetPair_positive[0][1])))\n",
        "    count = 0\n",
        "    for info in finalDrugTargetPair_positive:\n",
        "        X[count] = info[1]\n",
        "        count += 1\n",
        "    for info in finalDrugTargetPair_negative:\n",
        "        X[count] = info[1]\n",
        "        count += 1\n",
        "    Y = Y_positive + Y_negative\n",
        "\n",
        "    #Cell number 4\n",
        "\n",
        "    import random\n",
        "    from sklearn.model_selection import train_test_split\n",
        "    temp = []\n",
        "    for i in range(len(X)):\n",
        "        temp.append([X[i],Y[i]])\n",
        "    random.shuffle(temp)\n",
        "    for i in range(len(temp)):\n",
        "        X[i] = temp[i][0]\n",
        "        Y[i] = temp[i][1]\n",
        "    X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size=0.33)\n",
        "    standardScalerObject = StandardScaler().fit(X_train)\n",
        "    sample_train = standardScalerObject.transform(X_train)\n",
        "    sample_test = standardScalerObject.transform(X_test)\n",
        "\n",
        "    pca = PCA(n_components=config['qbits']).fit(sample_train)\n",
        "    sample_train = pca.transform(sample_train)\n",
        "    sample_test = pca.transform(sample_test)\n",
        "    samples = np.append(sample_train,sample_test,axis=0)\n",
        "    minmax_scale = MinMaxScaler((-1,1)).fit(samples)\n",
        "    sample_train = minmax_scale.transform(sample_train)\n",
        "    sample_test = minmax_scale.transform(sample_test)\n",
        "    class_labels = [1,0]\n",
        "    label_train, label_test = y_train, y_test\n",
        "    training_size = len(y_train)\n",
        "    test_size = len(y_test)\n",
        "    sample_train_positive = np.empty((len(finalDrugTargetPair_positive),config['qbits']))\n",
        "    sample_train_negative = np.empty((len(finalDrugTargetPair_negative),config['qbits']))\n",
        "    count_pos,count_neg = 0,0\n",
        "    for i in range(len(y_train)):\n",
        "        if y_train[i] == 1:\n",
        "            sample_train_positive[count_pos] = sample_train[i]\n",
        "            count_pos += 1\n",
        "        else:\n",
        "            sample_train_negative[count_neg] = sample_train[i]\n",
        "            count_neg += 1\n",
        "    sample_test_positive = np.empty((len(finalDrugTargetPair_positive),config['qbits']))\n",
        "    sample_test_negative = np.empty((len(finalDrugTargetPair_negative),config['qbits']))\n",
        "    count_pos,count_neg = 0,0        \n",
        "    for i in range(len(y_test)):\n",
        "        if y_test[i] == 1:\n",
        "            sample_test_positive[count_pos] = sample_train[i]\n",
        "            count_pos += 1\n",
        "        else:\n",
        "            sample_test_negative[count_neg] = sample_train[i]\n",
        "            count_neg += 1\n",
        "    print(sample_train_positive.shape) \n",
        "    print(sample_train_negative.shape)\n",
        "    print(sample_test_negative.shape)\n",
        "    print(sample_test_positive.shape)\n",
        "    training_input = {1:sample_train_positive,0:sample_train_negative}\n",
        "    test_input = {1:sample_test_positive,0:sample_test_negative}\n",
        "    sample_train = np.concatenate((sample_train_negative,sample_train_positive),axis=0)\n",
        "    sample_y_train = np.concatenate((np.zeros((sample_train_negative.shape[0],)),np.ones((sample_train_positive.shape[0],))),axis=0)\n",
        "    print(sample_train.shape)\n",
        "    print(sample_y_train.shape)\n",
        "    import random\n",
        "    from sklearn.model_selection import train_test_split\n",
        "    temp = []\n",
        "    for i in range(len(X)):\n",
        "        temp.append([sample_train[i],sample_y_train[i]])\n",
        "    random.shuffle(temp)\n",
        "    for i in range(len(temp)):\n",
        "        sample_train[i] = temp[i][0]\n",
        "        sample_y_train[i] = temp[i][1]\n",
        "    print(sample_train.shape)\n",
        "    print(sample_y_train.shape)\n",
        "    sample_test = np.concatenate((sample_test_negative,sample_test_positive),axis=0)\n",
        "    sample_y_test = np.concatenate((np.zeros((sample_test_negative.shape[0],)),np.ones((sample_test_positive.shape[0],))),axis=0)\n",
        "    print(sample_test.shape)\n",
        "    print(sample_y_test.shape)\n",
        "    import random\n",
        "    from sklearn.model_selection import train_test_split\n",
        "    temp = []\n",
        "    for i in range(len(X)):\n",
        "        temp.append([sample_test[i],sample_y_test[i]])\n",
        "    random.shuffle(temp)\n",
        "    for i in range(len(temp)):\n",
        "        sample_test[i] = temp[i][0]\n",
        "        sample_y_test[i] = temp[i][1]\n",
        "    print(sample_test.shape)\n",
        "    print(sample_y_test.shape)\n",
        "\n",
        "    import math\n",
        "    for i in training_input:\n",
        "        for j,_ in enumerate(training_input[i]):\n",
        "            for k,_ in enumerate(training_input[i][j]):\n",
        "                if math.isnan(training_input[i][j][k]):\n",
        "                    training_input[i][j][k] = 0\n",
        "    for i in test_input:\n",
        "        for j,_ in enumerate(test_input[i]):\n",
        "            for k,_ in enumerate(test_input[i][j]):\n",
        "                if math.isnan(test_input[i][j][k]):\n",
        "                    test_input[i][j][k] = 0\n",
        "\n",
        "    for i in training_input:\n",
        "        training_input[i] = training_input[i][:config['records']]\n",
        "    for i in test_input:\n",
        "        test_input[i] = test_input[i][:config['records']]\n",
        "    X_train = []\n",
        "    y_train = []\n",
        "    for i in training_input:\n",
        "        for j in training_input[i]:\n",
        "            X_train.append(j)\n",
        "            y_train.append(i)\n",
        "    c = list(zip(X_train, y_train))\n",
        "    random.shuffle(c)\n",
        "    X_train, y_train = zip(*c)\n",
        "\n",
        "    X_test = []\n",
        "    y_test = []\n",
        "    for i in test_input:\n",
        "        for j in test_input[i]:\n",
        "            X_test.append(j)\n",
        "            y_test.append(i)\n",
        "    c = list(zip(X_test, y_test))\n",
        "    random.shuffle(c)\n",
        "    X_test, y_test = zip(*c)\n",
        "    X_train = np.array(X_train)\n",
        "    y_train = np.array(y_train)\n",
        "    X_test = np.array(X_test)\n",
        "    y_test = np.array(y_test)\n",
        "    print(training_input)\n",
        "    print(test_input)\n",
        "    return X_train,X_test,y_train,y_test,training_input,test_input\n",
        "\n",
        "def dataPrep():\n",
        "    target = pd.read_csv(\"/content/drive/My Drive/all.csv\")\n",
        "    drugs = pd.read_csv(\"/content/drive/My Drive/drug_links.csv\")\n",
        "    structures = pd.read_csv(\"/content/drive/My Drive/structure_links.csv\")\n",
        "    structures = structures[['DrugBank ID','Name','SMILES']]\n",
        "    structures = structures.dropna()\n",
        "    smilesList = structures['SMILES'].tolist()\n",
        "    idList = structures['DrugBank ID'].tolist()\n",
        "    drugsInteracting = target['Drug IDs'].tolist()\n",
        "    drugsSet = set()\n",
        "    for i in drugsInteracting:\n",
        "        lsTemp = i.split('; ')\n",
        "        for drugId in lsTemp:\n",
        "            drugsSet.add(drugId)\n",
        "    drugsInteracting = list(drugsSet)\n",
        "    structures = structures[structures['DrugBank ID'].isin(drugsInteracting)]\n",
        "    smilesList = structures['SMILES'].tolist()\n",
        "    idList = structures['DrugBank ID'].tolist()\n",
        "    calc = Calculator(descriptors, ignore_3D=True)\n",
        "    descriptorMolecule = {}\n",
        "    discardCount = 0\n",
        "    newMoleculeList, newDescriptorList = [],[]\n",
        "    molList = []\n",
        "    for i in range(len(molList)):\n",
        "        tempDescriptorContainer = calc(molList[i])\n",
        "        if len(tempDescriptorContainer) == 1613:\n",
        "            newDescriptorList.append(tempDescriptorContainer)\n",
        "            newMoleculeList.append(molList[i])\n",
        "        else:\n",
        "            discardCount += 1\n",
        "    moleculesListDF = pd.DataFrame(newDescriptorList)\n",
        "    moleculesListDF.insert(0,'Molecules',molList)\n",
        "    moleculesListDF.insert(0,'ID',structures['DrugBank ID'].tolist())\n",
        "    export_csv = moleculesListDF.to_csv(r'/content/drive/My Drive/molecule_descriptor_full_dataset.csv',index = None, header=True)\n",
        "    descriptorsCSV = pd.read_csv(\"/content/drive/My Drive/molecule_descriptor_full_dataset.csv\")\n",
        "    drugsInDescriptors = descriptorsCSV['ID'].tolist()\n",
        "    data = []\n",
        "    proteinDescriptors = []\n",
        "    proteinList = []\n",
        "    drugTargetPair = {}\n",
        "    for seq_record in SeqIO.parse(\"/content/drive/My Drive/protein.fasta\", \"fasta\"):\n",
        "        targetDrug = seq_record.description\n",
        "        sequenceProtein = str(seq_record.seq)\n",
        "        proteinID = str(seq_record.id).split('|')[1]\n",
        "        targetDrugs = targetDrug[targetDrug.find(\"(\")+1:targetDrug.find(\")\")]\n",
        "        targetDrugs = targetDrugs.split(\"; \")\n",
        "        data.append([sequenceProtein,(proteinID,targetDrugs)])\n",
        "    drug_pair_negative = []\n",
        "    drug_pair_positive = []\n",
        "    drugDecoyStr_negative = []\n",
        "    drugDecoyStr_positive = []\n",
        "    for item in data:\n",
        "        decoyID = item[1][0]\n",
        "        drugsList = item[1][1]\n",
        "        for drug in drugsInDescriptors:\n",
        "            if drug not in drugsList:\n",
        "                drug_pair_negative.append([drug,decoyID])\n",
        "                drugDecoyStr_negative.append(drug+decoyID)\n",
        "            else:\n",
        "                drug_pair_positive.append([drug,decoyID])\n",
        "                drugDecoyStr_positive.append(drug+decoyID)\n",
        "    random.shuffle(drug_pair_negative)\n",
        "    random.shuffle(drug_pair_positive)\n",
        "    drug_pair_negative_dataframe = pd.DataFrame(drug_pair_negative)\n",
        "    drug_pair_positive_dataframe = pd.DataFrame(drug_pair_positive)\n",
        "    drug_pair_negative_dataframe = drug_pair_negative_dataframe.rename(columns={0: \"Drug\", 1: \"Target\"})\n",
        "    drug_pair_positive_dataframe = drug_pair_positive_dataframe.rename(columns={0: \"Drug\", 1: \"Target\"})\n",
        "    exported_drug_pair_negative = drug_pair_positive_dataframe.to_csv(r'/content/drive/My Drive/positive_pair.csv',index = None, header=True)\n",
        "    exported_drug_pair_positive = drug_pair_negative_dataframe.to_csv(r'/content/drive/My Drive/negative_pair.csv',index = None, header=True)\n",
        "    drugTargetList = []\n",
        "    proteinDescriptors = []\n",
        "    for i in range(len(data)):\n",
        "        mergedData = data[i]\n",
        "        proteinSequence = mergedData[0]\n",
        "        proteinId = mergedData[1][0]\n",
        "        drugs = mergedData[1][1]\n",
        "    #     print(i)        \n",
        "        try:\n",
        "            proteinDescriptors.append([proteinId,getProteinDescriptor(proteinSequence)])\n",
        "        except:\n",
        "            print(\"Exception Occured for proteinID {}\".format(proteinId))\n",
        "    descriptorsList = []        \n",
        "    proteinIDList = []\n",
        "    for descriptor in proteinDescriptors:\n",
        "        descriptorsList.append(descriptor[1])\n",
        "        proteinIDList.append(descriptor[0])\n",
        "    proteinDescriptorsDF = pd.DataFrame(descriptorsList)\n",
        "    proteinDescriptorsDF.insert(0,\"Protein ID\",proteinIDList)\n",
        "    export_csv = proteinDescriptorsDF.to_csv('/content/drive/My Drive/protein_descriptor_decoy.csv',index = None, header=True)\n"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "sRj4lTjqjaLl",
        "colab_type": "code",
        "outputId": "8ea2119b-6bfb-4057-d5f1-b4a5547638c1",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        }
      },
      "source": [
        "# dataPrep()\n",
        "X_train,X_test,y_train,y_test,training_input,test_input = preprocess()"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "/usr/local/lib/python3.6/dist-packages/IPython/core/interactiveshell.py:2822: DtypeWarning: Columns (54,55,56,57,58,59,60,61,62,90,91,92,93,94,95,96,97,98,108,109,110,111,112,113,114,115,116,136,137,145,146,153,154,155,163,164,172,173,181,182,189,190,191,199,200,207,208,209,217,218,226,227,261,262,263,264,265,266,267,268,269,297,298,299,300,301,302,303,304,305,315,316,317,318,319,320,321,322,323,352,353,361,362,369,370,371,379,380,388,389,397,398,405,406,407,415,416,423,424,425,433,434,442,443,458,459,466,467,474,475,482,483,490,491,498,499,506,507,514,515,522,523,530,531,538,539,554,555,562,563,570,571,578,579,586,587,594,595,602,603,610,611,618,619,626,627,634,635,808,818,826,827,828,834,835,836,842,843,844,853,855,861,863,1270,1271,1272,1273,1274,1275,1276,1277,1278,1279,1280,1281,1300,1301,1303,1304,1553,1581,1582,1584,1612) have mixed types. Specify dtype option on import or set low_memory=False.\n",
            "  if self.run_code(code, result):\n",
            "100%|██████████| 9144/9144 [00:11<00:00, 770.03it/s]\n",
            "  1%|          | 92/9144 [00:00<00:09, 906.91it/s]"
          ],
          "name": "stderr"
        },
        {
          "output_type": "stream",
          "text": [
            "6669\n",
            "1821\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "stream",
          "text": [
            "100%|██████████| 9144/9144 [00:10<00:00, 876.30it/s]\n"
          ],
          "name": "stderr"
        },
        {
          "output_type": "stream",
          "text": [
            "(6669, 10)\n",
            "(5387, 10)\n",
            "(5387, 10)\n",
            "(6669, 10)\n",
            "(12056, 10)\n",
            "(12056,)\n",
            "(12056, 10)\n",
            "(12056,)\n",
            "(12056, 10)\n",
            "(12056,)\n",
            "(12056, 10)\n",
            "(12056,)\n",
            "{1: array([[-0.59736586, -0.84287605, -0.21604904, -0.37968079, -0.17532904,\n",
            "        -0.46774585, -0.61702309, -0.54803341, -0.11118806, -0.29614966],\n",
            "       [-0.49652731, -0.88480255, -0.2380466 , -0.49388601,  0.006806  ,\n",
            "        -0.42797704, -0.57089429, -0.50289106, -0.01213286, -0.2382022 ],\n",
            "       [-0.3472722 , -0.88047998, -0.30654392, -0.58606654,  0.0288498 ,\n",
            "        -0.40931132, -0.51758397, -0.59252132, -0.09532696, -0.2077864 ],\n",
            "       [-0.47376175, -0.82561516, -0.22122838, -0.33452351, -0.07594053,\n",
            "        -0.62034981, -0.56983797, -0.57360479, -0.23558809, -0.1843496 ],\n",
            "       [-0.62464572, -0.8984194 , -0.17304913, -0.31931683,  0.16395037,\n",
            "        -0.58335579, -0.58959955, -0.47807808, -0.0263573 , -0.1615436 ],\n",
            "       [-0.49481028, -0.90782725, -0.07585673, -0.41960996, -0.1691485 ,\n",
            "        -0.34362659, -0.69683074, -0.54924067, -0.11282866, -0.24822583],\n",
            "       [-0.42998237, -0.86274909, -0.23280999, -0.38043282,  0.05228959,\n",
            "        -0.64664048, -0.65140637, -0.46828982, -0.11301707, -0.19730639],\n",
            "       [-0.60650947, -0.85528365, -0.16797018, -0.41839681, -0.00606932,\n",
            "        -0.44123273, -0.58563857, -0.56023411, -0.1614541 , -0.29380775],\n",
            "       [-0.38005207, -0.87256178, -0.24357398, -0.61665812,  0.44249468,\n",
            "        -0.47170895, -0.53975917, -0.54588484, -0.11042037, -0.26765475],\n",
            "       [-0.56247346, -0.80979135, -0.18740956, -0.4461678 ,  0.10399931,\n",
            "        -0.37810269, -0.5278702 , -0.59497504, -0.07741334, -0.25929557],\n",
            "       [-0.44106289, -0.8300251 , -0.30131652, -0.42060193, -0.21284934,\n",
            "        -0.40062716, -0.61081118, -0.51751675, -0.09901989, -0.27266052],\n",
            "       [-0.69507489, -0.86903479, -0.17374833, -0.4322206 , -0.15534452,\n",
            "        -0.63379253, -0.59459878, -0.50751856, -0.13109363, -0.2090643 ],\n",
            "       [-0.37495511, -0.86499912, -0.29316705, -0.56327911,  0.33283135,\n",
            "        -0.51459344, -0.51544827, -0.56912438, -0.09891331, -0.23735837],\n",
            "       [-0.53460531, -0.8603687 , -0.09771522, -0.40875318,  0.05457382,\n",
            "        -0.45665313, -0.56597458, -0.55088617, -0.17626257, -0.20085169],\n",
            "       [-0.27731422, -0.74323126, -0.07843286, -0.41765254,  0.06318853,\n",
            "        -0.48850686, -0.63491567, -0.44482422, -0.08403741, -0.18165881],\n",
            "       [-0.37774778, -0.86841666, -0.24329793, -0.61790826,  0.4447539 ,\n",
            "        -0.46995334, -0.5385439 , -0.54893259, -0.10884038, -0.27227371],\n",
            "       [-0.65371645, -0.86751256, -0.15995712, -0.45511655, -0.0121729 ,\n",
            "        -0.57365145, -0.53174257, -0.57721563, -0.26700915, -0.27743275],\n",
            "       [-0.34864708, -0.88432345, -0.36804002, -0.36623662, -0.09526137,\n",
            "        -0.54254104, -0.59616682, -0.5198837 ,  0.03543308, -0.21679148],\n",
            "       [-0.37112318, -0.89510739, -0.25803136, -0.56596346, -0.04009255,\n",
            "        -0.34757062, -0.54838531, -0.62638342, -0.08257342, -0.24349693],\n",
            "       [-0.53548606, -0.81853379, -0.18192701, -0.45097989, -0.18859529,\n",
            "        -0.48502046, -0.61037799, -0.55111365, -0.09255284, -0.2983631 ],\n",
            "       [-0.45599631, -0.8695159 , -0.2175562 , -0.47151705, -0.09563195,\n",
            "        -0.62207364, -0.50623542, -0.59167253, -0.2177129 , -0.32790701],\n",
            "       [-0.4087352 , -0.79479201, -0.25601054, -0.27237506, -0.03041825,\n",
            "        -0.55876655, -0.60676511, -0.49170882, -0.00751041, -0.26239641],\n",
            "       [-0.41273414, -0.93530551, -0.192477  , -0.48524646,  0.05663935,\n",
            "        -0.49016405, -0.64055936, -0.46182853, -0.16729286, -0.19002064],\n",
            "       [-0.42791331, -0.85975763, -0.27840036, -0.53895307, -0.0387406 ,\n",
            "        -0.31584613, -0.533135  , -0.66106179, -0.09406389, -0.29159445],\n",
            "       [-0.50868555, -0.90133932, -0.14745175, -0.42454155, -0.13965605,\n",
            "        -0.48411214, -0.67751674, -0.45862928, -0.02267904, -0.21374855],\n",
            "       [-0.57101235, -0.84023754, -0.2098617 , -0.52269259, -0.10976779,\n",
            "        -0.36873548, -0.66863112, -0.48004568, -0.11699709, -0.22477356],\n",
            "       [-0.51966435, -0.89712618, -0.29941152, -0.35953565,  0.18147219,\n",
            "        -0.4954719 , -0.70437882, -0.43056339, -0.10584611, -0.15441661],\n",
            "       [-0.36554903, -0.86563828, -0.1047937 , -0.41722077, -0.06271515,\n",
            "        -0.48820637, -0.59012815, -0.57182964, -0.04657355, -0.26816343],\n",
            "       [-0.49315962, -0.87326049, -0.19631187, -0.45796344,  0.05847622,\n",
            "        -0.36564873, -0.53461753, -0.59721066, -0.11854842, -0.28169041],\n",
            "       [ 0.54530663, -0.87944866, -0.38608435, -0.87604893, -0.95812202,\n",
            "        -0.75186382, -0.63194939, -0.54912971, -0.419634  , -0.10835131],\n",
            "       [-0.37852341, -0.87293394, -0.29371871, -0.56263105,  0.32990272,\n",
            "        -0.51161657, -0.51639741, -0.56814446, -0.10011578, -0.23551744],\n",
            "       [-0.52685399, -0.8766105 , -0.08750603, -0.40242055, -0.19394218,\n",
            "        -0.60333979, -0.65322283, -0.52032984, -0.11371886, -0.25815175],\n",
            "       [-0.64570741, -0.89793056, -0.18201916, -0.27609066, -0.10015413,\n",
            "        -0.66683315, -0.58172775, -0.50986462, -0.02847666, -0.22243098],\n",
            "       [-0.66692736, -0.8749859 , -0.15111019, -0.50283089,  0.00912681,\n",
            "        -0.53848572, -0.56747933, -0.55269638, -0.25295608, -0.26990381],\n",
            "       [-0.47091011, -0.87906252, -0.16022087, -0.3950789 , -0.12931581,\n",
            "        -0.55741992, -0.58506161, -0.52346862, -0.08280662, -0.21962334],\n",
            "       [-0.28216031, -0.89121166, -0.08718499, -0.49707535, -0.06953414,\n",
            "        -0.57395895, -0.52132559, -0.58316107, -0.10725707, -0.19840551],\n",
            "       [-0.80468755, -0.8638677 , -0.19596503, -0.50209525, -0.38077854,\n",
            "        -0.31206671, -0.50812822, -0.61389176,  0.00306551, -0.25615179],\n",
            "       [ 0.09001639, -0.86902457, -0.44886969, -0.38213983, -0.08522875,\n",
            "        -0.52254833, -0.4821468 , -0.70628488, -0.09790312, -0.32080704],\n",
            "       [-0.36825937, -0.81484103, -0.32118516, -0.42703043, -0.33792748,\n",
            "        -0.62745089, -0.51814753, -0.61779353, -0.11597471, -0.30113237],\n",
            "       [-0.42417333, -0.89516975,  0.0669227 , -0.43651945,  0.11298483,\n",
            "        -0.58032469, -0.64066331, -0.47499765, -0.11129901, -0.25447299],\n",
            "       [-0.55349359, -0.82600352, -0.09067208, -0.39236991, -0.03671569,\n",
            "        -0.54428483, -0.70567478, -0.48797062, -0.16267235, -0.26958441],\n",
            "       [-0.56275535, -0.87506346, -0.18406992, -0.41382907,  0.08589768,\n",
            "        -0.64942532, -0.61814513, -0.51762666, -0.16347586, -0.22329137],\n",
            "       [-0.50342706, -0.84471971, -0.19647271, -0.43372757, -0.04166723,\n",
            "        -0.47697783, -0.61210316, -0.51676564, -0.03601721, -0.27231528],\n",
            "       [-0.65659762, -0.89310314, -0.21144632, -0.43086576, -0.31866916,\n",
            "        -0.42349946, -0.59875085, -0.50546831, -0.08361971, -0.24767405],\n",
            "       [-0.58988254, -0.88923876, -0.25187756, -0.56845837, -0.36234968,\n",
            "        -0.17884928, -0.76752111, -0.51542917, -0.14583796, -0.29928584],\n",
            "       [-0.01449224, -0.93018713,  0.14446061, -0.42687862, -0.07279182,\n",
            "        -0.55137724, -0.56729039, -0.64005911, -0.05423968, -0.19709024],\n",
            "       [-0.29195389, -0.90036068,  0.02592118, -0.44634317, -0.01713415,\n",
            "        -0.51838439, -0.66820384, -0.52033203, -0.10436265, -0.24548087],\n",
            "       [-0.03899534, -0.91186896, -0.25398885, -0.65802623, -0.22174268,\n",
            "        -0.75325815, -0.5070924 , -0.6711485 , -0.27959085, -0.25766979],\n",
            "       [-0.52630133, -0.86067457, -0.18783295, -0.5205636 ,  0.104596  ,\n",
            "        -0.45258011, -0.62385417, -0.52859167, -0.10713589, -0.23912444],\n",
            "       [-0.54469549, -0.88965372, -0.11562308, -0.48003237,  0.0177207 ,\n",
            "        -0.64592288, -0.65182875, -0.47274177, -0.10144874, -0.22297945],\n",
            "       [-0.51329748, -0.78551635, -0.27263568, -0.4750972 , -0.0657062 ,\n",
            "        -0.28861158, -0.5540488 , -0.56939435, -0.07136655, -0.28653286],\n",
            "       [-0.39463878, -0.89727595,  0.01601242, -0.33483432, -0.0098728 ,\n",
            "        -0.52583203, -0.60744748, -0.52536741, -0.04779899, -0.24318592],\n",
            "       [-0.25915243, -0.90407161, -0.40683522, -0.24812393, -0.16554521,\n",
            "        -0.73102161, -0.55763495, -0.47267715,  0.06176744, -0.17254177],\n",
            "       [-0.57338402, -0.86048427, -0.17856409, -0.48054084,  0.294468  ,\n",
            "        -0.28629017, -0.6040495 , -0.56759135, -0.06632416, -0.29799958],\n",
            "       [-0.49102757, -0.88434713, -0.30901901, -0.36726179, -0.02486141,\n",
            "        -0.54361277, -0.60072821, -0.49966066, -0.13659267, -0.20659913],\n",
            "       [-0.48504701, -0.85219424, -0.30130636, -0.51607027,  0.20637937,\n",
            "        -0.37638089, -0.57395745, -0.52570251,  0.01048434, -0.24789913],\n",
            "       [-0.65225399, -0.89298971, -0.22187461, -0.39034625, -0.23280473,\n",
            "        -0.36375193, -0.65686237, -0.51122788, -0.20980222, -0.24401745],\n",
            "       [-0.7001894 , -0.87808545, -0.15866142, -0.45380634,  0.08816735,\n",
            "        -0.63520142, -0.55832839, -0.5394391 , -0.32538937, -0.29726295],\n",
            "       [-0.39847994, -0.79737428, -0.27774362, -0.25563146, -0.26107482,\n",
            "        -0.6052004 , -0.50857071, -0.54017032,  0.03170361, -0.2489708 ],\n",
            "       [-0.03586702, -0.88442415, -0.365153  , -0.56343761, -0.26358342,\n",
            "        -0.51373571, -0.58593011, -0.59022521, -0.19944162, -0.23566953],\n",
            "       [-0.49208337, -0.86377768, -0.30767024, -0.25882617,  0.02730462,\n",
            "        -0.6140201 , -0.50881664, -0.53949905,  0.01822509, -0.18165808],\n",
            "       [-0.47745585, -0.85900934, -0.28099707, -0.39686624,  0.02734219,\n",
            "        -0.43411171, -0.62244084, -0.49515306, -0.17239473, -0.2132924 ],\n",
            "       [-0.53092475, -0.8838942 , -0.3265838 , -0.2313063 , -0.22148624,\n",
            "        -0.51576918, -0.59260465, -0.52343009, -0.00221356, -0.19766434],\n",
            "       [-0.37334284, -0.84974909, -0.2255407 , -0.50543541,  0.17763872,\n",
            "        -0.47321183, -0.53445893, -0.59117409, -0.08660936, -0.30698779],\n",
            "       [-0.24676739, -0.82260022,  0.05356339, -0.41300019,  0.07532986,\n",
            "        -0.73882001, -0.76367123, -0.48452755, -0.05733498, -0.32252224],\n",
            "       [-0.40221276, -0.84825495, -0.18242841, -0.42690568,  0.01418368,\n",
            "        -0.51948542, -0.69114006, -0.45445654, -0.04690817, -0.29584791],\n",
            "       [-0.43173231, -0.83694304, -0.27573118, -0.57294631,  0.00525774,\n",
            "        -0.39793744, -0.72630731, -0.47706131, -0.23668843, -0.29031364],\n",
            "       [-0.22697708, -0.85126197, -0.20971474, -0.50223577, -0.05325396,\n",
            "        -0.50108647, -0.65492586, -0.53162679, -0.18691795, -0.26066086],\n",
            "       [-0.60928522, -0.8580276 , -0.16862227, -0.41146697, -0.0255846 ,\n",
            "        -0.45694745, -0.5947067 , -0.54008315, -0.15928877, -0.24443515],\n",
            "       [-0.56947561, -0.85729597, -0.19133914, -0.60056683, -0.09327945,\n",
            "        -0.52832834, -0.54803947, -0.48748591, -0.18030336, -0.17847209],\n",
            "       [-0.40963161, -0.88772715, -0.21321409, -0.39434664, -0.11919953,\n",
            "        -0.41897646, -0.57512577, -0.54032712, -0.15417611, -0.20335813],\n",
            "       [-0.55552562, -0.87276166, -0.24357067, -0.50035786, -0.11949711,\n",
            "        -0.50764516, -0.59789517, -0.51533414, -0.278765  , -0.23574897],\n",
            "       [-0.19829793, -0.91432935, -0.33939172, -0.62743894, -0.00587514,\n",
            "        -0.57155223, -0.42948384, -0.66120678, -0.02715559, -0.17854684],\n",
            "       [-0.54602718, -0.87170967, -0.18791082, -0.37315795, -0.14433349,\n",
            "        -0.52545558, -0.59987174, -0.58148847, -0.20126407, -0.20825916],\n",
            "       [-0.40508053, -0.86762527, -0.2424183 , -0.44432537, -0.10038984,\n",
            "        -0.51801049, -0.61709466, -0.51484783, -0.12905566, -0.22228608],\n",
            "       [-0.50805143, -0.86995061, -0.20869266, -0.39383139,  0.12070296,\n",
            "        -0.50717744, -0.53253097, -0.55481099, -0.03307873, -0.2416837 ],\n",
            "       [-0.22319641, -0.85488319, -0.09445843, -0.46890236, -0.18832127,\n",
            "        -0.6241784 , -0.59108438, -0.58997824, -0.12490055, -0.21582779],\n",
            "       [-0.4532813 , -0.78904158, -0.26769724, -0.29422889, -0.18194371,\n",
            "        -0.58630524, -0.53912796, -0.54099293, -0.08402014, -0.16671046],\n",
            "       [-0.66988254, -0.88967958, -0.25650799, -0.33048571, -0.35342715,\n",
            "        -0.47622602, -0.54344972, -0.56089606, -0.02640692, -0.25976649],\n",
            "       [-0.16175727, -0.90405681, -0.37032063, -0.38291718, -0.12188647,\n",
            "        -0.46902436, -0.57707804, -0.53722772, -0.13630207, -0.16613654],\n",
            "       [-0.32490349, -0.80983889, -0.24105324, -0.43623351, -0.22256211,\n",
            "        -0.44596496, -0.57866291, -0.55878155, -0.08925396, -0.28950933],\n",
            "       [-0.61916877, -0.84104359, -0.21734221, -0.44258387, -0.02067454,\n",
            "        -0.61411126, -0.61720633, -0.47593677, -0.163655  , -0.20673859],\n",
            "       [-0.66232483, -0.8641583 , -0.15090564, -0.50511748,  0.02268228,\n",
            "        -0.5421796 , -0.56041083, -0.55981148, -0.25302812, -0.26638586],\n",
            "       [-0.50714665, -0.86611348, -0.20844057, -0.39488476,  0.13050539,\n",
            "        -0.51555988, -0.52878711, -0.55984249, -0.03739374, -0.24334006],\n",
            "       [-0.45203554, -0.88011729, -0.19531126, -0.51774226,  0.02531522,\n",
            "        -0.47186367, -0.68663197, -0.47421767, -0.18381072, -0.26366721],\n",
            "       [-0.7213671 , -0.87093337, -0.15179419, -0.40324263, -0.09741651,\n",
            "        -0.51024432, -0.55622403, -0.56252585, -0.24104632, -0.31428086],\n",
            "       [-0.26460654, -0.92763386,  0.14973679, -0.42376429,  0.09334932,\n",
            "        -0.58108978, -0.65361198, -0.62795317, -0.07526124, -0.27281549],\n",
            "       [-0.49582403, -0.88546196, -0.21302156, -0.41216814,  0.08695295,\n",
            "        -0.42999454, -0.54043991, -0.57641683, -0.03414104, -0.2464811 ],\n",
            "       [-0.60804167, -0.90053776, -0.27406754, -0.58955234,  0.18416716,\n",
            "        -0.50725655, -0.70550182, -0.37292021, -0.2792186 , -0.15388232],\n",
            "       [-0.52810849, -0.86235218, -0.17633652, -0.54493431, -0.07235083,\n",
            "        -0.44386378, -0.72248445, -0.48551201, -0.20700314, -0.2446273 ],\n",
            "       [-0.39470334, -0.87477786, -0.28401141, -0.49735153, -0.00929948,\n",
            "        -0.31352023, -0.55372146, -0.63016176, -0.0309801 , -0.29155169],\n",
            "       [ 0.68725628, -0.97837167, -0.41390307, -0.91452353, -0.97801136,\n",
            "        -0.69958239, -0.66970878, -0.56102805, -0.52491893, -0.08109574],\n",
            "       [-0.57700397, -0.86101845, -0.2052088 , -0.53067833,  0.08602116,\n",
            "        -0.46937625, -0.57154859, -0.53305118, -0.1123737 , -0.22989988],\n",
            "       [-0.67096702, -0.8481323 , -0.21142468, -0.60961751, -0.19829042,\n",
            "        -0.35898287, -0.65602887, -0.4895422 , -0.08728556, -0.21727311],\n",
            "       [-0.69136212, -0.7771576 , -0.21780857, -0.39105701, -0.09273625,\n",
            "        -0.64265721, -0.50601115, -0.51966195, -0.09750339, -0.20555144],\n",
            "       [-0.11173457, -0.94172848, -0.68828412,  0.80406721,  0.39970432,\n",
            "        -0.27004805, -0.75857103, -0.4657158 , -0.4834182 , -0.38272008],\n",
            "       [-0.41669031, -0.90017611, -0.24561477, -0.44367906, -0.10230407,\n",
            "        -0.50762013, -0.61917302, -0.52177208, -0.14244343, -0.23983841],\n",
            "       [-0.51862575, -0.88829447, -0.13785105, -0.41743994, -0.05508137,\n",
            "        -0.49558947, -0.62378826, -0.52077211, -0.1252158 , -0.20603696],\n",
            "       [-0.39570229, -0.83110182, -0.02676536, -0.36596534, -0.13843307,\n",
            "        -0.48775794, -0.64719145, -0.53040573, -0.13053553, -0.27463158],\n",
            "       [-0.32785885, -0.86467783, -0.35216643, -0.50684136, -0.03685446,\n",
            "        -0.40896769, -0.66758076, -0.43485226,  0.02447085, -0.20727454]]), 0: array([[-5.54265741e-01, -8.54673345e-01, -2.27631281e-01,\n",
            "        -4.50708461e-01, -1.89498269e-01, -2.74616103e-01,\n",
            "        -6.74830839e-01, -5.05911546e-01, -1.91460165e-01,\n",
            "        -2.63270804e-01],\n",
            "       [-5.96791883e-01, -8.78953424e-01, -2.06984954e-01,\n",
            "        -4.88593117e-01, -8.07564097e-02, -4.51003984e-01,\n",
            "        -5.92305849e-01, -5.60168085e-01, -1.74358266e-01,\n",
            "        -2.40733287e-01],\n",
            "       [-5.10573541e-01, -8.29049605e-01, -1.75942901e-01,\n",
            "        -4.07140322e-01,  8.37389113e-04, -5.82524332e-01,\n",
            "        -6.19801034e-01, -4.96385582e-01, -9.97121299e-02,\n",
            "        -2.20347281e-01],\n",
            "       [-2.51610589e-01, -8.87572176e-01, -1.56627156e-01,\n",
            "        -4.19468501e-01, -6.04160972e-02, -4.52718385e-01,\n",
            "        -5.81508916e-01, -5.32908857e-01, -1.16014591e-01,\n",
            "        -2.72372727e-01],\n",
            "       [-6.74708006e-01, -8.58821065e-01, -2.46121519e-01,\n",
            "        -3.54443688e-01,  1.57169349e-01, -6.55353543e-01,\n",
            "        -5.75717254e-01, -4.87380348e-01, -1.01470834e-01,\n",
            "        -1.66666578e-01],\n",
            "       [-5.77937268e-01, -8.36062404e-01, -2.06091499e-01,\n",
            "        -4.03672223e-01, -4.14319911e-03, -6.42211638e-01,\n",
            "        -5.31350199e-01, -5.48212600e-01, -2.89629654e-01,\n",
            "        -2.74515153e-01],\n",
            "       [-6.51133229e-01, -8.51353298e-01, -1.63992291e-01,\n",
            "        -5.44136840e-01, -2.38129019e-01, -4.83902681e-01,\n",
            "        -6.93441542e-01, -5.33539025e-01, -3.28682013e-01,\n",
            "        -2.25558655e-01],\n",
            "       [-4.20168699e-01, -8.94924783e-01, -3.35100120e-01,\n",
            "        -4.72323215e-01, -1.54733371e-01, -7.66851786e-01,\n",
            "        -4.99263216e-01, -5.43637182e-01, -8.79104291e-02,\n",
            "        -1.87033212e-01],\n",
            "       [-4.50273672e-01, -8.52704255e-01, -2.75202636e-01,\n",
            "        -4.46944442e-01,  1.43590822e-01, -5.43568524e-01,\n",
            "        -6.52543206e-01, -4.51678779e-01, -1.30856784e-01,\n",
            "        -1.85550910e-01],\n",
            "       [-1.60491535e-01, -8.42818069e-01, -3.05188004e-01,\n",
            "        -5.79928261e-01, -2.09692131e-01, -4.96968222e-01,\n",
            "        -6.16380108e-01, -5.48251476e-01, -1.05080808e-01,\n",
            "        -2.49901473e-01],\n",
            "       [-2.76296921e-01, -8.64904379e-01, -1.97585258e-01,\n",
            "        -4.39280547e-01, -8.95699183e-02, -6.05653769e-01,\n",
            "        -6.22895809e-01, -5.20613228e-01, -1.68560311e-01,\n",
            "        -2.50378349e-01],\n",
            "       [-2.73557143e-01, -8.81437995e-01,  4.67858572e-02,\n",
            "        -4.09045097e-01,  5.92889090e-02, -7.42585891e-01,\n",
            "        -7.69931971e-01, -4.69481762e-01, -5.46414319e-02,\n",
            "        -2.86143825e-01],\n",
            "       [-5.44015166e-01, -8.88466149e-01, -2.84087537e-01,\n",
            "        -3.99663186e-01,  1.97602894e-01, -6.50842268e-01,\n",
            "        -5.46973470e-01, -4.75665829e-01, -5.75850030e-03,\n",
            "        -1.49651105e-01],\n",
            "       [-5.37913060e-01, -8.53792933e-01, -2.34561807e-01,\n",
            "        -3.65436361e-01,  1.49368211e-01, -4.20336597e-01,\n",
            "        -5.35742741e-01, -5.57484121e-01, -5.14011350e-02,\n",
            "        -2.09170910e-01],\n",
            "       [-5.05541888e-01, -9.04182690e-01, -1.94650919e-01,\n",
            "        -4.03928990e-01, -1.80573794e-01, -4.19286103e-01,\n",
            "        -6.26988807e-01, -5.21549771e-01, -1.30945000e-01,\n",
            "        -2.21704956e-01],\n",
            "       [-6.73462061e-01, -9.02273438e-01, -2.44973708e-01,\n",
            "        -3.95451437e-01, -2.18209817e-01, -5.16415049e-01,\n",
            "        -5.78685163e-01, -5.51222793e-01, -7.81103154e-02,\n",
            "        -2.05230395e-01],\n",
            "       [-6.75814136e-01, -8.75412087e-01, -1.66106043e-01,\n",
            "        -3.25932041e-01, -8.28484533e-02, -6.32039249e-01,\n",
            "        -5.68988981e-01, -5.43225515e-01, -1.81071320e-01,\n",
            "        -2.59643033e-01],\n",
            "       [-4.05014341e-01, -8.82363044e-01, -2.36533423e-01,\n",
            "        -3.10119459e-01, -2.61034822e-01, -4.65057185e-01,\n",
            "        -6.48275305e-01, -5.19407497e-01, -5.54096176e-02,\n",
            "        -2.46348032e-01],\n",
            "       [-5.64346922e-01, -8.70111021e-01, -2.02815969e-01,\n",
            "        -5.02208126e-01, -7.83553591e-02, -5.10328243e-01,\n",
            "        -5.46861563e-01, -5.75138422e-01, -1.82880833e-01,\n",
            "        -2.25794616e-01],\n",
            "       [-5.26240055e-01, -8.84093411e-01, -1.51539398e-01,\n",
            "        -5.73999565e-01, -4.78748613e-02, -4.86152637e-01,\n",
            "        -6.85474516e-01, -4.58410769e-01, -2.21883023e-01,\n",
            "        -2.07048625e-01],\n",
            "       [-5.11549106e-01, -8.94215318e-01, -2.87275757e-01,\n",
            "        -4.87731748e-01, -1.40267239e-01, -5.33406358e-01,\n",
            "        -7.35437368e-01, -4.62690995e-01, -2.82790359e-01,\n",
            "        -1.92159590e-01],\n",
            "       [-5.03698859e-01, -9.08791260e-01, -1.52347858e-01,\n",
            "        -4.66572807e-01, -7.39086976e-02, -5.00506498e-01,\n",
            "        -6.01742101e-01, -5.25704077e-01, -1.45222740e-01,\n",
            "        -2.30331182e-01],\n",
            "       [-5.27973177e-01, -7.99093605e-01, -3.15874107e-01,\n",
            "        -4.23867001e-01, -2.20362528e-01, -2.06523058e-01,\n",
            "        -7.69613049e-01, -4.78277460e-01, -1.08470978e-01,\n",
            "        -3.13451253e-01],\n",
            "       [ 1.08323182e-01, -7.73476191e-01, -3.75151845e-01,\n",
            "        -7.64653806e-01, -5.10793190e-01, -6.09389054e-01,\n",
            "        -6.53954242e-01, -5.01514472e-01, -3.94904215e-01,\n",
            "        -4.10319500e-02],\n",
            "       [-4.05492712e-01, -9.01042149e-01, -2.36005905e-01,\n",
            "        -4.15606536e-01, -1.68134302e-01, -5.50437672e-01,\n",
            "        -6.12327023e-01, -5.29249198e-01, -1.41119825e-01,\n",
            "        -2.64276170e-01],\n",
            "       [-4.79245131e-01, -6.87675006e-01, -1.12733082e-01,\n",
            "        -3.95870865e-01, -1.12305605e-02, -3.75644115e-01,\n",
            "        -5.31346010e-01, -6.39961549e-01, -2.34420024e-01,\n",
            "        -3.24189710e-01],\n",
            "       [-4.08681484e-01, -8.78622138e-01, -1.76927892e-01,\n",
            "        -8.27668925e-01, -1.99598018e-02, -1.26913770e-01,\n",
            "        -8.88451757e-01, -3.41358982e-01, -1.85020525e-01,\n",
            "        -2.80450445e-01],\n",
            "       [-3.95823723e-01, -8.75981669e-01, -2.83703451e-01,\n",
            "        -4.95636417e-01, -1.98102661e-02, -3.14316056e-01,\n",
            "        -5.58634603e-01, -6.23492352e-01, -2.72920998e-02,\n",
            "        -2.84936642e-01],\n",
            "       [-6.24746601e-01, -8.65606853e-01, -2.71801581e-01,\n",
            "        -4.54690397e-01, -1.54560004e-02, -4.61268982e-01,\n",
            "        -6.60733729e-01, -4.57053611e-01, -2.85452515e-01,\n",
            "        -2.74847711e-01],\n",
            "       [-5.63876122e-01, -9.17648876e-01, -2.20440712e-01,\n",
            "        -4.66104740e-01, -2.64912693e-01, -4.85546924e-01,\n",
            "        -6.60816294e-01, -5.15918542e-01, -2.01911605e-01,\n",
            "        -2.25955097e-01],\n",
            "       [-2.69753362e-01, -8.86512626e-01, -6.22819260e-01,\n",
            "         9.25032334e-01,  5.32002931e-01, -4.61070245e-01,\n",
            "        -6.18791266e-01, -4.96923472e-01, -2.27275112e-01,\n",
            "        -3.17127127e-01],\n",
            "       [-4.85375327e-01, -9.09234687e-01,  3.48292056e-02,\n",
            "        -3.82078851e-01, -7.74956339e-02, -5.49550302e-01,\n",
            "        -7.85440988e-01, -4.99712548e-01, -1.49719672e-01,\n",
            "        -3.46408190e-01],\n",
            "       [-5.18131463e-01, -9.18560480e-01, -1.75755542e-01,\n",
            "        -3.63009554e-01,  7.57618188e-03, -4.81026773e-01,\n",
            "        -6.56406928e-01, -5.11611325e-01, -1.82969836e-01,\n",
            "        -2.06222603e-01],\n",
            "       [-4.20750322e-01, -8.85241599e-01, -2.40368475e-01,\n",
            "        -3.63813629e-01, -1.95698399e-01, -5.13545126e-01,\n",
            "        -6.04228172e-01, -5.44533955e-01, -9.50269841e-02,\n",
            "        -1.82264368e-01],\n",
            "       [-2.21401798e-01, -8.62385564e-01,  1.06673193e-01,\n",
            "        -3.54233483e-01, -2.05137446e-01, -6.15248408e-01,\n",
            "        -6.32967611e-01, -4.83923105e-01, -3.23469700e-02,\n",
            "        -2.52712728e-01],\n",
            "       [-4.12484629e-01, -7.38386018e-01, -1.81128330e-01,\n",
            "        -5.43087394e-01,  1.64127399e-01, -3.80481713e-01,\n",
            "        -5.35703457e-01, -6.05678665e-01, -1.15299907e-01,\n",
            "        -3.19576601e-01],\n",
            "       [-6.69606907e-01, -8.63074077e-01, -1.45162752e-01,\n",
            "        -3.93723023e-01,  8.15873420e-02, -6.35154781e-01,\n",
            "        -5.88567311e-01, -5.46601846e-01, -1.91250464e-01,\n",
            "        -2.36389149e-01],\n",
            "       [-2.09243260e-03, -8.91094997e-01, -8.92817189e-02,\n",
            "        -4.18542562e-01,  3.17087198e-02, -5.53306323e-01,\n",
            "        -6.20140475e-01, -5.65368950e-01, -9.55620433e-02,\n",
            "        -2.58345498e-01],\n",
            "       [-4.41957721e-01, -9.21627462e-01, -1.92219854e-01,\n",
            "        -4.13235881e-01, -1.82422605e-01, -5.40276452e-01,\n",
            "        -6.24813397e-01, -5.00124592e-01, -7.69035993e-02,\n",
            "        -2.26721276e-01],\n",
            "       [-5.89037343e-01, -8.78472727e-01, -1.33992001e-01,\n",
            "        -4.92147487e-01, -2.30889405e-01, -4.17734660e-01,\n",
            "        -6.44543104e-01, -5.28854030e-01, -2.31860804e-01,\n",
            "        -2.56073993e-01],\n",
            "       [-4.97745292e-01, -7.95626047e-01, -2.18475269e-01,\n",
            "        -5.32916167e-01, -1.49599668e-02, -5.29861544e-01,\n",
            "        -7.01541003e-01, -4.70403938e-01, -2.22494689e-01,\n",
            "        -2.55479199e-01],\n",
            "       [-6.07195072e-01, -8.81367522e-01, -2.33437573e-01,\n",
            "        -4.36032795e-01, -3.73474444e-03, -4.41916483e-01,\n",
            "        -5.52928328e-01, -5.43339715e-01,  5.64737414e-02,\n",
            "        -2.10199853e-01],\n",
            "       [-6.03509350e-01, -8.80827414e-01, -1.98375306e-01,\n",
            "        -4.77022412e-01, -3.49740520e-02, -3.97102635e-01,\n",
            "        -5.61189045e-01, -5.65143090e-01, -1.39958795e-01,\n",
            "        -2.59523431e-01],\n",
            "       [-6.94298243e-01, -8.84411030e-01, -2.48598127e-01,\n",
            "        -4.75812909e-01, -3.73772183e-01, -1.64802667e-01,\n",
            "        -6.92258429e-01, -4.91316093e-01,  1.64065137e-01,\n",
            "        -2.77730089e-01],\n",
            "       [-6.09751818e-01, -8.98106789e-01, -1.98847841e-01,\n",
            "        -3.95092968e-01, -2.18435746e-01, -4.98654611e-01,\n",
            "        -6.03337633e-01, -4.89015444e-01, -8.19673585e-02,\n",
            "        -2.14238624e-01],\n",
            "       [-5.77285664e-01, -8.47995593e-01, -2.77144947e-01,\n",
            "        -5.71030969e-01, -5.64947056e-02, -3.20782853e-01,\n",
            "        -8.47549283e-01, -4.15977189e-01, -2.44783824e-01,\n",
            "        -2.24316252e-01],\n",
            "       [-4.86241074e-01, -8.97854877e-01, -2.79544617e-01,\n",
            "        -3.45219955e-01, -1.96029439e-01, -4.79396479e-01,\n",
            "        -6.19325862e-01, -5.10766681e-01, -4.01599328e-02,\n",
            "        -2.38106655e-01],\n",
            "       [-3.86965708e-01, -8.51259522e-01, -3.46171807e-01,\n",
            "        -3.49208876e-01, -2.82849425e-01, -4.75491843e-01,\n",
            "        -5.60785766e-01, -4.97718716e-01,  9.83540982e-02,\n",
            "        -1.34937785e-01],\n",
            "       [-5.49247478e-01, -8.24367609e-01, -2.13121850e-01,\n",
            "        -3.26187420e-01, -1.42151868e-01, -5.11374248e-01,\n",
            "        -6.32285002e-01, -5.47671194e-01, -1.34709324e-01,\n",
            "        -2.60123850e-01],\n",
            "       [-5.40552025e-01, -8.88753763e-01, -2.66259718e-01,\n",
            "        -4.15395840e-01, -3.44261477e-01, -4.85473346e-01,\n",
            "        -4.96565687e-01, -5.61767810e-01, -6.32099948e-02,\n",
            "        -2.50518520e-01],\n",
            "       [-4.38266921e-01, -8.89338665e-01, -2.84981702e-01,\n",
            "        -3.61596267e-01, -2.48925954e-01, -5.45904530e-01,\n",
            "        -5.80943488e-01, -4.89509053e-01,  5.33976002e-02,\n",
            "        -2.17131232e-01],\n",
            "       [-1.93843344e-01, -8.95439971e-01, -7.03931307e-02,\n",
            "        -3.83722553e-01, -1.01832908e-01, -5.77268271e-01,\n",
            "        -6.92893841e-01, -5.37881748e-01, -1.33419464e-01,\n",
            "        -2.47255987e-01],\n",
            "       [-4.95206482e-01, -8.55874202e-01, -2.16550597e-01,\n",
            "        -5.41245136e-01,  1.26640543e-01, -2.95066254e-01,\n",
            "        -5.34636097e-01, -5.94566491e-01, -4.48172762e-02,\n",
            "        -2.54220788e-01],\n",
            "       [-4.64262538e-01, -9.05198044e-01, -4.37498576e-02,\n",
            "        -4.49337266e-01, -1.53168360e-02, -6.03243313e-01,\n",
            "        -5.54248132e-01, -5.58655197e-01, -5.63850094e-02,\n",
            "        -2.05223279e-01],\n",
            "       [-5.00413778e-01, -8.27024278e-01, -2.23147175e-01,\n",
            "        -5.07624710e-01,  2.08254578e-03, -3.77888972e-01,\n",
            "        -6.20427624e-01, -5.46654890e-01, -1.28502649e-01,\n",
            "        -2.74820501e-01],\n",
            "       [-4.52548417e-01, -8.56846174e-01, -3.11328826e-01,\n",
            "        -3.04706562e-01, -3.37725740e-01, -5.52439347e-01,\n",
            "        -5.98025347e-01, -5.17227309e-01, -1.68925751e-02,\n",
            "        -2.24461238e-01],\n",
            "       [-2.65477733e-01, -8.43303124e-01, -1.94110424e-01,\n",
            "        -4.48074703e-01, -8.16885429e-02, -5.67700365e-01,\n",
            "        -6.18492159e-01, -5.38252918e-01, -1.52052231e-01,\n",
            "        -3.14908683e-01],\n",
            "       [-5.00344700e-01, -8.86947169e-01, -1.79566450e-01,\n",
            "        -3.70653485e-01,  4.20294085e-02, -3.46549892e-01,\n",
            "        -5.95180945e-01, -5.40008553e-01, -7.88652348e-03,\n",
            "        -2.44430759e-01],\n",
            "       [-6.16768005e-01, -9.04226312e-01, -2.38447841e-01,\n",
            "        -4.75935215e-01,  1.21930934e-01, -4.56700925e-01,\n",
            "        -5.47246938e-01, -5.35918094e-01, -1.05766115e-01,\n",
            "        -2.07432525e-01],\n",
            "       [-5.79475234e-01, -8.74234451e-01, -1.96971545e-01,\n",
            "        -4.20192865e-01, -1.74767250e-01, -5.57884889e-01,\n",
            "        -5.44058827e-01, -5.98894859e-01, -1.96094353e-01,\n",
            "        -2.97375504e-01],\n",
            "       [-4.36553663e-01, -8.59727280e-01, -1.81550398e-01,\n",
            "        -5.70856702e-01, -5.34538410e-03, -5.21179151e-01,\n",
            "        -5.10250055e-01, -5.80992107e-01, -1.17826078e-01,\n",
            "        -2.29591318e-01],\n",
            "       [-5.54322847e-01, -8.62607174e-01, -1.65078886e-01,\n",
            "        -5.58572408e-01,  2.35981590e-02, -3.89172592e-01,\n",
            "        -5.78306488e-01, -5.23619000e-01, -1.51270351e-01,\n",
            "        -1.74239320e-01],\n",
            "       [-5.74911341e-01, -8.08683909e-01, -1.70290867e-01,\n",
            "        -4.11488195e-01, -2.80373130e-01, -5.10506531e-01,\n",
            "        -6.07901472e-01, -5.95517655e-01, -1.81918089e-01,\n",
            "        -3.11832611e-01],\n",
            "       [-5.63343674e-01, -8.76659495e-01, -2.16172400e-01,\n",
            "        -5.53501235e-01,  1.16369846e-01, -3.43376115e-01,\n",
            "        -5.67911747e-01, -5.76929447e-01, -1.06787136e-01,\n",
            "        -2.51521437e-01],\n",
            "       [-2.86155612e-01, -8.05378252e-01, -2.90986683e-01,\n",
            "        -3.61941872e-01,  2.43639119e-01, -6.85985558e-01,\n",
            "        -5.93646616e-01, -3.96399699e-01, -7.35914560e-02,\n",
            "        -2.08468099e-01],\n",
            "       [-4.13499584e-01, -8.93339383e-01, -2.57194819e-01,\n",
            "        -5.75495886e-01,  4.39658940e-02, -5.62937404e-01,\n",
            "        -7.46765012e-01, -4.29094908e-01, -8.44013689e-02,\n",
            "        -2.21504340e-01],\n",
            "       [-5.21611462e-01, -8.62278002e-01, -3.70663060e-02,\n",
            "        -3.00146990e-01, -1.78646240e-01, -4.95083587e-01,\n",
            "        -6.25252455e-01, -5.76834447e-01, -1.67003919e-01,\n",
            "        -2.46925412e-01],\n",
            "       [-3.86104948e-01, -8.21434178e-01, -2.02635879e-01,\n",
            "        -4.71588679e-01,  1.92451056e-01, -4.36929369e-01,\n",
            "        -6.19704662e-01, -4.96375499e-01, -1.39060149e-01,\n",
            "        -2.15199442e-01],\n",
            "       [-5.54540340e-01, -8.53008354e-01, -8.42492000e-02,\n",
            "        -4.30645917e-01, -1.42987535e-01, -4.59800415e-01,\n",
            "        -5.08619273e-01, -6.14090576e-01, -9.88790431e-02,\n",
            "        -2.70803649e-01],\n",
            "       [-3.68112489e-01, -8.36101537e-01, -3.88438401e-01,\n",
            "        -3.64282779e-01, -4.88250706e-02, -2.95042293e-01,\n",
            "        -6.77808427e-01, -4.97624788e-01,  5.56746629e-02,\n",
            "        -2.48170583e-01],\n",
            "       [-1.98297932e-01, -9.14329353e-01, -3.39391718e-01,\n",
            "        -6.27438943e-01, -5.87513989e-03, -5.71552226e-01,\n",
            "        -4.29483844e-01, -6.61206784e-01, -2.71555879e-02,\n",
            "        -1.78546839e-01],\n",
            "       [-5.23633906e-01, -8.32591358e-01, -2.06327920e-01,\n",
            "        -5.05896361e-01, -8.42474641e-02, -4.01117271e-01,\n",
            "        -6.62735977e-01, -4.94586971e-01, -1.77859211e-01,\n",
            "        -2.03786984e-01],\n",
            "       [-5.98246171e-01, -7.97000930e-01, -1.96088217e-01,\n",
            "        -4.38576492e-01,  7.25566207e-02, -5.45480981e-01,\n",
            "        -6.07576652e-01, -4.90218105e-01, -9.59565692e-02,\n",
            "        -2.28633183e-01],\n",
            "       [-5.91533585e-01, -7.54541970e-01, -2.32458210e-01,\n",
            "        -3.25896987e-01, -1.80279461e-01, -5.56390221e-01,\n",
            "        -5.71028226e-01, -5.26227324e-01,  8.70839594e-02,\n",
            "        -2.54248447e-01],\n",
            "       [-4.65118571e-01, -8.51990938e-01, -2.42436127e-01,\n",
            "        -5.05034356e-01,  4.88272725e-02, -2.76854450e-01,\n",
            "        -5.48999372e-01, -6.31290217e-01, -1.66595152e-01,\n",
            "        -3.19942857e-01],\n",
            "       [-5.11690015e-01, -8.80592229e-01, -2.33752788e-01,\n",
            "        -3.48328998e-01, -1.05796664e-01, -5.96030084e-01,\n",
            "        -5.90838431e-01, -5.14059409e-01, -1.23772386e-01,\n",
            "        -1.85387508e-01],\n",
            "       [-4.69076574e-01, -8.74570908e-01, -2.85981219e-01,\n",
            "        -4.33681423e-01, -1.53412826e-01, -4.35236945e-01,\n",
            "        -6.75412621e-01, -5.08471780e-01, -1.55422355e-01,\n",
            "        -2.06632150e-01],\n",
            "       [-5.39481816e-01, -8.54313413e-01, -1.92654730e-01,\n",
            "        -6.24837486e-01,  1.20409781e-01, -2.43811103e-01,\n",
            "        -5.98374503e-01, -5.61864867e-01, -7.84912968e-02,\n",
            "        -2.54245888e-01],\n",
            "       [-5.84886555e-01, -8.86769490e-01, -1.90773805e-01,\n",
            "        -4.53298417e-01,  4.97395140e-02, -5.87172965e-01,\n",
            "        -6.80331145e-01, -4.85742253e-01, -2.70584528e-01,\n",
            "        -2.09129110e-01],\n",
            "       [-3.26569732e-01, -8.72010794e-01,  8.88237687e-02,\n",
            "        -3.78289628e-01, -1.48897237e-01, -4.85058632e-01,\n",
            "        -7.10566392e-01, -5.17816441e-01, -4.82510981e-02,\n",
            "        -2.63265482e-01],\n",
            "       [-1.64622035e-01, -8.99642577e-01, -3.38664975e-01,\n",
            "        -4.76169378e-01,  1.43820346e-01, -4.17705942e-01,\n",
            "        -5.40335409e-01, -4.85649012e-01, -1.20626917e-01,\n",
            "        -1.47957672e-01],\n",
            "       [-3.99811671e-01, -8.83202502e-01, -2.55690708e-01,\n",
            "        -4.52954451e-01, -6.78321643e-02, -5.44689611e-01,\n",
            "        -6.04168296e-01, -5.44772248e-01, -2.06969463e-01,\n",
            "        -2.52651575e-01],\n",
            "       [-5.61132171e-01, -8.76159704e-01, -2.40250323e-01,\n",
            "        -4.71946651e-01, -1.63681247e-01, -4.70680901e-01,\n",
            "        -6.28648320e-01, -4.98074650e-01, -1.44649546e-01,\n",
            "        -2.06641532e-01],\n",
            "       [-5.41672465e-01, -8.93025228e-01, -2.87136931e-01,\n",
            "        -5.25717904e-01, -4.32181328e-01, -4.30563235e-01,\n",
            "        -5.97246148e-01, -5.17286675e-01, -7.88924616e-02,\n",
            "        -2.41646764e-01],\n",
            "       [-4.07288317e-01, -8.95832841e-01, -2.57836514e-01,\n",
            "        -4.51952509e-01, -6.54369644e-02, -5.56516901e-01,\n",
            "        -6.02731439e-01, -5.46098577e-01, -2.20531657e-01,\n",
            "        -2.50609138e-01],\n",
            "       [-5.67353292e-01, -7.86565852e-01, -1.36046147e-01,\n",
            "        -4.59113745e-01,  1.82513206e-02, -4.86866147e-01,\n",
            "        -6.54574627e-01, -4.81944504e-01, -6.60484189e-02,\n",
            "        -1.65018760e-01],\n",
            "       [-3.00168936e-01, -9.03358328e-01, -3.85414290e-01,\n",
            "        -4.50020387e-01, -4.30685295e-01, -5.65411569e-01,\n",
            "        -6.05515278e-01, -4.92133239e-01, -1.03766299e-01,\n",
            "        -2.08076119e-01],\n",
            "       [-6.27160585e-01, -8.27649141e-01, -2.03987393e-01,\n",
            "        -4.37727936e-01, -2.88680047e-01, -4.21957701e-01,\n",
            "        -5.85737383e-01, -5.31224138e-01, -8.79958237e-02,\n",
            "        -2.90532585e-01],\n",
            "       [-6.08829497e-01, -8.20113019e-01, -2.28518708e-01,\n",
            "        -5.45127723e-01,  1.71303238e-02, -4.13269978e-01,\n",
            "        -6.27579114e-01, -5.06326726e-01, -1.51977402e-01,\n",
            "        -2.42632973e-01],\n",
            "       [-5.85571589e-01, -8.75267709e-01, -1.74412170e-01,\n",
            "        -4.54432075e-01, -3.60196627e-03, -2.15071119e-01,\n",
            "        -6.60902625e-01, -4.90799679e-01, -1.56448838e-01,\n",
            "        -2.44503764e-01],\n",
            "       [-4.64689927e-01, -8.80499385e-01, -2.09481638e-01,\n",
            "        -4.41207876e-01, -8.18870335e-02, -5.28555447e-01,\n",
            "        -6.79657435e-01, -4.98949930e-01, -1.07288458e-01,\n",
            "        -2.57753236e-01],\n",
            "       [-6.81546522e-01, -8.86498872e-01, -1.87510454e-01,\n",
            "        -3.83427192e-01, -6.36613055e-02, -4.13880591e-01,\n",
            "        -6.24717993e-01, -5.16608981e-01, -1.10225030e-01,\n",
            "        -2.36103366e-01],\n",
            "       [-4.78285803e-01, -8.93866511e-01, -3.61567144e-01,\n",
            "        -7.49926575e-01,  2.21400196e-01, -2.29670889e-01,\n",
            "        -8.32102270e-01, -2.63860639e-01, -3.05906621e-01,\n",
            "        -3.57017814e-02],\n",
            "       [-5.30460871e-01, -8.49822579e-01, -2.58839012e-01,\n",
            "        -4.95960443e-01, -9.24566834e-02, -5.79348175e-01,\n",
            "        -5.95911354e-01, -5.18820374e-01, -1.66876431e-01,\n",
            "        -2.00454965e-01],\n",
            "       [-2.36213629e-01, -8.46441920e-01, -3.06732044e-01,\n",
            "        -3.53767722e-01, -4.02669443e-01, -5.00801493e-01,\n",
            "        -6.38479520e-01, -4.82716344e-01, -8.51819378e-02,\n",
            "        -2.30368263e-01],\n",
            "       [-4.70465374e-01, -9.14459017e-01, -2.52085658e-01,\n",
            "        -5.00506742e-01, -5.76403577e-02, -4.90210774e-01,\n",
            "        -5.53722463e-01, -5.06304939e-01, -2.74535174e-01,\n",
            "        -1.61747766e-01],\n",
            "       [-3.08158645e-01, -9.19542049e-01,  1.15992653e-01,\n",
            "        -3.98639138e-01,  8.96964708e-02, -5.44154938e-01,\n",
            "        -6.07516348e-01, -6.29840547e-01, -1.04129070e-01,\n",
            "        -1.93022345e-01],\n",
            "       [-4.17892384e-01, -8.68149413e-01, -2.72340780e-01,\n",
            "        -4.66304924e-01,  2.79488994e-02, -5.53257184e-01,\n",
            "        -6.91852588e-01, -4.34672260e-01, -1.46629327e-01,\n",
            "        -1.70626855e-01],\n",
            "       [-6.11174608e-01, -9.01889630e-01, -1.98443611e-01,\n",
            "        -3.97537992e-01, -2.08370757e-01, -4.99683000e-01,\n",
            "        -5.99249221e-01, -4.98478789e-01, -8.56539190e-02,\n",
            "        -2.31886397e-01],\n",
            "       [-5.29613637e-01, -8.51756982e-01, -1.68187232e-01,\n",
            "        -3.95164374e-01, -1.01235299e-01, -5.55623233e-01,\n",
            "        -5.68718772e-01, -5.48441130e-01, -1.99588520e-01,\n",
            "        -2.41850799e-01]])}\n",
            "{1: array([[-0.59736586, -0.84287605, -0.21604904, -0.37968079, -0.17532904,\n",
            "        -0.46774585, -0.61702309, -0.54803341, -0.11118806, -0.29614966],\n",
            "       [-0.55426574, -0.85467335, -0.22763128, -0.45070846, -0.18949827,\n",
            "        -0.2746161 , -0.67483084, -0.50591155, -0.19146016, -0.2632708 ],\n",
            "       [-0.59679188, -0.87895342, -0.20698495, -0.48859312, -0.08075641,\n",
            "        -0.45100398, -0.59230585, -0.56016808, -0.17435827, -0.24073329],\n",
            "       [-0.67470801, -0.85882106, -0.24612152, -0.35444369,  0.15716935,\n",
            "        -0.65535354, -0.57571725, -0.48738035, -0.10147083, -0.16666658],\n",
            "       [-0.47376175, -0.82561516, -0.22122838, -0.33452351, -0.07594053,\n",
            "        -0.62034981, -0.56983797, -0.57360479, -0.23558809, -0.1843496 ],\n",
            "       [-0.62464572, -0.8984194 , -0.17304913, -0.31931683,  0.16395037,\n",
            "        -0.58335579, -0.58959955, -0.47807808, -0.0263573 , -0.1615436 ],\n",
            "       [-0.49481028, -0.90782725, -0.07585673, -0.41960996, -0.1691485 ,\n",
            "        -0.34362659, -0.69683074, -0.54924067, -0.11282866, -0.24822583],\n",
            "       [-0.42998237, -0.86274909, -0.23280999, -0.38043282,  0.05228959,\n",
            "        -0.64664048, -0.65140637, -0.46828982, -0.11301707, -0.19730639],\n",
            "       [-0.57793727, -0.8360624 , -0.2060915 , -0.40367222, -0.0041432 ,\n",
            "        -0.64221164, -0.5313502 , -0.5482126 , -0.28962965, -0.27451515],\n",
            "       [-0.65113323, -0.8513533 , -0.16399229, -0.54413684, -0.23812902,\n",
            "        -0.48390268, -0.69344154, -0.53353903, -0.32868201, -0.22555866],\n",
            "       [-0.60650947, -0.85528365, -0.16797018, -0.41839681, -0.00606932,\n",
            "        -0.44123273, -0.58563857, -0.56023411, -0.1614541 , -0.29380775],\n",
            "       [-0.45027367, -0.85270426, -0.27520264, -0.44694444,  0.14359082,\n",
            "        -0.54356852, -0.65254321, -0.45167878, -0.13085678, -0.18555091],\n",
            "       [-0.16049153, -0.84281807, -0.305188  , -0.57992826, -0.20969213,\n",
            "        -0.49696822, -0.61638011, -0.54825148, -0.10508081, -0.24990147],\n",
            "       [-0.27629692, -0.86490438, -0.19758526, -0.43928055, -0.08956992,\n",
            "        -0.60565377, -0.62289581, -0.52061323, -0.16856031, -0.25037835],\n",
            "       [-0.27355714, -0.881438  ,  0.04678586, -0.4090451 ,  0.05928891,\n",
            "        -0.74258589, -0.76993197, -0.46948176, -0.05464143, -0.28614383],\n",
            "       [-0.44106289, -0.8300251 , -0.30131652, -0.42060193, -0.21284934,\n",
            "        -0.40062716, -0.61081118, -0.51751675, -0.09901989, -0.27266052],\n",
            "       [-0.69507489, -0.86903479, -0.17374833, -0.4322206 , -0.15534452,\n",
            "        -0.63379253, -0.59459878, -0.50751856, -0.13109363, -0.2090643 ],\n",
            "       [-0.40501434, -0.88236304, -0.23653342, -0.31011946, -0.26103482,\n",
            "        -0.46505718, -0.6482753 , -0.5194075 , -0.05540962, -0.24634803],\n",
            "       [-0.56434692, -0.87011102, -0.20281597, -0.50220813, -0.07835536,\n",
            "        -0.51032824, -0.54686156, -0.57513842, -0.18288083, -0.22579462],\n",
            "       [-0.52624005, -0.88409341, -0.1515394 , -0.57399956, -0.04787486,\n",
            "        -0.48615264, -0.68547452, -0.45841077, -0.22188302, -0.20704863],\n",
            "       [-0.53460531, -0.8603687 , -0.09771522, -0.40875318,  0.05457382,\n",
            "        -0.45665313, -0.56597458, -0.55088617, -0.17626257, -0.20085169],\n",
            "       [-0.51154911, -0.89421532, -0.28727576, -0.48773175, -0.14026724,\n",
            "        -0.53340636, -0.73543737, -0.46269099, -0.28279036, -0.19215959],\n",
            "       [-0.50369886, -0.90879126, -0.15234786, -0.46657281, -0.0739087 ,\n",
            "        -0.5005065 , -0.6017421 , -0.52570408, -0.14522274, -0.23033118],\n",
            "       [-0.52797318, -0.79909361, -0.31587411, -0.423867  , -0.22036253,\n",
            "        -0.20652306, -0.76961305, -0.47827746, -0.10847098, -0.31345125],\n",
            "       [-0.65371645, -0.86751256, -0.15995712, -0.45511655, -0.0121729 ,\n",
            "        -0.57365145, -0.53174257, -0.57721563, -0.26700915, -0.27743275],\n",
            "       [-0.34864708, -0.88432345, -0.36804002, -0.36623662, -0.09526137,\n",
            "        -0.54254104, -0.59616682, -0.5198837 ,  0.03543308, -0.21679148],\n",
            "       [ 0.10832318, -0.77347619, -0.37515184, -0.76465381, -0.51079319,\n",
            "        -0.60938905, -0.65395424, -0.50151447, -0.39490422, -0.04103195],\n",
            "       [-0.40549271, -0.90104215, -0.23600591, -0.41560654, -0.1681343 ,\n",
            "        -0.55043767, -0.61232702, -0.5292492 , -0.14111982, -0.26427617],\n",
            "       [-0.47924513, -0.68767501, -0.11273308, -0.39587086, -0.01123056,\n",
            "        -0.37564412, -0.53134601, -0.63996155, -0.23442002, -0.32418971],\n",
            "       [-0.53548606, -0.81853379, -0.18192701, -0.45097989, -0.18859529,\n",
            "        -0.48502046, -0.61037799, -0.55111365, -0.09255284, -0.2983631 ],\n",
            "       [-0.40868148, -0.87862214, -0.17692789, -0.82766893, -0.0199598 ,\n",
            "        -0.12691377, -0.88845176, -0.34135898, -0.18502052, -0.28045045],\n",
            "       [-0.39582372, -0.87598167, -0.28370345, -0.49563642, -0.01981027,\n",
            "        -0.31431606, -0.5586346 , -0.62349235, -0.0272921 , -0.28493664],\n",
            "       [-0.45599631, -0.8695159 , -0.2175562 , -0.47151705, -0.09563195,\n",
            "        -0.62207364, -0.50623542, -0.59167253, -0.2177129 , -0.32790701],\n",
            "       [-0.6247466 , -0.86560685, -0.27180158, -0.4546904 , -0.015456  ,\n",
            "        -0.46126898, -0.66073373, -0.45705361, -0.28545252, -0.27484771],\n",
            "       [-0.56387612, -0.91764888, -0.22044071, -0.46610474, -0.26491269,\n",
            "        -0.48554692, -0.66081629, -0.51591854, -0.2019116 , -0.2259551 ],\n",
            "       [-0.26975336, -0.88651263, -0.62281926,  0.92503233,  0.53200293,\n",
            "        -0.46107024, -0.61879127, -0.49692347, -0.22727511, -0.31712713],\n",
            "       [-0.42075032, -0.8852416 , -0.24036848, -0.36381363, -0.1956984 ,\n",
            "        -0.51354513, -0.60422817, -0.54453396, -0.09502698, -0.18226437],\n",
            "       [-0.41248463, -0.73838602, -0.18112833, -0.54308739,  0.1641274 ,\n",
            "        -0.38048171, -0.53570346, -0.60567867, -0.11529991, -0.3195766 ],\n",
            "       [-0.49315962, -0.87326049, -0.19631187, -0.45796344,  0.05847622,\n",
            "        -0.36564873, -0.53461753, -0.59721066, -0.11854842, -0.28169041],\n",
            "       [ 0.54530663, -0.87944866, -0.38608435, -0.87604893, -0.95812202,\n",
            "        -0.75186382, -0.63194939, -0.54912971, -0.419634  , -0.10835131],\n",
            "       [-0.66960691, -0.86307408, -0.14516275, -0.39372302,  0.08158734,\n",
            "        -0.63515478, -0.58856731, -0.54660185, -0.19125046, -0.23638915],\n",
            "       [-0.37852341, -0.87293394, -0.29371871, -0.56263105,  0.32990272,\n",
            "        -0.51161657, -0.51639741, -0.56814446, -0.10011578, -0.23551744],\n",
            "       [-0.00209243, -0.891095  , -0.08928172, -0.41854256,  0.03170872,\n",
            "        -0.55330632, -0.62014048, -0.56536895, -0.09556204, -0.2583455 ],\n",
            "       [-0.44195772, -0.92162746, -0.19221985, -0.41323588, -0.18242261,\n",
            "        -0.54027645, -0.6248134 , -0.50012459, -0.0769036 , -0.22672128],\n",
            "       [-0.66692736, -0.8749859 , -0.15111019, -0.50283089,  0.00912681,\n",
            "        -0.53848572, -0.56747933, -0.55269638, -0.25295608, -0.26990381],\n",
            "       [ 0.09001639, -0.86902457, -0.44886969, -0.38213983, -0.08522875,\n",
            "        -0.52254833, -0.4821468 , -0.70628488, -0.09790312, -0.32080704],\n",
            "       [-0.58903734, -0.87847273, -0.133992  , -0.49214749, -0.2308894 ,\n",
            "        -0.41773466, -0.6445431 , -0.52885403, -0.2318608 , -0.25607399],\n",
            "       [-0.49774529, -0.79562605, -0.21847527, -0.53291617, -0.01495997,\n",
            "        -0.52986154, -0.701541  , -0.47040394, -0.22249469, -0.2554792 ],\n",
            "       [-0.42417333, -0.89516975,  0.0669227 , -0.43651945,  0.11298483,\n",
            "        -0.58032469, -0.64066331, -0.47499765, -0.11129901, -0.25447299],\n",
            "       [-0.55349359, -0.82600352, -0.09067208, -0.39236991, -0.03671569,\n",
            "        -0.54428483, -0.70567478, -0.48797062, -0.16267235, -0.26958441],\n",
            "       [-0.50342706, -0.84471971, -0.19647271, -0.43372757, -0.04166723,\n",
            "        -0.47697783, -0.61210316, -0.51676564, -0.03601721, -0.27231528],\n",
            "       [-0.58988254, -0.88923876, -0.25187756, -0.56845837, -0.36234968,\n",
            "        -0.17884928, -0.76752111, -0.51542917, -0.14583796, -0.29928584],\n",
            "       [-0.60975182, -0.89810679, -0.19884784, -0.39509297, -0.21843575,\n",
            "        -0.49865461, -0.60333763, -0.48901544, -0.08196736, -0.21423862],\n",
            "       [-0.48624107, -0.89785488, -0.27954462, -0.34521996, -0.19602944,\n",
            "        -0.47939648, -0.61932586, -0.51076668, -0.04015993, -0.23810665],\n",
            "       [-0.54924748, -0.82436761, -0.21312185, -0.32618742, -0.14215187,\n",
            "        -0.51137425, -0.632285  , -0.54767119, -0.13470932, -0.26012385],\n",
            "       [-0.43826692, -0.88933867, -0.2849817 , -0.36159627, -0.24892595,\n",
            "        -0.54590453, -0.58094349, -0.48950905,  0.0533976 , -0.21713123],\n",
            "       [-0.19384334, -0.89543997, -0.07039313, -0.38372255, -0.10183291,\n",
            "        -0.57726827, -0.69289384, -0.53788175, -0.13341946, -0.24725599],\n",
            "       [-0.03899534, -0.91186896, -0.25398885, -0.65802623, -0.22174268,\n",
            "        -0.75325815, -0.5070924 , -0.6711485 , -0.27959085, -0.25766979],\n",
            "       [-0.52630133, -0.86067457, -0.18783295, -0.5205636 ,  0.104596  ,\n",
            "        -0.45258011, -0.62385417, -0.52859167, -0.10713589, -0.23912444],\n",
            "       [-0.46426254, -0.90519804, -0.04374986, -0.44933727, -0.01531684,\n",
            "        -0.60324331, -0.55424813, -0.5586552 , -0.05638501, -0.20522328],\n",
            "       [-0.54469549, -0.88965372, -0.11562308, -0.48003237,  0.0177207 ,\n",
            "        -0.64592288, -0.65182875, -0.47274177, -0.10144874, -0.22297945],\n",
            "       [-0.51329748, -0.78551635, -0.27263568, -0.4750972 , -0.0657062 ,\n",
            "        -0.28861158, -0.5540488 , -0.56939435, -0.07136655, -0.28653286],\n",
            "       [-0.25915243, -0.90407161, -0.40683522, -0.24812393, -0.16554521,\n",
            "        -0.73102161, -0.55763495, -0.47267715,  0.06176744, -0.17254177],\n",
            "       [-0.5003447 , -0.88694717, -0.17956645, -0.37065348,  0.04202941,\n",
            "        -0.34654989, -0.59518095, -0.54000855, -0.00788652, -0.24443076],\n",
            "       [-0.49102757, -0.88434713, -0.30901901, -0.36726179, -0.02486141,\n",
            "        -0.54361277, -0.60072821, -0.49966066, -0.13659267, -0.20659913],\n",
            "       [-0.61676801, -0.90422631, -0.23844784, -0.47593521,  0.12193093,\n",
            "        -0.45670093, -0.54724694, -0.53591809, -0.10576611, -0.20743253],\n",
            "       [-0.57947523, -0.87423445, -0.19697155, -0.42019287, -0.17476725,\n",
            "        -0.55788489, -0.54405883, -0.59889486, -0.19609435, -0.2973755 ],\n",
            "       [-0.57491134, -0.80868391, -0.17029087, -0.41148819, -0.28037313,\n",
            "        -0.51050653, -0.60790147, -0.59551766, -0.18191809, -0.31183261],\n",
            "       [-0.56334367, -0.87665949, -0.2161724 , -0.55350123,  0.11636985,\n",
            "        -0.34337611, -0.56791175, -0.57692945, -0.10678714, -0.25152144],\n",
            "       [-0.28615561, -0.80537825, -0.29098668, -0.36194187,  0.24363912,\n",
            "        -0.68598556, -0.59364662, -0.3963997 , -0.07359146, -0.2084681 ],\n",
            "       [-0.47745585, -0.85900934, -0.28099707, -0.39686624,  0.02734219,\n",
            "        -0.43411171, -0.62244084, -0.49515306, -0.17239473, -0.2132924 ],\n",
            "       [-0.41349958, -0.89333938, -0.25719482, -0.57549589,  0.04396589,\n",
            "        -0.5629374 , -0.74676501, -0.42909491, -0.08440137, -0.22150434],\n",
            "       [-0.37334284, -0.84974909, -0.2255407 , -0.50543541,  0.17763872,\n",
            "        -0.47321183, -0.53445893, -0.59117409, -0.08660936, -0.30698779],\n",
            "       [-0.24676739, -0.82260022,  0.05356339, -0.41300019,  0.07532986,\n",
            "        -0.73882001, -0.76367123, -0.48452755, -0.05733498, -0.32252224],\n",
            "       [-0.40221276, -0.84825495, -0.18242841, -0.42690568,  0.01418368,\n",
            "        -0.51948542, -0.69114006, -0.45445654, -0.04690817, -0.29584791],\n",
            "       [-0.43173231, -0.83694304, -0.27573118, -0.57294631,  0.00525774,\n",
            "        -0.39793744, -0.72630731, -0.47706131, -0.23668843, -0.29031364],\n",
            "       [-0.60928522, -0.8580276 , -0.16862227, -0.41146697, -0.0255846 ,\n",
            "        -0.45694745, -0.5947067 , -0.54008315, -0.15928877, -0.24443515],\n",
            "       [-0.55454034, -0.85300835, -0.0842492 , -0.43064592, -0.14298753,\n",
            "        -0.45980042, -0.50861927, -0.61409058, -0.09887904, -0.27080365],\n",
            "       [-0.36811249, -0.83610154, -0.3884384 , -0.36428278, -0.04882507,\n",
            "        -0.29504229, -0.67780843, -0.49762479,  0.05567466, -0.24817058],\n",
            "       [-0.55552562, -0.87276166, -0.24357067, -0.50035786, -0.11949711,\n",
            "        -0.50764516, -0.59789517, -0.51533414, -0.278765  , -0.23574897],\n",
            "       [-0.19829793, -0.91432935, -0.33939172, -0.62743894, -0.00587514,\n",
            "        -0.57155223, -0.42948384, -0.66120678, -0.02715559, -0.17854684],\n",
            "       [-0.54602718, -0.87170967, -0.18791082, -0.37315795, -0.14433349,\n",
            "        -0.52545558, -0.59987174, -0.58148847, -0.20126407, -0.20825916],\n",
            "       [-0.52363391, -0.83259136, -0.20632792, -0.50589636, -0.08424746,\n",
            "        -0.40111727, -0.66273598, -0.49458697, -0.17785921, -0.20378698],\n",
            "       [-0.59824617, -0.79700093, -0.19608822, -0.43857649,  0.07255662,\n",
            "        -0.54548098, -0.60757665, -0.4902181 , -0.09595657, -0.22863318],\n",
            "       [-0.46511857, -0.85199094, -0.24243613, -0.50503436,  0.04882727,\n",
            "        -0.27685445, -0.54899937, -0.63129022, -0.16659515, -0.31994286],\n",
            "       [-0.51169001, -0.88059223, -0.23375279, -0.348329  , -0.10579666,\n",
            "        -0.59603008, -0.59083843, -0.51405941, -0.12377239, -0.18538751],\n",
            "       [-0.22319641, -0.85488319, -0.09445843, -0.46890236, -0.18832127,\n",
            "        -0.6241784 , -0.59108438, -0.58997824, -0.12490055, -0.21582779],\n",
            "       [-0.46907657, -0.87457091, -0.28598122, -0.43368142, -0.15341283,\n",
            "        -0.43523694, -0.67541262, -0.50847178, -0.15542236, -0.20663215],\n",
            "       [-0.66988254, -0.88967958, -0.25650799, -0.33048571, -0.35342715,\n",
            "        -0.47622602, -0.54344972, -0.56089606, -0.02640692, -0.25976649],\n",
            "       [-0.58488655, -0.88676949, -0.19077381, -0.45329842,  0.04973951,\n",
            "        -0.58717297, -0.68033114, -0.48574225, -0.27058453, -0.20912911],\n",
            "       [-0.16175727, -0.90405681, -0.37032063, -0.38291718, -0.12188647,\n",
            "        -0.46902436, -0.57707804, -0.53722772, -0.13630207, -0.16613654],\n",
            "       [-0.66232483, -0.8641583 , -0.15090564, -0.50511748,  0.02268228,\n",
            "        -0.5421796 , -0.56041083, -0.55981148, -0.25302812, -0.26638586],\n",
            "       [-0.45203554, -0.88011729, -0.19531126, -0.51774226,  0.02531522,\n",
            "        -0.47186367, -0.68663197, -0.47421767, -0.18381072, -0.26366721],\n",
            "       [-0.16462204, -0.89964258, -0.33866498, -0.47616938,  0.14382035,\n",
            "        -0.41770594, -0.54033541, -0.48564901, -0.12062692, -0.14795767],\n",
            "       [-0.39981167, -0.8832025 , -0.25569071, -0.45295445, -0.06783216,\n",
            "        -0.54468961, -0.6041683 , -0.54477225, -0.20696946, -0.25265158],\n",
            "       [-0.54167247, -0.89302523, -0.28713693, -0.5257179 , -0.43218133,\n",
            "        -0.43056324, -0.59724615, -0.51728668, -0.07889246, -0.24164676],\n",
            "       [-0.7213671 , -0.87093337, -0.15179419, -0.40324263, -0.09741651,\n",
            "        -0.51024432, -0.55622403, -0.56252585, -0.24104632, -0.31428086],\n",
            "       [-0.40728832, -0.89583284, -0.25783651, -0.45195251, -0.06543696,\n",
            "        -0.5565169 , -0.60273144, -0.54609858, -0.22053166, -0.25060914],\n",
            "       [-0.26460654, -0.92763386,  0.14973679, -0.42376429,  0.09334932,\n",
            "        -0.58108978, -0.65361198, -0.62795317, -0.07526124, -0.27281549],\n",
            "       [-0.60804167, -0.90053776, -0.27406754, -0.58955234,  0.18416716,\n",
            "        -0.50725655, -0.70550182, -0.37292021, -0.2792186 , -0.15388232]]), 0: array([[-4.96527310e-01, -8.84802553e-01, -2.38046599e-01,\n",
            "        -4.93886007e-01,  6.80599967e-03, -4.27977038e-01,\n",
            "        -5.70894293e-01, -5.02891055e-01, -1.21328621e-02,\n",
            "        -2.38202197e-01],\n",
            "       [-5.10573541e-01, -8.29049605e-01, -1.75942901e-01,\n",
            "        -4.07140322e-01,  8.37389113e-04, -5.82524332e-01,\n",
            "        -6.19801034e-01, -4.96385582e-01, -9.97121299e-02,\n",
            "        -2.20347281e-01],\n",
            "       [-2.51610589e-01, -8.87572176e-01, -1.56627156e-01,\n",
            "        -4.19468501e-01, -6.04160972e-02, -4.52718385e-01,\n",
            "        -5.81508916e-01, -5.32908857e-01, -1.16014591e-01,\n",
            "        -2.72372727e-01],\n",
            "       [-3.47272197e-01, -8.80479985e-01, -3.06543916e-01,\n",
            "        -5.86066544e-01,  2.88498000e-02, -4.09311319e-01,\n",
            "        -5.17583965e-01, -5.92521320e-01, -9.53269619e-02,\n",
            "        -2.07786397e-01],\n",
            "       [-3.80052069e-01, -8.72561784e-01, -2.43573978e-01,\n",
            "        -6.16658118e-01,  4.42494679e-01, -4.71708950e-01,\n",
            "        -5.39759167e-01, -5.45884841e-01, -1.10420368e-01,\n",
            "        -2.67654746e-01],\n",
            "       [-4.20168699e-01, -8.94924783e-01, -3.35100120e-01,\n",
            "        -4.72323215e-01, -1.54733371e-01, -7.66851786e-01,\n",
            "        -4.99263216e-01, -5.43637182e-01, -8.79104291e-02,\n",
            "        -1.87033212e-01],\n",
            "       [-5.44015166e-01, -8.88466149e-01, -2.84087537e-01,\n",
            "        -3.99663186e-01,  1.97602894e-01, -6.50842268e-01,\n",
            "        -5.46973470e-01, -4.75665829e-01, -5.75850030e-03,\n",
            "        -1.49651105e-01],\n",
            "       [-5.37913060e-01, -8.53792933e-01, -2.34561807e-01,\n",
            "        -3.65436361e-01,  1.49368211e-01, -4.20336597e-01,\n",
            "        -5.35742741e-01, -5.57484121e-01, -5.14011350e-02,\n",
            "        -2.09170910e-01],\n",
            "       [-5.62473464e-01, -8.09791353e-01, -1.87409561e-01,\n",
            "        -4.46167804e-01,  1.03999305e-01, -3.78102687e-01,\n",
            "        -5.27870197e-01, -5.94975042e-01, -7.74133365e-02,\n",
            "        -2.59295570e-01],\n",
            "       [-5.05541888e-01, -9.04182690e-01, -1.94650919e-01,\n",
            "        -4.03928990e-01, -1.80573794e-01, -4.19286103e-01,\n",
            "        -6.26988807e-01, -5.21549771e-01, -1.30945000e-01,\n",
            "        -2.21704956e-01],\n",
            "       [-3.74955112e-01, -8.64999123e-01, -2.93167053e-01,\n",
            "        -5.63279113e-01,  3.32831349e-01, -5.14593439e-01,\n",
            "        -5.15448272e-01, -5.69124380e-01, -9.89133052e-02,\n",
            "        -2.37358368e-01],\n",
            "       [-6.73462061e-01, -9.02273438e-01, -2.44973708e-01,\n",
            "        -3.95451437e-01, -2.18209817e-01, -5.16415049e-01,\n",
            "        -5.78685163e-01, -5.51222793e-01, -7.81103154e-02,\n",
            "        -2.05230395e-01],\n",
            "       [-6.75814136e-01, -8.75412087e-01, -1.66106043e-01,\n",
            "        -3.25932041e-01, -8.28484533e-02, -6.32039249e-01,\n",
            "        -5.68988981e-01, -5.43225515e-01, -1.81071320e-01,\n",
            "        -2.59643033e-01],\n",
            "       [-2.77314224e-01, -7.43231262e-01, -7.84328584e-02,\n",
            "        -4.17652542e-01,  6.31885251e-02, -4.88506857e-01,\n",
            "        -6.34915669e-01, -4.44824218e-01, -8.40374140e-02,\n",
            "        -1.81658808e-01],\n",
            "       [-3.77747776e-01, -8.68416662e-01, -2.43297934e-01,\n",
            "        -6.17908260e-01,  4.44753895e-01, -4.69953338e-01,\n",
            "        -5.38543901e-01, -5.48932591e-01, -1.08840376e-01,\n",
            "        -2.72273709e-01],\n",
            "       [-3.71123182e-01, -8.95107395e-01, -2.58031356e-01,\n",
            "        -5.65963461e-01, -4.00925457e-02, -3.47570618e-01,\n",
            "        -5.48385307e-01, -6.26383424e-01, -8.25734222e-02,\n",
            "        -2.43496931e-01],\n",
            "       [-4.08735200e-01, -7.94792015e-01, -2.56010543e-01,\n",
            "        -2.72375057e-01, -3.04182507e-02, -5.58766549e-01,\n",
            "        -6.06765114e-01, -4.91708815e-01, -7.51041312e-03,\n",
            "        -2.62396409e-01],\n",
            "       [-4.12734143e-01, -9.35305510e-01, -1.92477000e-01,\n",
            "        -4.85246461e-01,  5.66393478e-02, -4.90164052e-01,\n",
            "        -6.40559357e-01, -4.61828525e-01, -1.67292856e-01,\n",
            "        -1.90020641e-01],\n",
            "       [-4.27913307e-01, -8.59757634e-01, -2.78400362e-01,\n",
            "        -5.38953071e-01, -3.87405951e-02, -3.15846135e-01,\n",
            "        -5.33134997e-01, -6.61061790e-01, -9.40638859e-02,\n",
            "        -2.91594450e-01],\n",
            "       [-5.08685546e-01, -9.01339322e-01, -1.47451750e-01,\n",
            "        -4.24541545e-01, -1.39656045e-01, -4.84112144e-01,\n",
            "        -6.77516738e-01, -4.58629283e-01, -2.26790424e-02,\n",
            "        -2.13748552e-01],\n",
            "       [-5.71012347e-01, -8.40237536e-01, -2.09861702e-01,\n",
            "        -5.22692588e-01, -1.09767794e-01, -3.68735484e-01,\n",
            "        -6.68631117e-01, -4.80045682e-01, -1.16997094e-01,\n",
            "        -2.24773559e-01],\n",
            "       [-5.19664349e-01, -8.97126184e-01, -2.99411519e-01,\n",
            "        -3.59535652e-01,  1.81472194e-01, -4.95471896e-01,\n",
            "        -7.04378822e-01, -4.30563391e-01, -1.05846109e-01,\n",
            "        -1.54416606e-01],\n",
            "       [-4.85375327e-01, -9.09234687e-01,  3.48292056e-02,\n",
            "        -3.82078851e-01, -7.74956339e-02, -5.49550302e-01,\n",
            "        -7.85440988e-01, -4.99712548e-01, -1.49719672e-01,\n",
            "        -3.46408190e-01],\n",
            "       [-5.18131463e-01, -9.18560480e-01, -1.75755542e-01,\n",
            "        -3.63009554e-01,  7.57618188e-03, -4.81026773e-01,\n",
            "        -6.56406928e-01, -5.11611325e-01, -1.82969836e-01,\n",
            "        -2.06222603e-01],\n",
            "       [-3.65549031e-01, -8.65638283e-01, -1.04793696e-01,\n",
            "        -4.17220769e-01, -6.27151545e-02, -4.88206374e-01,\n",
            "        -5.90128152e-01, -5.71829638e-01, -4.65735455e-02,\n",
            "        -2.68163429e-01],\n",
            "       [-2.21401798e-01, -8.62385564e-01,  1.06673193e-01,\n",
            "        -3.54233483e-01, -2.05137446e-01, -6.15248408e-01,\n",
            "        -6.32967611e-01, -4.83923105e-01, -3.23469700e-02,\n",
            "        -2.52712728e-01],\n",
            "       [-5.26853989e-01, -8.76610498e-01, -8.75060346e-02,\n",
            "        -4.02420551e-01, -1.93942180e-01, -6.03339793e-01,\n",
            "        -6.53222830e-01, -5.20329838e-01, -1.13718860e-01,\n",
            "        -2.58151749e-01],\n",
            "       [-6.45707407e-01, -8.97930562e-01, -1.82019165e-01,\n",
            "        -2.76090657e-01, -1.00154132e-01, -6.66833147e-01,\n",
            "        -5.81727753e-01, -5.09864624e-01, -2.84766612e-02,\n",
            "        -2.22430980e-01],\n",
            "       [-4.70910105e-01, -8.79062519e-01, -1.60220868e-01,\n",
            "        -3.95078899e-01, -1.29315811e-01, -5.57419922e-01,\n",
            "        -5.85061610e-01, -5.23468617e-01, -8.28066218e-02,\n",
            "        -2.19623342e-01],\n",
            "       [-2.82160307e-01, -8.91211662e-01, -8.71849850e-02,\n",
            "        -4.97075348e-01, -6.95341361e-02, -5.73958955e-01,\n",
            "        -5.21325591e-01, -5.83161074e-01, -1.07257071e-01,\n",
            "        -1.98405509e-01],\n",
            "       [-8.04687552e-01, -8.63867701e-01, -1.95965027e-01,\n",
            "        -5.02095249e-01, -3.80778542e-01, -3.12066710e-01,\n",
            "        -5.08128222e-01, -6.13891762e-01,  3.06551499e-03,\n",
            "        -2.56151788e-01],\n",
            "       [-3.68259374e-01, -8.14841029e-01, -3.21185160e-01,\n",
            "        -4.27030430e-01, -3.37927477e-01, -6.27450890e-01,\n",
            "        -5.18147527e-01, -6.17793527e-01, -1.15974711e-01,\n",
            "        -3.01132367e-01],\n",
            "       [-5.62755345e-01, -8.75063459e-01, -1.84069922e-01,\n",
            "        -4.13829069e-01,  8.58976830e-02, -6.49425320e-01,\n",
            "        -6.18145131e-01, -5.17626657e-01, -1.63475864e-01,\n",
            "        -2.23291374e-01],\n",
            "       [-6.56597623e-01, -8.93103141e-01, -2.11446324e-01,\n",
            "        -4.30865760e-01, -3.18669162e-01, -4.23499455e-01,\n",
            "        -5.98750847e-01, -5.05468312e-01, -8.36197086e-02,\n",
            "        -2.47674046e-01],\n",
            "       [-6.07195072e-01, -8.81367522e-01, -2.33437573e-01,\n",
            "        -4.36032795e-01, -3.73474444e-03, -4.41916483e-01,\n",
            "        -5.52928328e-01, -5.43339715e-01,  5.64737414e-02,\n",
            "        -2.10199853e-01],\n",
            "       [-6.03509350e-01, -8.80827414e-01, -1.98375306e-01,\n",
            "        -4.77022412e-01, -3.49740520e-02, -3.97102635e-01,\n",
            "        -5.61189045e-01, -5.65143090e-01, -1.39958795e-01,\n",
            "        -2.59523431e-01],\n",
            "       [-6.94298243e-01, -8.84411030e-01, -2.48598127e-01,\n",
            "        -4.75812909e-01, -3.73772183e-01, -1.64802667e-01,\n",
            "        -6.92258429e-01, -4.91316093e-01,  1.64065137e-01,\n",
            "        -2.77730089e-01],\n",
            "       [-5.77285664e-01, -8.47995593e-01, -2.77144947e-01,\n",
            "        -5.71030969e-01, -5.64947056e-02, -3.20782853e-01,\n",
            "        -8.47549283e-01, -4.15977189e-01, -2.44783824e-01,\n",
            "        -2.24316252e-01],\n",
            "       [-1.44922370e-02, -9.30187129e-01,  1.44460615e-01,\n",
            "        -4.26878615e-01, -7.27918191e-02, -5.51377243e-01,\n",
            "        -5.67290389e-01, -6.40059107e-01, -5.42396763e-02,\n",
            "        -1.97090242e-01],\n",
            "       [-2.91953889e-01, -9.00360676e-01,  2.59211797e-02,\n",
            "        -4.46343171e-01, -1.71341457e-02, -5.18384385e-01,\n",
            "        -6.68203842e-01, -5.20332029e-01, -1.04362645e-01,\n",
            "        -2.45480866e-01],\n",
            "       [-3.86965708e-01, -8.51259522e-01, -3.46171807e-01,\n",
            "        -3.49208876e-01, -2.82849425e-01, -4.75491843e-01,\n",
            "        -5.60785766e-01, -4.97718716e-01,  9.83540982e-02,\n",
            "        -1.34937785e-01],\n",
            "       [-5.40552025e-01, -8.88753763e-01, -2.66259718e-01,\n",
            "        -4.15395840e-01, -3.44261477e-01, -4.85473346e-01,\n",
            "        -4.96565687e-01, -5.61767810e-01, -6.32099948e-02,\n",
            "        -2.50518520e-01],\n",
            "       [-4.95206482e-01, -8.55874202e-01, -2.16550597e-01,\n",
            "        -5.41245136e-01,  1.26640543e-01, -2.95066254e-01,\n",
            "        -5.34636097e-01, -5.94566491e-01, -4.48172762e-02,\n",
            "        -2.54220788e-01],\n",
            "       [-5.00413778e-01, -8.27024278e-01, -2.23147175e-01,\n",
            "        -5.07624710e-01,  2.08254578e-03, -3.77888972e-01,\n",
            "        -6.20427624e-01, -5.46654890e-01, -1.28502649e-01,\n",
            "        -2.74820501e-01],\n",
            "       [-4.52548417e-01, -8.56846174e-01, -3.11328826e-01,\n",
            "        -3.04706562e-01, -3.37725740e-01, -5.52439347e-01,\n",
            "        -5.98025347e-01, -5.17227309e-01, -1.68925751e-02,\n",
            "        -2.24461238e-01],\n",
            "       [-3.94638781e-01, -8.97275952e-01,  1.60124169e-02,\n",
            "        -3.34834319e-01, -9.87280147e-03, -5.25832030e-01,\n",
            "        -6.07447484e-01, -5.25367414e-01, -4.77989944e-02,\n",
            "        -2.43185918e-01],\n",
            "       [-2.65477733e-01, -8.43303124e-01, -1.94110424e-01,\n",
            "        -4.48074703e-01, -8.16885429e-02, -5.67700365e-01,\n",
            "        -6.18492159e-01, -5.38252918e-01, -1.52052231e-01,\n",
            "        -3.14908683e-01],\n",
            "       [-5.73384018e-01, -8.60484273e-01, -1.78564089e-01,\n",
            "        -4.80540842e-01,  2.94467996e-01, -2.86290172e-01,\n",
            "        -6.04049498e-01, -5.67591353e-01, -6.63241566e-02,\n",
            "        -2.97999583e-01],\n",
            "       [-4.85047014e-01, -8.52194236e-01, -3.01306357e-01,\n",
            "        -5.16070272e-01,  2.06379374e-01, -3.76380887e-01,\n",
            "        -5.73957453e-01, -5.25702510e-01,  1.04843431e-02,\n",
            "        -2.47899132e-01],\n",
            "       [-6.52253986e-01, -8.92989709e-01, -2.21874607e-01,\n",
            "        -3.90346250e-01, -2.32804733e-01, -3.63751931e-01,\n",
            "        -6.56862367e-01, -5.11227884e-01, -2.09802224e-01,\n",
            "        -2.44017453e-01],\n",
            "       [-7.00189401e-01, -8.78085454e-01, -1.58661424e-01,\n",
            "        -4.53806341e-01,  8.81673493e-02, -6.35201424e-01,\n",
            "        -5.58328388e-01, -5.39439104e-01, -3.25389368e-01,\n",
            "        -2.97262946e-01],\n",
            "       [-3.98479942e-01, -7.97374279e-01, -2.77743623e-01,\n",
            "        -2.55631462e-01, -2.61074824e-01, -6.05200401e-01,\n",
            "        -5.08570715e-01, -5.40170324e-01,  3.17036098e-02,\n",
            "        -2.48970803e-01],\n",
            "       [-4.36553663e-01, -8.59727280e-01, -1.81550398e-01,\n",
            "        -5.70856702e-01, -5.34538410e-03, -5.21179151e-01,\n",
            "        -5.10250055e-01, -5.80992107e-01, -1.17826078e-01,\n",
            "        -2.29591318e-01],\n",
            "       [-3.58670244e-02, -8.84424155e-01, -3.65153005e-01,\n",
            "        -5.63437607e-01, -2.63583416e-01, -5.13735706e-01,\n",
            "        -5.85930108e-01, -5.90225207e-01, -1.99441616e-01,\n",
            "        -2.35669530e-01],\n",
            "       [-4.92083366e-01, -8.63777675e-01, -3.07670244e-01,\n",
            "        -2.58826170e-01,  2.73046161e-02, -6.14020101e-01,\n",
            "        -5.08816638e-01, -5.39499049e-01,  1.82250943e-02,\n",
            "        -1.81658084e-01],\n",
            "       [-5.54322847e-01, -8.62607174e-01, -1.65078886e-01,\n",
            "        -5.58572408e-01,  2.35981590e-02, -3.89172592e-01,\n",
            "        -5.78306488e-01, -5.23619000e-01, -1.51270351e-01,\n",
            "        -1.74239320e-01],\n",
            "       [-5.30924745e-01, -8.83894200e-01, -3.26583801e-01,\n",
            "        -2.31306296e-01, -2.21486237e-01, -5.15769183e-01,\n",
            "        -5.92604645e-01, -5.23430094e-01, -2.21355811e-03,\n",
            "        -1.97664345e-01],\n",
            "       [-5.21611462e-01, -8.62278002e-01, -3.70663060e-02,\n",
            "        -3.00146990e-01, -1.78646240e-01, -4.95083587e-01,\n",
            "        -6.25252455e-01, -5.76834447e-01, -1.67003919e-01,\n",
            "        -2.46925412e-01],\n",
            "       [-3.86104948e-01, -8.21434178e-01, -2.02635879e-01,\n",
            "        -4.71588679e-01,  1.92451056e-01, -4.36929369e-01,\n",
            "        -6.19704662e-01, -4.96375499e-01, -1.39060149e-01,\n",
            "        -2.15199442e-01],\n",
            "       [-2.26977083e-01, -8.51261968e-01, -2.09714743e-01,\n",
            "        -5.02235772e-01, -5.32539559e-02, -5.01086473e-01,\n",
            "        -6.54925859e-01, -5.31626792e-01, -1.86917954e-01,\n",
            "        -2.60660862e-01],\n",
            "       [-5.69475610e-01, -8.57295967e-01, -1.91339136e-01,\n",
            "        -6.00566827e-01, -9.32794507e-02, -5.28328337e-01,\n",
            "        -5.48039473e-01, -4.87485911e-01, -1.80303361e-01,\n",
            "        -1.78472087e-01],\n",
            "       [-1.98297932e-01, -9.14329353e-01, -3.39391718e-01,\n",
            "        -6.27438943e-01, -5.87513989e-03, -5.71552226e-01,\n",
            "        -4.29483844e-01, -6.61206784e-01, -2.71555879e-02,\n",
            "        -1.78546839e-01],\n",
            "       [-4.09631611e-01, -8.87727151e-01, -2.13214090e-01,\n",
            "        -3.94346638e-01, -1.19199532e-01, -4.18976464e-01,\n",
            "        -5.75125768e-01, -5.40327122e-01, -1.54176108e-01,\n",
            "        -2.03358131e-01],\n",
            "       [-4.05080532e-01, -8.67625265e-01, -2.42418300e-01,\n",
            "        -4.44325365e-01, -1.00389840e-01, -5.18010490e-01,\n",
            "        -6.17094665e-01, -5.14847832e-01, -1.29055664e-01,\n",
            "        -2.22286083e-01],\n",
            "       [-5.91533585e-01, -7.54541970e-01, -2.32458210e-01,\n",
            "        -3.25896987e-01, -1.80279461e-01, -5.56390221e-01,\n",
            "        -5.71028226e-01, -5.26227324e-01,  8.70839594e-02,\n",
            "        -2.54248447e-01],\n",
            "       [-5.08051432e-01, -8.69950614e-01, -2.08692660e-01,\n",
            "        -3.93831385e-01,  1.20702958e-01, -5.07177442e-01,\n",
            "        -5.32530975e-01, -5.54810991e-01, -3.30787302e-02,\n",
            "        -2.41683701e-01],\n",
            "       [-5.39481816e-01, -8.54313413e-01, -1.92654730e-01,\n",
            "        -6.24837486e-01,  1.20409781e-01, -2.43811103e-01,\n",
            "        -5.98374503e-01, -5.61864867e-01, -7.84912968e-02,\n",
            "        -2.54245888e-01],\n",
            "       [-4.53281299e-01, -7.89041581e-01, -2.67697244e-01,\n",
            "        -2.94228891e-01, -1.81943709e-01, -5.86305242e-01,\n",
            "        -5.39127961e-01, -5.40992927e-01, -8.40201370e-02,\n",
            "        -1.66710463e-01],\n",
            "       [-3.24903492e-01, -8.09838886e-01, -2.41053243e-01,\n",
            "        -4.36233513e-01, -2.22562110e-01, -4.45964955e-01,\n",
            "        -5.78662912e-01, -5.58781550e-01, -8.92539595e-02,\n",
            "        -2.89509332e-01],\n",
            "       [-3.26569732e-01, -8.72010794e-01,  8.88237687e-02,\n",
            "        -3.78289628e-01, -1.48897237e-01, -4.85058632e-01,\n",
            "        -7.10566392e-01, -5.17816441e-01, -4.82510981e-02,\n",
            "        -2.63265482e-01],\n",
            "       [-6.19168768e-01, -8.41043591e-01, -2.17342208e-01,\n",
            "        -4.42583869e-01, -2.06745423e-02, -6.14111260e-01,\n",
            "        -6.17206334e-01, -4.75936768e-01, -1.63655001e-01,\n",
            "        -2.06738589e-01],\n",
            "       [-5.07146649e-01, -8.66113480e-01, -2.08440573e-01,\n",
            "        -3.94884761e-01,  1.30505387e-01, -5.15559884e-01,\n",
            "        -5.28787114e-01, -5.59842485e-01, -3.73937394e-02,\n",
            "        -2.43340061e-01],\n",
            "       [-5.61132171e-01, -8.76159704e-01, -2.40250323e-01,\n",
            "        -4.71946651e-01, -1.63681247e-01, -4.70680901e-01,\n",
            "        -6.28648320e-01, -4.98074650e-01, -1.44649546e-01,\n",
            "        -2.06641532e-01],\n",
            "       [-4.95824029e-01, -8.85461955e-01, -2.13021560e-01,\n",
            "        -4.12168144e-01,  8.69529547e-02, -4.29994537e-01,\n",
            "        -5.40439906e-01, -5.76416827e-01, -3.41410417e-02,\n",
            "        -2.46481100e-01],\n",
            "       [-3.94703340e-01, -8.74777858e-01, -2.84011406e-01,\n",
            "        -4.97351528e-01, -9.29947668e-03, -3.13520226e-01,\n",
            "        -5.53721460e-01, -6.30161761e-01, -3.09800956e-02,\n",
            "        -2.91551695e-01],\n",
            "       [-5.77003975e-01, -8.61018447e-01, -2.05208801e-01,\n",
            "        -5.30678325e-01,  8.60211620e-02, -4.69376252e-01,\n",
            "        -5.71548592e-01, -5.33051182e-01, -1.12373697e-01,\n",
            "        -2.29899881e-01],\n",
            "       [-1.11734571e-01, -9.41728475e-01, -6.88284119e-01,\n",
            "         8.04067210e-01,  3.99704324e-01, -2.70048048e-01,\n",
            "        -7.58571030e-01, -4.65715803e-01, -4.83418199e-01,\n",
            "        -3.82720075e-01],\n",
            "       [-6.08829497e-01, -8.20113019e-01, -2.28518708e-01,\n",
            "        -5.45127723e-01,  1.71303238e-02, -4.13269978e-01,\n",
            "        -6.27579114e-01, -5.06326726e-01, -1.51977402e-01,\n",
            "        -2.42632973e-01],\n",
            "       [-4.16690306e-01, -9.00176108e-01, -2.45614766e-01,\n",
            "        -4.43679061e-01, -1.02304071e-01, -5.07620130e-01,\n",
            "        -6.19173020e-01, -5.21772076e-01, -1.42443429e-01,\n",
            "        -2.39838409e-01],\n",
            "       [-5.85571589e-01, -8.75267709e-01, -1.74412170e-01,\n",
            "        -4.54432075e-01, -3.60196627e-03, -2.15071119e-01,\n",
            "        -6.60902625e-01, -4.90799679e-01, -1.56448838e-01,\n",
            "        -2.44503764e-01],\n",
            "       [-4.64689927e-01, -8.80499385e-01, -2.09481638e-01,\n",
            "        -4.41207876e-01, -8.18870335e-02, -5.28555447e-01,\n",
            "        -6.79657435e-01, -4.98949930e-01, -1.07288458e-01,\n",
            "        -2.57753236e-01],\n",
            "       [-6.81546522e-01, -8.86498872e-01, -1.87510454e-01,\n",
            "        -3.83427192e-01, -6.36613055e-02, -4.13880591e-01,\n",
            "        -6.24717993e-01, -5.16608981e-01, -1.10225030e-01,\n",
            "        -2.36103366e-01],\n",
            "       [-4.33087348e-01, -7.05533606e-01, -2.11458871e-01,\n",
            "        -3.62521103e-01, -5.75063236e-02, -5.63848812e-01,\n",
            "        -5.69222093e-01, -5.58487070e-01, -1.31847535e-01,\n",
            "        -3.17336391e-01],\n",
            "       [-5.30460871e-01, -8.49822579e-01, -2.58839012e-01,\n",
            "        -4.95960443e-01, -9.24566834e-02, -5.79348175e-01,\n",
            "        -5.95911354e-01, -5.18820374e-01, -1.66876431e-01,\n",
            "        -2.00454965e-01],\n",
            "       [-2.36213629e-01, -8.46441920e-01, -3.06732044e-01,\n",
            "        -3.53767722e-01, -4.02669443e-01, -5.00801493e-01,\n",
            "        -6.38479520e-01, -4.82716344e-01, -8.51819378e-02,\n",
            "        -2.30368263e-01],\n",
            "       [-7.64602477e-01, -7.28481743e-01, -1.63834895e-01,\n",
            "        -4.47305419e-01, -3.17278301e-01, -3.81117116e-01,\n",
            "        -6.11395704e-01, -5.02492511e-01, -7.39949943e-02,\n",
            "        -2.00048072e-01],\n",
            "       [-4.70465374e-01, -9.14459017e-01, -2.52085658e-01,\n",
            "        -5.00506742e-01, -5.76403577e-02, -4.90210774e-01,\n",
            "        -5.53722463e-01, -5.06304939e-01, -2.74535174e-01,\n",
            "        -1.61747766e-01],\n",
            "       [-4.17892384e-01, -8.68149413e-01, -2.72340780e-01,\n",
            "        -4.66304924e-01,  2.79488994e-02, -5.53257184e-01,\n",
            "        -6.91852588e-01, -4.34672260e-01, -1.46629327e-01,\n",
            "        -1.70626855e-01],\n",
            "       [-4.99399261e-01, -8.42973066e-01, -7.59955520e-02,\n",
            "        -4.94396695e-01, -3.71759328e-02, -6.90101259e-01,\n",
            "        -6.70412303e-01, -4.79188837e-01, -7.09208258e-02,\n",
            "        -2.48384116e-01],\n",
            "       [-6.11174608e-01, -9.01889630e-01, -1.98443611e-01,\n",
            "        -3.97537992e-01, -2.08370757e-01, -4.99683000e-01,\n",
            "        -5.99249221e-01, -4.98478789e-01, -8.56539190e-02,\n",
            "        -2.31886397e-01],\n",
            "       [-5.29613637e-01, -8.51756982e-01, -1.68187232e-01,\n",
            "        -3.95164374e-01, -1.01235299e-01, -5.55623233e-01,\n",
            "        -5.68718772e-01, -5.48441130e-01, -1.99588520e-01,\n",
            "        -2.41850799e-01],\n",
            "       [-6.03318958e-01, -7.05877930e-01, -1.88871150e-01,\n",
            "        -4.29691826e-01, -2.89007394e-01, -3.05194720e-01,\n",
            "        -6.49407534e-01, -5.34759996e-01, -6.66268797e-03,\n",
            "        -2.94221842e-01],\n",
            "       [-5.75279484e-01, -9.03547724e-01, -2.10080247e-01,\n",
            "        -4.84467143e-01,  4.12164934e-02, -4.52523244e-01,\n",
            "        -6.52044895e-01, -5.00903646e-01, -1.89071069e-01,\n",
            "        -2.08020294e-01],\n",
            "       [-4.17444392e-01, -9.35245372e-01, -1.94177463e-01,\n",
            "        -3.92798114e-01, -1.48201293e-01, -2.63126916e-01,\n",
            "        -5.55180217e-01, -5.40526534e-01, -2.71111966e-01,\n",
            "        -1.43915513e-01],\n",
            "       [-6.90181024e-01, -8.79615888e-01, -1.96405133e-01,\n",
            "        -4.84351949e-01,  2.65888737e-02, -6.74196112e-01,\n",
            "        -5.71807960e-01, -4.86032789e-01, -2.19845834e-01,\n",
            "        -1.43401669e-01],\n",
            "       [-4.33944552e-01, -8.89379392e-01, -3.74281742e-01,\n",
            "        -3.09923952e-01,  1.19715186e-02, -5.03053251e-01,\n",
            "        -5.29553265e-01, -5.55591528e-01, -7.50622730e-02,\n",
            "        -2.01806840e-01],\n",
            "       [-4.59612385e-01, -8.41467836e-01, -2.65680120e-01,\n",
            "        -3.35965347e-01, -2.84949220e-01, -6.25748621e-01,\n",
            "        -5.61115189e-01, -5.29561740e-01,  6.93995907e-02,\n",
            "        -1.83220113e-01],\n",
            "       [-6.15419482e-01, -8.57714693e-01, -1.64131013e-01,\n",
            "        -3.92057176e-01, -9.20210849e-02, -5.83352209e-01,\n",
            "        -6.36455001e-01, -5.19696042e-01, -1.55874566e-01,\n",
            "        -2.48910348e-01],\n",
            "       [-3.51658867e-01, -8.73143258e-01, -6.06293016e-02,\n",
            "        -4.01340971e-01, -2.44667003e-01, -5.04821432e-01,\n",
            "        -6.32530803e-01, -5.41202023e-01, -1.15156975e-01,\n",
            "        -2.36742021e-01],\n",
            "       [-4.15527088e-01, -8.64456874e-01, -2.63232268e-01,\n",
            "        -4.29454312e-01, -1.45661710e-01, -4.99269710e-01,\n",
            "        -6.71869478e-01, -5.10008878e-01, -1.43993902e-01,\n",
            "        -2.75218196e-01]])}\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "gA0QAf-zOYZG",
        "colab_type": "text"
      },
      "source": [
        "## Classical SVM \n"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "eNP5hBF6OXg0",
        "colab_type": "code",
        "outputId": "42c531a6-42d9-4cc1-a620-492d01dceed0",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 34
        }
      },
      "source": [
        "svcClassifier = SVC(gamma = 'auto')\n",
        "svcClassifier.fit(X_train,y_train)\n",
        "print(svcClassifier.score(X_test,y_test))"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "0.5\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "4qdpLFBIPApZ",
        "colab_type": "text"
      },
      "source": [
        "## Classical DNN\n"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "_xMR8orzPDU9",
        "colab_type": "code",
        "outputId": "c9657d7b-ee65-433c-8d7a-4ab96bb614c9",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        }
      },
      "source": [
        "n = config['qbits']\n",
        "model = tf.keras.Sequential([\n",
        "    tf.keras.layers.Dense(1024,activation='relu',input_shape=(n,)),\n",
        "    tf.keras.layers.Dense(512,activation='relu'),\n",
        "    tf.keras.layers.Dense(512,activation='relu'),\n",
        "    tf.keras.layers.Dense(256,activation='relu'),\n",
        "    tf.keras.layers.Dense(128,activation='relu'),\n",
        "    tf.keras.layers.Dense(64,activation='relu'),\n",
        "    tf.keras.layers.Dense(1,activation='sigmoid')\n",
        "])\n",
        "model.compile(optimizer='adam',loss='binary_crossentropy',metrics=['accuracy'])\n",
        "result = model.fit(X_train,y_train,batch_size=32,epochs=100)\n",
        "model.evaluate(X_test,y_test)"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "WARNING:tensorflow:From /usr/local/lib/python3.6/dist-packages/tensorflow/python/ops/init_ops.py:1251: calling VarianceScaling.__init__ (from tensorflow.python.ops.init_ops) with dtype is deprecated and will be removed in a future version.\n",
            "Instructions for updating:\n",
            "Call initializer instance with the dtype argument instead of passing it to the constructor\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "stream",
          "text": [
            "WARNING:tensorflow:From /usr/local/lib/python3.6/dist-packages/tensorflow/python/ops/init_ops.py:1251: calling VarianceScaling.__init__ (from tensorflow.python.ops.init_ops) with dtype is deprecated and will be removed in a future version.\n",
            "Instructions for updating:\n",
            "Call initializer instance with the dtype argument instead of passing it to the constructor\n"
          ],
          "name": "stderr"
        },
        {
          "output_type": "stream",
          "text": [
            "WARNING:tensorflow:From /usr/local/lib/python3.6/dist-packages/tensorflow/python/ops/nn_impl.py:180: add_dispatch_support.<locals>.wrapper (from tensorflow.python.ops.array_ops) is deprecated and will be removed in a future version.\n",
            "Instructions for updating:\n",
            "Use tf.where in 2.0, which has the same broadcast rule as np.where\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "stream",
          "text": [
            "WARNING:tensorflow:From /usr/local/lib/python3.6/dist-packages/tensorflow/python/ops/nn_impl.py:180: add_dispatch_support.<locals>.wrapper (from tensorflow.python.ops.array_ops) is deprecated and will be removed in a future version.\n",
            "Instructions for updating:\n",
            "Use tf.where in 2.0, which has the same broadcast rule as np.where\n"
          ],
          "name": "stderr"
        },
        {
          "output_type": "stream",
          "text": [
            "Epoch 1/100\n",
            "200/200 [==============================] - 1s 3ms/sample - loss: 0.6995 - acc: 0.5000\n",
            "Epoch 2/100\n",
            "200/200 [==============================] - 0s 527us/sample - loss: 0.6938 - acc: 0.4850\n",
            "Epoch 3/100\n",
            "200/200 [==============================] - 0s 568us/sample - loss: 0.7016 - acc: 0.5000\n",
            "Epoch 4/100\n",
            "200/200 [==============================] - 0s 512us/sample - loss: 0.6951 - acc: 0.4600\n",
            "Epoch 5/100\n",
            "200/200 [==============================] - 0s 513us/sample - loss: 0.6933 - acc: 0.5000\n",
            "Epoch 6/100\n",
            "200/200 [==============================] - 0s 509us/sample - loss: 0.6934 - acc: 0.4850\n",
            "Epoch 7/100\n",
            "200/200 [==============================] - 0s 498us/sample - loss: 0.6923 - acc: 0.5400\n",
            "Epoch 8/100\n",
            "200/200 [==============================] - 0s 513us/sample - loss: 0.6912 - acc: 0.5550\n",
            "Epoch 9/100\n",
            "200/200 [==============================] - 0s 518us/sample - loss: 0.6894 - acc: 0.5150\n",
            "Epoch 10/100\n",
            "200/200 [==============================] - 0s 500us/sample - loss: 0.6921 - acc: 0.4550\n",
            "Epoch 11/100\n",
            "200/200 [==============================] - 0s 493us/sample - loss: 0.6892 - acc: 0.5250\n",
            "Epoch 12/100\n",
            "200/200 [==============================] - 0s 525us/sample - loss: 0.6999 - acc: 0.5200\n",
            "Epoch 13/100\n",
            "200/200 [==============================] - 0s 567us/sample - loss: 0.6857 - acc: 0.5550\n",
            "Epoch 14/100\n",
            "200/200 [==============================] - 0s 490us/sample - loss: 0.6919 - acc: 0.5200\n",
            "Epoch 15/100\n",
            "200/200 [==============================] - 0s 527us/sample - loss: 0.6868 - acc: 0.5300\n",
            "Epoch 16/100\n",
            "200/200 [==============================] - 0s 488us/sample - loss: 0.6835 - acc: 0.5300\n",
            "Epoch 17/100\n",
            "200/200 [==============================] - 0s 495us/sample - loss: 0.6790 - acc: 0.5550\n",
            "Epoch 18/100\n",
            "200/200 [==============================] - 0s 488us/sample - loss: 0.6731 - acc: 0.5950\n",
            "Epoch 19/100\n",
            "200/200 [==============================] - 0s 466us/sample - loss: 0.6746 - acc: 0.6000\n",
            "Epoch 20/100\n",
            "200/200 [==============================] - 0s 526us/sample - loss: 0.6686 - acc: 0.5750\n",
            "Epoch 21/100\n",
            "200/200 [==============================] - 0s 506us/sample - loss: 0.6823 - acc: 0.5550\n",
            "Epoch 22/100\n",
            "200/200 [==============================] - 0s 554us/sample - loss: 0.6587 - acc: 0.6000\n",
            "Epoch 23/100\n",
            "200/200 [==============================] - 0s 512us/sample - loss: 0.6711 - acc: 0.6050\n",
            "Epoch 24/100\n",
            "200/200 [==============================] - 0s 523us/sample - loss: 0.6586 - acc: 0.6000\n",
            "Epoch 25/100\n",
            "200/200 [==============================] - 0s 478us/sample - loss: 0.6595 - acc: 0.6350\n",
            "Epoch 26/100\n",
            "200/200 [==============================] - 0s 500us/sample - loss: 0.6568 - acc: 0.5950\n",
            "Epoch 27/100\n",
            "200/200 [==============================] - 0s 495us/sample - loss: 0.6467 - acc: 0.6350\n",
            "Epoch 28/100\n",
            "200/200 [==============================] - 0s 538us/sample - loss: 0.6497 - acc: 0.6150\n",
            "Epoch 29/100\n",
            "200/200 [==============================] - 0s 516us/sample - loss: 0.6629 - acc: 0.5650\n",
            "Epoch 30/100\n",
            "200/200 [==============================] - 0s 542us/sample - loss: 0.6577 - acc: 0.5650\n",
            "Epoch 31/100\n",
            "200/200 [==============================] - 0s 487us/sample - loss: 0.6456 - acc: 0.6200\n",
            "Epoch 32/100\n",
            "200/200 [==============================] - 0s 512us/sample - loss: 0.6364 - acc: 0.6100\n",
            "Epoch 33/100\n",
            "200/200 [==============================] - 0s 494us/sample - loss: 0.6334 - acc: 0.6300\n",
            "Epoch 34/100\n",
            "200/200 [==============================] - 0s 526us/sample - loss: 0.6444 - acc: 0.6000\n",
            "Epoch 35/100\n",
            "200/200 [==============================] - 0s 523us/sample - loss: 0.6351 - acc: 0.6350\n",
            "Epoch 36/100\n",
            "200/200 [==============================] - 0s 612us/sample - loss: 0.6728 - acc: 0.6400\n",
            "Epoch 37/100\n",
            "200/200 [==============================] - 0s 501us/sample - loss: 0.6508 - acc: 0.5750\n",
            "Epoch 38/100\n",
            "200/200 [==============================] - 0s 540us/sample - loss: 0.6449 - acc: 0.5950\n",
            "Epoch 39/100\n",
            "200/200 [==============================] - 0s 466us/sample - loss: 0.6488 - acc: 0.6100\n",
            "Epoch 40/100\n",
            "200/200 [==============================] - 0s 538us/sample - loss: 0.6465 - acc: 0.6150\n",
            "Epoch 41/100\n",
            "200/200 [==============================] - 0s 483us/sample - loss: 0.6234 - acc: 0.6300\n",
            "Epoch 42/100\n",
            "200/200 [==============================] - 0s 499us/sample - loss: 0.6526 - acc: 0.5550\n",
            "Epoch 43/100\n",
            "200/200 [==============================] - 0s 542us/sample - loss: 0.6467 - acc: 0.5800\n",
            "Epoch 44/100\n",
            "200/200 [==============================] - 0s 507us/sample - loss: 0.6229 - acc: 0.6600\n",
            "Epoch 45/100\n",
            "200/200 [==============================] - 0s 509us/sample - loss: 0.6396 - acc: 0.6000\n",
            "Epoch 46/100\n",
            "200/200 [==============================] - 0s 511us/sample - loss: 0.6199 - acc: 0.6050\n",
            "Epoch 47/100\n",
            "200/200 [==============================] - 0s 525us/sample - loss: 0.6082 - acc: 0.6050\n",
            "Epoch 48/100\n",
            "200/200 [==============================] - 0s 547us/sample - loss: 0.6072 - acc: 0.6500\n",
            "Epoch 49/100\n",
            "200/200 [==============================] - 0s 530us/sample - loss: 0.6433 - acc: 0.6150\n",
            "Epoch 50/100\n",
            "200/200 [==============================] - 0s 519us/sample - loss: 0.6120 - acc: 0.6200\n",
            "Epoch 51/100\n",
            "200/200 [==============================] - 0s 499us/sample - loss: 0.6067 - acc: 0.6250\n",
            "Epoch 52/100\n",
            "200/200 [==============================] - 0s 565us/sample - loss: 0.6843 - acc: 0.5750\n",
            "Epoch 53/100\n",
            "200/200 [==============================] - 0s 500us/sample - loss: 0.6214 - acc: 0.6200\n",
            "Epoch 54/100\n",
            "200/200 [==============================] - 0s 492us/sample - loss: 0.6257 - acc: 0.6200\n",
            "Epoch 55/100\n",
            "200/200 [==============================] - 0s 541us/sample - loss: 0.6282 - acc: 0.6150\n",
            "Epoch 56/100\n",
            "200/200 [==============================] - 0s 547us/sample - loss: 0.6308 - acc: 0.6050\n",
            "Epoch 57/100\n",
            "200/200 [==============================] - 0s 525us/sample - loss: 0.6299 - acc: 0.6350\n",
            "Epoch 58/100\n",
            "200/200 [==============================] - 0s 503us/sample - loss: 0.6236 - acc: 0.5850\n",
            "Epoch 59/100\n",
            "200/200 [==============================] - 0s 507us/sample - loss: 0.6396 - acc: 0.6050\n",
            "Epoch 60/100\n",
            "200/200 [==============================] - 0s 536us/sample - loss: 0.6277 - acc: 0.5950\n",
            "Epoch 61/100\n",
            "200/200 [==============================] - 0s 507us/sample - loss: 0.6135 - acc: 0.5950\n",
            "Epoch 62/100\n",
            "200/200 [==============================] - 0s 537us/sample - loss: 0.6212 - acc: 0.6150\n",
            "Epoch 63/100\n",
            "200/200 [==============================] - 0s 497us/sample - loss: 0.6256 - acc: 0.5850\n",
            "Epoch 64/100\n",
            "200/200 [==============================] - 0s 510us/sample - loss: 0.6277 - acc: 0.6100\n",
            "Epoch 65/100\n",
            "200/200 [==============================] - 0s 498us/sample - loss: 0.6124 - acc: 0.6150\n",
            "Epoch 66/100\n",
            "200/200 [==============================] - 0s 522us/sample - loss: 0.6192 - acc: 0.5950\n",
            "Epoch 67/100\n",
            "200/200 [==============================] - 0s 511us/sample - loss: 0.6109 - acc: 0.6050\n",
            "Epoch 68/100\n",
            "200/200 [==============================] - 0s 506us/sample - loss: 0.5918 - acc: 0.6300\n",
            "Epoch 69/100\n",
            "200/200 [==============================] - 0s 493us/sample - loss: 0.6394 - acc: 0.5350\n",
            "Epoch 70/100\n",
            "200/200 [==============================] - 0s 512us/sample - loss: 0.6374 - acc: 0.5250\n",
            "Epoch 71/100\n",
            "200/200 [==============================] - 0s 493us/sample - loss: 0.6393 - acc: 0.6250\n",
            "Epoch 72/100\n",
            "200/200 [==============================] - 0s 506us/sample - loss: 0.6252 - acc: 0.6350\n",
            "Epoch 73/100\n",
            "200/200 [==============================] - 0s 530us/sample - loss: 0.6360 - acc: 0.6100\n",
            "Epoch 74/100\n",
            "200/200 [==============================] - 0s 514us/sample - loss: 0.6359 - acc: 0.6300\n",
            "Epoch 75/100\n",
            "200/200 [==============================] - 0s 507us/sample - loss: 0.6267 - acc: 0.6200\n",
            "Epoch 76/100\n",
            "200/200 [==============================] - 0s 512us/sample - loss: 0.6374 - acc: 0.5900\n",
            "Epoch 77/100\n",
            "200/200 [==============================] - 0s 517us/sample - loss: 0.6134 - acc: 0.6250\n",
            "Epoch 78/100\n",
            "200/200 [==============================] - 0s 538us/sample - loss: 0.6234 - acc: 0.6400\n",
            "Epoch 79/100\n",
            "200/200 [==============================] - 0s 536us/sample - loss: 0.6114 - acc: 0.6300\n",
            "Epoch 80/100\n",
            "200/200 [==============================] - 0s 485us/sample - loss: 0.6091 - acc: 0.6400\n",
            "Epoch 81/100\n",
            "200/200 [==============================] - 0s 526us/sample - loss: 0.6317 - acc: 0.6200\n",
            "Epoch 82/100\n",
            "200/200 [==============================] - 0s 581us/sample - loss: 0.6187 - acc: 0.6050\n",
            "Epoch 83/100\n",
            "200/200 [==============================] - 0s 485us/sample - loss: 0.6264 - acc: 0.6200\n",
            "Epoch 84/100\n",
            "200/200 [==============================] - 0s 540us/sample - loss: 0.6017 - acc: 0.6400\n",
            "Epoch 85/100\n",
            "200/200 [==============================] - 0s 517us/sample - loss: 0.6041 - acc: 0.6250\n",
            "Epoch 86/100\n",
            "200/200 [==============================] - 0s 520us/sample - loss: 0.6244 - acc: 0.6350\n",
            "Epoch 87/100\n",
            "200/200 [==============================] - 0s 519us/sample - loss: 0.6113 - acc: 0.6400\n",
            "Epoch 88/100\n",
            "200/200 [==============================] - 0s 498us/sample - loss: 0.6055 - acc: 0.6150\n",
            "Epoch 89/100\n",
            "200/200 [==============================] - 0s 501us/sample - loss: 0.5939 - acc: 0.6400\n",
            "Epoch 90/100\n",
            "200/200 [==============================] - 0s 495us/sample - loss: 0.5718 - acc: 0.6500\n",
            "Epoch 91/100\n",
            "200/200 [==============================] - 0s 563us/sample - loss: 0.5738 - acc: 0.6400\n",
            "Epoch 92/100\n",
            "200/200 [==============================] - 0s 509us/sample - loss: 0.5748 - acc: 0.6550\n",
            "Epoch 93/100\n",
            "200/200 [==============================] - 0s 484us/sample - loss: 0.5713 - acc: 0.6400\n",
            "Epoch 94/100\n",
            "200/200 [==============================] - 0s 553us/sample - loss: 0.5758 - acc: 0.6300\n",
            "Epoch 95/100\n",
            "200/200 [==============================] - 0s 490us/sample - loss: 0.5583 - acc: 0.6750\n",
            "Epoch 96/100\n",
            "200/200 [==============================] - 0s 505us/sample - loss: 0.5952 - acc: 0.6600\n",
            "Epoch 97/100\n",
            "200/200 [==============================] - 0s 515us/sample - loss: 0.6482 - acc: 0.6200\n",
            "Epoch 98/100\n",
            "200/200 [==============================] - 0s 488us/sample - loss: 0.6002 - acc: 0.6250\n",
            "Epoch 99/100\n",
            "200/200 [==============================] - 0s 526us/sample - loss: 0.6508 - acc: 0.6100\n",
            "Epoch 100/100\n",
            "200/200 [==============================] - 0s 516us/sample - loss: 0.6739 - acc: 0.6050\n",
            "200/200 [==============================] - 0s 237us/sample - loss: 0.9906 - acc: 0.5100\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "[0.9905993366241455, 0.51]"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 9
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "qnqTZ77wOolM",
        "colab_type": "text"
      },
      "source": [
        "## Quantum SVM\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "XRL9TepiM1pJ",
        "colab_type": "text"
      },
      "source": [
        "Here we are using the qiskit library by IBM which can be used to run quantum algorithms on IBM's quantum computers or quantum simulators. We chose to use qiskit because of aqua, a sub-library of the qiskit framework. Aqua has some pre-implemented quantum algorithms such as QSVM which makes it easy to run and try the algorithm for people who are currently exploring the domain of quantum computing.\n",
        "\n",
        "It may take time to run the Quantum SVM using the code below. This may be because of long queues for accessing the servers and running your request on the server.\n",
        "\n",
        "***Remember to store your IBMQ account token in the variable \"IBM_TOKEN\".***"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "h0kMfcTuOqop",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "# To enable on publication\n",
        "# IBM_TOKEN = 'YOUR_TOKEN'\n",
        "\n",
        "\n",
        "config_IBMQ = {\n",
        "    'hub': 'ibm-q',\n",
        "    'group': 'open',\n",
        "    'project': 'main',\n",
        "    'entanglement': 'full',\n",
        "    'token': IBM_TOKEN,\n",
        "    'server': 'qasm_simulator',\n",
        "    #'server': 'ibmq_qasm_simulator',\n",
        "    #'server': 'ibmq_16_melbourne',\n",
        "    'shots': 512,\n",
        "    'depth': 2,\n",
        "    'seed': 10598,\n",
        "    'local': True,\n",
        "    #'local': False,\n",
        "    'qbits': 10,\n",
        "    'records': 20\n",
        "}\n",
        "\n",
        "IBMQ.save_account(config_IBMQ['token'])\n",
        "if config_IBMQ['local']:\n",
        "  #IBMQ.load_accounts(hub=None)\n",
        "  feature_map = SecondOrderExpansion(feature_dimension=config['qbits'], depth=config_IBMQ['depth'], entanglement=config_IBMQ['entanglement'])\n",
        "  qsvm = QSVM(feature_map, training_input, test_input)\n",
        "  backend = BasicAer.get_backend(config_IBMQ['server'])\n",
        "  quantum_instance = QuantumInstance(backend, shots=config_IBMQ['shots'], seed_transpiler=config_IBMQ['seed'])\n",
        "  result = qsvm.run(quantum_instance)\n",
        "  print(result)\n",
        "else:\n",
        "  IBMQ.load_accounts(hub=None)\n",
        "  feature_map = SecondOrderExpansion(feature_dimension=config_IBMQ['qbits'], depth=config_IBMQ['depth'], entanglement=config_IBMQ['entanglement'])\n",
        "  qsvm = QSVM(feature_map, training_input, test_input)\n",
        "  provider = IBMQ.get_provider()\n",
        "  backend = provider.get_backend(config_IBMQ['server'])\n",
        "  quantum_instance = QuantumInstance(backend, shots=config_IBMQ['shots'], seed_transpiler=config_IBMQ['seed'], skip_qobj_validation=False)\n",
        "  result = qsvm.run(quantum_instance)\n",
        "  print(result)"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "jlaRpZJ3PScj",
        "colab_type": "text"
      },
      "source": [
        "## Quantum Neural Networks\n"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "LgCK9fltcUuZ",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "import pickle\n",
        "pickle.dump(X_train, open('X_train.pkl','wb'))\n",
        "pickle.dump(y_train, open('y_train.pkl','wb'))"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "CbV0ZK2T_koD",
        "colab_type": "code",
        "outputId": "fab71245-a7ea-4b7f-db5e-e7f916c1e72a",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        }
      },
      "source": [
        "%%time\n",
        "!git clone https://github.com/kkhosla96/QuantumImageClassification.git\n",
        "!mv -f QuantumImageClassification/* .\n",
        "!wget -O forest.tar.bz2 https://u8193507.ct.sendgrid.net/wf/click?upn=DVKTQCcBwH4CT-2FFvSWVL6I8GFSrqY0JvhZilrlFpk6lSWAqkKzSQZCfsqrBpen2137Fh1pWiF-2BeEq6eQOu9jPPGlL-2F0pZyLc4a9q2CpBuGQ-3D_L-2BfbHitP5D1pgZUrqcep5r44S8QKIpb03x-2Fd-2FtOrEiFZuJLLF0oXQyHcAAbvtcugKWy-2BlR4Kzqro3bSbn3SX1pbw86WGJCkYpnwepwXT-2BBsER4n0lqBTp-2FbCVK2VFLz1HAHsucc-2Be5Kb6-2Btt7BCFtccUQEkxCVAsR2gpMpGRxfkCTS6pntYaFEa9PTw4ZIsjBxJLM0eevDPZnIMydBiHP-2F7q-2BMti-2BTNGeGx7TfVbWwer-2FlFFdl0PQG0PM1lEooq7IxmDzNDAA1Yu6rd2qlqTWEkG-2BXejYm0jAOHOvj6n1XQC-2F6VpQjxO7An7tL-2BgM2vMLEgitquZjgqBZ5s2GaZ-2FH4RTzlk9gA02fg-2Bd-2FGo-2FiFZ812yd-2FBGuRRQBgiGQu71u2w-2FsQrKlhuQiXi-2B-2BsamcPeXY1yxPzmhz8LG4iYogCQA-3D\n",
        "!tar -xvf forest.tar.bz2"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Cloning into 'QuantumImageClassification'...\n",
            "remote: Enumerating objects: 130, done.\u001b[K\n",
            "remote: Counting objects: 100% (130/130), done.\u001b[K\n",
            "remote: Compressing objects: 100% (99/99), done.\u001b[K\n",
            "remote: Total 130 (delta 48), reused 97 (delta 21), pack-reused 0\u001b[K\n",
            "Receiving objects: 100% (130/130), 4.25 MiB | 12.23 MiB/s, done.\n",
            "Resolving deltas: 100% (48/48), done.\n",
            "--2019-09-13 12:51:22--  https://u8193507.ct.sendgrid.net/wf/click?upn=DVKTQCcBwH4CT-2FFvSWVL6I8GFSrqY0JvhZilrlFpk6lSWAqkKzSQZCfsqrBpen2137Fh1pWiF-2BeEq6eQOu9jPPGlL-2F0pZyLc4a9q2CpBuGQ-3D_L-2BfbHitP5D1pgZUrqcep5r44S8QKIpb03x-2Fd-2FtOrEiFZuJLLF0oXQyHcAAbvtcugKWy-2BlR4Kzqro3bSbn3SX1pbw86WGJCkYpnwepwXT-2BBsER4n0lqBTp-2FbCVK2VFLz1HAHsucc-2Be5Kb6-2Btt7BCFtccUQEkxCVAsR2gpMpGRxfkCTS6pntYaFEa9PTw4ZIsjBxJLM0eevDPZnIMydBiHP-2F7q-2BMti-2BTNGeGx7TfVbWwer-2FlFFdl0PQG0PM1lEooq7IxmDzNDAA1Yu6rd2qlqTWEkG-2BXejYm0jAOHOvj6n1XQC-2F6VpQjxO7An7tL-2BgM2vMLEgitquZjgqBZ5s2GaZ-2FH4RTzlk9gA02fg-2Bd-2FGo-2FiFZ812yd-2FBGuRRQBgiGQu71u2w-2FsQrKlhuQiXi-2B-2BsamcPeXY1yxPzmhz8LG4iYogCQA-3D\n",
            "Resolving u8193507.ct.sendgrid.net (u8193507.ct.sendgrid.net)... 167.89.123.16, 167.89.118.35\n",
            "Connecting to u8193507.ct.sendgrid.net (u8193507.ct.sendgrid.net)|167.89.123.16|:443... connected.\n",
            "HTTP request sent, awaiting response... 302 Found\n",
            "Location: https://downloads.rigetti.com/qcs-sdk/forest-sdk-linux-deb.tar.bz2 [following]\n",
            "--2019-09-13 12:51:22--  https://downloads.rigetti.com/qcs-sdk/forest-sdk-linux-deb.tar.bz2\n",
            "Resolving downloads.rigetti.com (downloads.rigetti.com)... 13.224.29.29, 13.224.29.15, 13.224.29.70, ...\n",
            "Connecting to downloads.rigetti.com (downloads.rigetti.com)|13.224.29.29|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 54161549 (52M) [application/x-bzip2]\n",
            "Saving to: ‘forest.tar.bz2’\n",
            "\n",
            "forest.tar.bz2      100%[===================>]  51.65M  79.4MB/s    in 0.7s    \n",
            "\n",
            "2019-09-13 12:51:23 (79.4 MB/s) - ‘forest.tar.bz2’ saved [54161549/54161549]\n",
            "\n",
            "forest-sdk-2.11.0-linux-deb/\n",
            "forest-sdk-2.11.0-linux-deb/forest-sdk-2.11.0-linux-deb.run\n",
            "forest-sdk-2.11.0-linux-deb/README.txt\n",
            "/bin/bash: /content/forest-sdk-2.8.0-linux-deb/forest-sdk-2.8.0-linux-deb.run: No such file or directory\n",
            "Collecting pyquil\n",
            "\u001b[?25l  Downloading https://files.pythonhosted.org/packages/9b/65/40d512a067ce1ec4278d6735518438f88ef4b22a21e0cacc73febd1b49a3/pyquil-2.11.0.tar.gz (173kB)\n",
            "\u001b[K     |████████████████████████████████| 174kB 5.0MB/s \n",
            "\u001b[?25hRequirement already satisfied, skipping upgrade: numpy in /usr/local/lib/python3.6/site-packages (from pyquil) (1.17.2)\n",
            "Collecting antlr4-python3-runtime (from pyquil)\n",
            "\u001b[?25l  Downloading https://files.pythonhosted.org/packages/29/14/8ac135ec7cc9db3f768e2d032776718c6b23f74e63543f0974b4873500b2/antlr4-python3-runtime-4.7.2.tar.gz (112kB)\n",
            "\u001b[K     |████████████████████████████████| 112kB 49.4MB/s \n",
            "\u001b[?25hRequirement already satisfied, skipping upgrade: requests in /usr/local/lib/python3.6/site-packages (from pyquil) (2.22.0)\n",
            "Requirement already satisfied, skipping upgrade: six in /usr/local/lib/python3.6/site-packages (from pyquil) (1.11.0)\n",
            "Requirement already satisfied, skipping upgrade: networkx>=2.0.0 in /usr/local/lib/python3.6/site-packages (from pyquil) (2.3)\n",
            "Collecting rpcq>=2.7.2 (from pyquil)\n",
            "  Downloading https://files.pythonhosted.org/packages/52/d8/62f38429b2ffd19c09a67271dfacf1a112e8ff73774e9a64453ed8c10110/rpcq-2.7.3.tar.gz\n",
            "Collecting immutables==0.6 (from pyquil)\n",
            "\u001b[?25l  Downloading https://files.pythonhosted.org/packages/93/85/14400a6cadf89d9cb369b3b7013baa766ea2a9fabe50d7a711ed8a6cc5d9/immutables-0.6-cp36-cp36m-manylinux1_x86_64.whl (70kB)\n",
            "\u001b[K     |████████████████████████████████| 71kB 25.4MB/s \n",
            "\u001b[?25hRequirement already satisfied, skipping upgrade: idna<2.9,>=2.5 in /usr/local/lib/python3.6/site-packages (from requests->pyquil) (2.6)\n",
            "Requirement already satisfied, skipping upgrade: urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1 in /usr/local/lib/python3.6/site-packages (from requests->pyquil) (1.22)\n",
            "Requirement already satisfied, skipping upgrade: chardet<3.1.0,>=3.0.2 in /usr/local/lib/python3.6/site-packages (from requests->pyquil) (3.0.4)\n",
            "Requirement already satisfied, skipping upgrade: certifi>=2017.4.17 in /usr/local/lib/python3.6/site-packages (from requests->pyquil) (2019.6.16)\n",
            "Requirement already satisfied, skipping upgrade: decorator>=4.3.0 in /usr/local/lib/python3.6/site-packages (from networkx>=2.0.0->pyquil) (4.4.0)\n",
            "Requirement already satisfied, skipping upgrade: future in /usr/local/lib/python3.6/site-packages (from rpcq>=2.7.2->pyquil) (0.17.1)\n",
            "Collecting msgpack>=0.6 (from rpcq>=2.7.2->pyquil)\n",
            "\u001b[?25l  Downloading https://files.pythonhosted.org/packages/92/7e/ae9e91c1bb8d846efafd1f353476e3fd7309778b582d2fb4cea4cc15b9a2/msgpack-0.6.1-cp36-cp36m-manylinux1_x86_64.whl (248kB)\n",
            "\u001b[K     |████████████████████████████████| 256kB 46.4MB/s \n",
            "\u001b[?25hCollecting python-rapidjson (from rpcq>=2.7.2->pyquil)\n",
            "\u001b[?25l  Downloading https://files.pythonhosted.org/packages/dd/32/7da92d409426dde28cdf844e7b1df1f8a5a790e0d53b45b850641946ddd6/python_rapidjson-0.8.0-cp36-cp36m-manylinux1_x86_64.whl (1.2MB)\n",
            "\u001b[K     |████████████████████████████████| 1.2MB 46.3MB/s \n",
            "\u001b[?25hCollecting pyzmq>=17 (from rpcq>=2.7.2->pyquil)\n",
            "\u001b[?25l  Downloading https://files.pythonhosted.org/packages/75/89/6f0ea51ffa9c2c00c0ab0460f137b16a5ab5b47e3b060c5b1fc9ca425836/pyzmq-18.1.0-cp36-cp36m-manylinux1_x86_64.whl (1.1MB)\n",
            "\u001b[K     |████████████████████████████████| 1.1MB 41.7MB/s \n",
            "\u001b[?25hCollecting ruamel.yaml (from rpcq>=2.7.2->pyquil)\n",
            "\u001b[?25l  Downloading https://files.pythonhosted.org/packages/fa/90/ecff85a2e9c497e2fa7142496e10233556b5137db5bd46f3f3b006935ca8/ruamel.yaml-0.16.5-py2.py3-none-any.whl (123kB)\n",
            "\u001b[K     |████████████████████████████████| 133kB 46.8MB/s \n",
            "\u001b[?25hRequirement already satisfied, skipping upgrade: typing in /usr/local/lib/python3.6/site-packages (from rpcq>=2.7.2->pyquil) (3.6.2)\n",
            "Collecting ruamel.yaml.clib>=0.1.2; platform_python_implementation == \"CPython\" and python_version < \"3.8\" (from ruamel.yaml->rpcq>=2.7.2->pyquil)\n",
            "\u001b[?25l  Downloading https://files.pythonhosted.org/packages/96/62/ed93cb8ae7e2ad8c5fe874e8027306aeee0c6a02c04fa015b5f99d14b3db/ruamel.yaml.clib-0.1.2-cp36-cp36m-manylinux1_x86_64.whl (549kB)\n",
            "\u001b[K     |████████████████████████████████| 552kB 49.2MB/s \n",
            "\u001b[?25hBuilding wheels for collected packages: pyquil, antlr4-python3-runtime, rpcq\n",
            "  Building wheel for pyquil (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for pyquil: filename=pyquil-2.11.0-cp36-none-any.whl size=199819 sha256=a96ce464384057aa16d4a75ea685c7c180691b2216febc447a4afecbb8dc141b\n",
            "  Stored in directory: /root/.cache/pip/wheels/eb/7b/21/bbf57d58f34c48af6aa8dbf6392ac3c3b40b6b55ddf6e5af8d\n",
            "  Building wheel for antlr4-python3-runtime (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for antlr4-python3-runtime: filename=antlr4_python3_runtime-4.7.2-cp36-none-any.whl size=140952 sha256=28466c14b8435329a3da28bbd2c2135ec75f2ab1812053e0b25a56fee1293f91\n",
            "  Stored in directory: /root/.cache/pip/wheels/03/65/47/b6700d23b7529a49513f7601b4f96174a4fa9efdbec03e7ea3\n",
            "  Building wheel for rpcq (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for rpcq: filename=rpcq-2.7.3-cp36-none-any.whl size=42293 sha256=399a07f43d552922230b25a765d5481d9cfe3e57e7de84211eb848f0711789fe\n",
            "  Stored in directory: /root/.cache/pip/wheels/e6/e3/62/24c8c5cd9d661ae478bcfd0030a8598ba9299ab27391bfed0e\n",
            "Successfully built pyquil antlr4-python3-runtime rpcq\n",
            "Installing collected packages: antlr4-python3-runtime, msgpack, python-rapidjson, pyzmq, ruamel.yaml.clib, ruamel.yaml, rpcq, immutables, pyquil\n",
            "  Found existing installation: pyzmq 16.0.3\n",
            "    Uninstalling pyzmq-16.0.3:\n",
            "      Successfully uninstalled pyzmq-16.0.3\n",
            "Successfully installed antlr4-python3-runtime-4.7.2 immutables-0.6 msgpack-0.6.1 pyquil-2.11.0 python-rapidjson-0.8.0 pyzmq-18.1.0 rpcq-2.7.3 ruamel.yaml-0.16.5 ruamel.yaml.clib-0.1.2\n",
            "Cloning into 'pennylane-forest'...\n",
            "remote: Enumerating objects: 81, done.\u001b[K\n",
            "remote: Counting objects: 100% (81/81), done.\u001b[K\n",
            "remote: Compressing objects: 100% (45/45), done.\u001b[K\n",
            "remote: Total 378 (delta 62), reused 45 (delta 36), pack-reused 297\u001b[K\n",
            "Receiving objects: 100% (378/378), 2.48 MiB | 7.44 MiB/s, done.\n",
            "Resolving deltas: 100% (252/252), done.\n",
            "/content/pennylane-forest\n",
            "Obtaining file:///content/pennylane-forest\n",
            "Requirement already satisfied: pyquil>=2.7 in /usr/local/lib/python3.6/site-packages (from PennyLane-Forest==0.4.0.dev0) (2.11.0)\n",
            "Requirement already satisfied: pennylane>=0.4 in /usr/local/lib/python3.6/site-packages (from PennyLane-Forest==0.4.0.dev0) (0.5.0)\n",
            "Requirement already satisfied: six in /usr/local/lib/python3.6/site-packages (from pyquil>=2.7->PennyLane-Forest==0.4.0.dev0) (1.11.0)\n",
            "Requirement already satisfied: rpcq>=2.7.2 in /usr/local/lib/python3.6/site-packages (from pyquil>=2.7->PennyLane-Forest==0.4.0.dev0) (2.7.3)\n",
            "Requirement already satisfied: requests in /usr/local/lib/python3.6/site-packages (from pyquil>=2.7->PennyLane-Forest==0.4.0.dev0) (2.22.0)\n",
            "Requirement already satisfied: immutables==0.6 in /usr/local/lib/python3.6/site-packages (from pyquil>=2.7->PennyLane-Forest==0.4.0.dev0) (0.6)\n",
            "Requirement already satisfied: networkx>=2.0.0 in /usr/local/lib/python3.6/site-packages (from pyquil>=2.7->PennyLane-Forest==0.4.0.dev0) (2.3)\n",
            "Requirement already satisfied: numpy in /usr/local/lib/python3.6/site-packages (from pyquil>=2.7->PennyLane-Forest==0.4.0.dev0) (1.17.2)\n",
            "Requirement already satisfied: antlr4-python3-runtime in /usr/local/lib/python3.6/site-packages (from pyquil>=2.7->PennyLane-Forest==0.4.0.dev0) (4.7.2)\n",
            "Requirement already satisfied: toml in /usr/local/lib/python3.6/site-packages (from pennylane>=0.4->PennyLane-Forest==0.4.0.dev0) (0.10.0)\n",
            "Requirement already satisfied: appdirs in /usr/local/lib/python3.6/site-packages (from pennylane>=0.4->PennyLane-Forest==0.4.0.dev0) (1.4.3)\n",
            "Requirement already satisfied: scipy in /usr/local/lib/python3.6/site-packages (from pennylane>=0.4->PennyLane-Forest==0.4.0.dev0) (1.0.0)\n",
            "Requirement already satisfied: autograd in /usr/local/lib/python3.6/site-packages (from pennylane>=0.4->PennyLane-Forest==0.4.0.dev0) (1.3)\n",
            "Requirement already satisfied: semantic-version==2.6 in /usr/local/lib/python3.6/site-packages (from pennylane>=0.4->PennyLane-Forest==0.4.0.dev0) (2.6.0)\n",
            "Requirement already satisfied: msgpack>=0.6 in /usr/local/lib/python3.6/site-packages (from rpcq>=2.7.2->pyquil>=2.7->PennyLane-Forest==0.4.0.dev0) (0.6.1)\n",
            "Requirement already satisfied: future in /usr/local/lib/python3.6/site-packages (from rpcq>=2.7.2->pyquil>=2.7->PennyLane-Forest==0.4.0.dev0) (0.17.1)\n",
            "Requirement already satisfied: typing in /usr/local/lib/python3.6/site-packages (from rpcq>=2.7.2->pyquil>=2.7->PennyLane-Forest==0.4.0.dev0) (3.6.2)\n",
            "Requirement already satisfied: pyzmq>=17 in /usr/local/lib/python3.6/site-packages (from rpcq>=2.7.2->pyquil>=2.7->PennyLane-Forest==0.4.0.dev0) (18.1.0)\n",
            "Requirement already satisfied: python-rapidjson in /usr/local/lib/python3.6/site-packages (from rpcq>=2.7.2->pyquil>=2.7->PennyLane-Forest==0.4.0.dev0) (0.8.0)\n",
            "Requirement already satisfied: ruamel.yaml in /usr/local/lib/python3.6/site-packages (from rpcq>=2.7.2->pyquil>=2.7->PennyLane-Forest==0.4.0.dev0) (0.16.5)\n",
            "Requirement already satisfied: chardet<3.1.0,>=3.0.2 in /usr/local/lib/python3.6/site-packages (from requests->pyquil>=2.7->PennyLane-Forest==0.4.0.dev0) (3.0.4)\n",
            "Requirement already satisfied: idna<2.9,>=2.5 in /usr/local/lib/python3.6/site-packages (from requests->pyquil>=2.7->PennyLane-Forest==0.4.0.dev0) (2.6)\n",
            "Requirement already satisfied: urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1 in /usr/local/lib/python3.6/site-packages (from requests->pyquil>=2.7->PennyLane-Forest==0.4.0.dev0) (1.22)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.6/site-packages (from requests->pyquil>=2.7->PennyLane-Forest==0.4.0.dev0) (2019.6.16)\n",
            "Requirement already satisfied: decorator>=4.3.0 in /usr/local/lib/python3.6/site-packages (from networkx>=2.0.0->pyquil>=2.7->PennyLane-Forest==0.4.0.dev0) (4.4.0)\n",
            "Requirement already satisfied: ruamel.yaml.clib>=0.1.2; platform_python_implementation == \"CPython\" and python_version < \"3.8\" in /usr/local/lib/python3.6/site-packages (from ruamel.yaml->rpcq>=2.7.2->pyquil>=2.7->PennyLane-Forest==0.4.0.dev0) (0.1.2)\n",
            "Installing collected packages: PennyLane-Forest\n",
            "  Running setup.py develop for PennyLane-Forest\n",
            "Successfully installed PennyLane-Forest\n",
            "/content\n",
            "CPU times: user 257 ms, sys: 207 ms, total: 464 ms\n",
            "Wall time: 24.2 s\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "LiKavxe_r4UF",
        "colab_type": "code",
        "outputId": "5e8edd35-f3d2-4b7a-db94-57eee703cb86",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 996
        }
      },
      "source": [
        "# VERIFY THE VERSION DOWNLOADED AND UPDATE THE NUMBER IN THE PATH\n",
        "!/content/forest-sdk-2.11.0-linux-deb/forest-sdk-2.11.0-linux-deb.run --quiet --accept\n",
        "!pip install --upgrade pyquil\n",
        "!git clone https://github.com/rigetti/pennylane-forest\n",
        "%cd pennylane-forest\n",
        "!pip install -e .\n",
        "%cd .."
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "                      End User License Agreement\n",
            "                Forest(TM) 2.0 Software Development Kit\n",
            "\n",
            "Thank you for your interest in Rigetti & Co, Inc., a Delaware corporation (commo\n",
            "nly referred to as Rigetti Computing) (“Rigetti”), and our Forest(TM) Software D\n",
            "evelopment Kit and associated downloadable materials (collectively, “Forest SDK”\n",
            "). Please read this End User License Agreement (“EULA”) carefully, as it constit\n",
            "utes a legally binding agreement governing your use of the Forest SDK (as such t\n",
            "erms are defined below).\n",
            "\n",
            "BY CLICKING “I ACCEPT,” OR BY OTHERWISE DOWNLOADING OR USING THE FOREST SDK, YOU\n",
            " AGREE THAT YOU HAVE READ, UNDERSTOOD AND AGREE TO BY BOUND BY THE FOLLOWING TER\n",
            "MS AND CONDITIONS. IF YOU DO NOT AGREE, YOU MAY NOT DOWNLOAD, REPRODUCE, INSTALL\n",
            ", OR OTHERWISE ACCESS THE FOREST SDK.\n",
            "\n",
            "If you are using the software on behalf of an entity, such as a company or organ\n",
            "ization, you are agreeing to this EULA on behalf of such entity (and you represe\n",
            "nt that you have the legal authority to do so), and the words “You,” “Your” and \n",
            "\"Yourself\" in this EULA will refer not only to you as an individual, but also to\n",
            " such entity and any other person using the software on such entity's behalf.\n",
            "\n",
            "Summaries of certain sections of this EULA have been included to help You unders\n",
            "tand them. These summaries are for Your convenience only; they will not affect, \n",
            "and they may not be used to interpret, the meaning of this EULA.\n",
            "\u001b[K\n",
            "1.      The QCS Service.\n",
            "Summary: This EULA grants a license only under the Forest Software Development K\n",
            "it, designed for use in conjunction with the Rigetti QCS Service.  The QCS Terms\n",
            " of Service still apply, and must be accepted separately when you sign up for an\n",
            " account.\n",
            "a.      QCS Access. In order to run your software on actual quantum hardware, yo\n",
            "u must also visit https://www.rigetti.com and sign up for a QCS account, unless \n",
            "one has been provided through your organization.\n",
            "b.      Order of Precedence. As between this EULA and the QCS Terms of Service, \n",
            "this EULA shall govern only use of the Forest SDK.  The Terms of Service shall g\n",
            "overn all other aspects of your online account, if applicable, except as amended\n",
            " or supplemented by other agreements.\n",
            "\u001b[K\n",
            "2.      Your License to the Forest SDK.\n",
            "Summary: Your license to use the Forest SDK is subject to certain restrictions.\n",
            "a.      License Grant. Rigetti grants You a non-exclusive, non-sublicenseable, n\n",
            "on-assignable, worldwide, terminable license, under its patents and copyrights f\n",
            "ully embodied in the Forest SDK to install and run a single instance of the Fore\n",
            "st SDK on a single computer.\n",
            "b.      Restrictions & Limitations. You acknowledge and agree that the license o\n",
            "f Section 2(a) above excludes the right to use the Forest SDK in connection with\n",
            " the design, development, marketing, promotion, sale, support or maintenance of \n",
            "any of the following: (i) any third-party quantum processors or other quantum co\n",
            "mputing hardware (collectively, \"Quantum Hardware\"); (ii) any third-party softwa\n",
            "re or hardware that virtualizes or otherwise simulates any Quantum Hardware; or \n",
            "(iii) any third-party hardware or software (including any tool, library or other\n",
            " computer code) that enables the development of software for, or the execution o\n",
            "f software on, any non-Rigetti Quantum Hardware, or otherwise facilitates access\n",
            " to or use of any non-Rigetti Quantum Hardware. You further acknowledge and agre\n",
            "e that the licenses of Section 2(a) exclude the right to use the Forest SDK in c\n",
            "onjunction with or running on top of any third-party Quantum Hardware or softwar\n",
            "\u001b[K MD5 checksums are OK."
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "7X96pqiZaGHR",
        "colab_type": "text"
      },
      "source": [
        "RESTART THE RUNTIME\n",
        "\n",
        "Click on \"**Runtime > Restart Runtime...**\""
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "93kvmMiwkVNg",
        "colab_type": "code",
        "outputId": "292307f2-203c-49c9-d5a5-e79b127f5c7b",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 221
        }
      },
      "source": [
        "!quilc --version\n",
        "!qvm --version\n",
        "!python -c \"import pyquil; print(pyquil.__version__)\"\n",
        "!nohup quilc -S &\n",
        "!nohup qvm -S &\n",
        "!ps -aux"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "quilc: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.28' not found (required by quilc)\n",
            "qvm: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.28' not found (required by qvm)\n",
            "2.11.0\n",
            "nohup: appending output to 'nohup.out'\n",
            "nohup: appending output to 'nohup.out'\n",
            "USER         PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND\n",
            "root           1  0.0  0.0  39192  6264 ?        Ss   12:36   0:00 /bin/bash -e \n",
            "root           8  0.3  0.4 687764 56316 ?        Sl   12:36   0:04 /tools/node/b\n",
            "root          25  0.5  0.7 406248 102188 ?       Sl   12:36   0:07 /usr/bin/pyth\n",
            "root         116  0.0  0.0  35884  4888 ?        Ss   12:37   0:00 tail -n +0 -F\n",
            "root        3106  9.9  1.1 618544 147472 ?       Ssl  12:59   0:01 /usr/bin/pyth\n",
            "root        3138  0.0  0.0  54836  6516 ?        R    13:00   0:00 ps -aux\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "xosfXwwDtSds",
        "colab_type": "code",
        "outputId": "66acc397-a525-47fa-8221-3980781ec57d",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 34
        }
      },
      "source": [
        "from pkg_resources import iter_entry_points\n",
        "print(sum(1 for _ in iter_entry_points(\"pennylane.plugins\")))\n",
        "for i in iter_entry_points(\"pennylane.plugins\"):\n",
        "  print(i)"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "0\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "W4VzayPTc57Z",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "X_train = pickle.load(open('X_train.pkl','r'))\n",
        "y_train = pickle.load(open('y_train.pkl','r'))"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "gWaAtJ3Aazvz",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "import pennylane as qml\n",
        "from pennylane import numpy as np\n",
        "import pennylane.optimize\n",
        "import os\n",
        "import time\n",
        "import random\n",
        "import pickle\n",
        "from tqdm import tqdm\n",
        "import matplotlib.pyplot as plt\n",
        "import torch\n",
        "import torch.nn as nn\n",
        "import torch.optim as optim\n",
        "from torch.autograd import Variable\n",
        "from torchsummary import summary\n",
        "import torchvision.datasets as dset\n",
        "import torchvision.transforms as transforms"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "1yCvFbI2PUXX",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "ENCODING_SIZE = 10\n",
        "NUM_QUBITS = 10\n",
        "EPOCHS = 100\n",
        "dev = qml.device(\"forest.qvm\", device=\"{}q-pyqvm\".format(NUM_QUBITS), shots=1600)\n",
        "#dev = qml.device('default.qubit', wires=NUM_QUBITS)\n",
        "from pennylane.ops import RX, RY, CNOT\n",
        "# x will be a length ENCODING_SIZE vector that represents\n",
        "# the encoding of a MNIST image\n",
        "# thetas is of size 2 * NUM_QUBITS\n",
        "@qml.qnode(dev)\n",
        "def circuit(x, thetas):\n",
        "    for i in range(ENCODING_SIZE):\n",
        "        RX(x[i], wires=i)\n",
        "    for i in range(NUM_QUBITS - 1):\n",
        "        CNOT(wires=[i, i+1])\n",
        "    for i in range(NUM_QUBITS):\n",
        "        RX(thetas[i], wires=i)\n",
        "    for i in range(NUM_QUBITS, 2 * NUM_QUBITS):\n",
        "        RY(thetas[i], wires=(i - NUM_QUBITS))\n",
        "    return tuple(qml.expval.PauliZ(wires=i) for i in range(NUM_QUBITS))\n",
        "\n",
        "def cost(X, actual_labels, thetas):\n",
        "    b = X.shape[0]\n",
        "    yhats = []\n",
        "    for i in range(b):\n",
        "        yhat = circuit(X[i], thetas)\n",
        "        yhats.append(yhat)\n",
        "    st = np.stack(yhats)\n",
        "    actual_class_vals = st[range(b), actual_labels]\n",
        "    shifted = st - np.max(st, axis=1)[:, np.newaxis]\n",
        "    the_sum = np.log(np.sum(np.exp(shifted), axis=1))\n",
        "    return np.mean(-actual_class_vals + the_sum)\n",
        "\n",
        "def accuracy(thetas, number_batches=3):\n",
        "    correct = 0\n",
        "    total = 0\n",
        "    for i, data in enumerate(X_test):\n",
        "        if i == number_batches:\n",
        "            break\n",
        "        predicted = []\n",
        "        for record in X_test:\n",
        "            ys = circuit(record, thetas)\n",
        "            predicted.append(np.argmax(ys))\n",
        "        correct += (np.asarray(predicted) == y_test).sum()\n",
        "        total += len(predicted)\n",
        "    return(correct / total)\n",
        "\n",
        "thetas = 2 * np.random.rand(2 * NUM_QUBITS) - 1\n",
        "print(thetas)\n",
        "print(thetas.dtype)\n",
        "\n",
        "#print(circuit(inputs[0].view(1, -1).numpy(), thetas))\n",
        "\n",
        "print(thetas.shape)\n",
        "start = time.time()\n",
        "grad = qml.grad(cost, argnum=2)(X_train, y_train, thetas) \n",
        "print(grad)\n",
        "print(grad.shape)\n",
        "print(time.time() - start)\n",
        "\n",
        "eta = .1\n",
        "start = time.time()\n",
        "accuracies = []\n",
        "for epoch in range(1):\n",
        "    for i in range(EPOCHS):\n",
        "        grad = qml.grad(cost, argnum=2)(X_train, y_train, thetas)\n",
        "        thetas -= eta * grad\n",
        "        acc = accuracy(thetas)\n",
        "        accuracies.append(acc)\n",
        "        print(\"%d batches took %f seconds and has accuracy %f\" % (i + 1, time.time() - start, acc))\n",
        "\n",
        "plt.plot(accuracies)\n",
        "plt.show()\n",
        "print(thetas)"
      ],
      "execution_count": 0,
      "outputs": []
    }
  ]
}